// @copilotkit/a2ui-renderer/dist/react-renderer/core/A2UIRenderer.cjs
const require_runtime = require('../../_virtual/_rolldown/runtime.cjs');
const require_useA2UI = require('../hooks/useA2UI.cjs');
const require_ComponentNode = require('./ComponentNode.cjs');
const require_utils = require('../lib/utils.cjs');
let react = require("react");
let react_jsx_runtime = require("react/jsx-runtime");

//#region src/react-renderer/core/A2UIRenderer.tsx
/** Default loading fallback - memoized to prevent recreation */
const DefaultLoadingFallback = (0, react.memo)(function DefaultLoadingFallback() {
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
		className: "a2ui-loading",
		style: {
			padding: "16px",
			opacity: .5
		},
		children: "Loading..."
	});
});
/**
* A2UIRenderer - renders an A2UI surface.
*
* This is the main entry point for rendering A2UI content in your React app.
* It reads the surface state from the A2UI store and renders the component tree.
*
* Memoized to prevent unnecessary re-renders when props haven't changed.
*
* @example
* ```tsx
* function App() {
*   return (
*     <A2UIProvider onAction={handleAction}>
*       <A2UIRenderer surfaceId="main" />
*     </A2UIProvider>
*   );
* }
* ```
*/
const A2UIRenderer = (0, react.memo)(function A2UIRenderer({ surfaceId, className, fallback = null, loadingFallback, registry }) {
	const { getSurface, version } = require_useA2UI.useA2UI();
	const surface = getSurface(surfaceId);
	const surfaceStyles = (0, react.useMemo)(() => {
		if (!surface?.styles) return {};
		const styles = {};
		for (const [key, value] of Object.entries(surface.styles)) switch (key) {
			case "primaryColor":
				styles["--p-100"] = "#ffffff";
				styles["--p-99"] = `color-mix(in srgb, ${value} 2%, white 98%)`;
				styles["--p-98"] = `color-mix(in srgb, ${value} 4%, white 96%)`;
				styles["--p-95"] = `color-mix(in srgb, ${value} 10%, white 90%)`;
				styles["--p-90"] = `color-mix(in srgb, ${value} 20%, white 80%)`;
				styles["--p-80"] = `color-mix(in srgb, ${value} 40%, white 60%)`;
				styles["--p-70"] = `color-mix(in srgb, ${value} 60%, white 40%)`;
				styles["--p-60"] = `color-mix(in srgb, ${value} 80%, white 20%)`;
				styles["--p-50"] = String(value);
				styles["--p-40"] = `color-mix(in srgb, ${value} 80%, black 20%)`;
				styles["--p-35"] = `color-mix(in srgb, ${value} 70%, black 30%)`;
				styles["--p-30"] = `color-mix(in srgb, ${value} 60%, black 40%)`;
				styles["--p-25"] = `color-mix(in srgb, ${value} 50%, black 50%)`;
				styles["--p-20"] = `color-mix(in srgb, ${value} 40%, black 60%)`;
				styles["--p-15"] = `color-mix(in srgb, ${value} 30%, black 70%)`;
				styles["--p-10"] = `color-mix(in srgb, ${value} 20%, black 80%)`;
				styles["--p-5"] = `color-mix(in srgb, ${value} 10%, black 90%)`;
				styles["--p-0"] = "#000000";
				break;
			case "font":
				styles["--font-family"] = String(value);
				styles["--font-family-flex"] = String(value);
				break;
		}
		return styles;
	}, [surface?.styles]);
	if (!surface || !surface.componentTree) return /* @__PURE__ */ (0, react_jsx_runtime.jsx)(react_jsx_runtime.Fragment, { children: fallback });
	const actualLoadingFallback = loadingFallback ?? /* @__PURE__ */ (0, react_jsx_runtime.jsx)(DefaultLoadingFallback, {});
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
		className: require_utils.cn("a2ui-surface", className),
		style: surfaceStyles,
		"data-surface-id": surfaceId,
		"data-version": version,
		children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(react.Suspense, {
			fallback: actualLoadingFallback,
			children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_ComponentNode.default, {
				node: surface.componentTree,
				surfaceId,
				registry
			})
		})
	});
});

//#endregion
exports.default = A2UIRenderer;
