// @copilotkit/a2ui-renderer/dist/react-renderer/core/A2UIProvider.cjs
const require_runtime = require('../../_virtual/_rolldown/runtime.cjs');
const require_ThemeContext = require('../theme/ThemeContext.cjs');
let react = require("react");
let _a2ui_lit_0_8 = require("@a2ui/lit/0.8");
let react_jsx_runtime = require("react/jsx-runtime");

//#region src/react-renderer/core/A2UIProvider.tsx
/**
* Context for stable actions (never changes reference, prevents re-renders).
* Components that only need to dispatch actions or read data won't re-render.
*/
const A2UIActionsContext = (0, react.createContext)(null);
/**
* Context for reactive state (changes trigger re-renders).
* Only components that need to react to state changes subscribe to this.
*/
const A2UIStateContext = (0, react.createContext)(null);
/**
* Provider component that sets up the A2UI context for descendant components.
*
* This provider uses a two-context architecture for performance:
* - A2UIActionsContext: Stable actions that never change (no re-renders)
* - A2UIStateContext: Reactive state that triggers re-renders when needed
*
* @example
* ```tsx
* function App() {
*   const handleAction = async (message) => {
*     const response = await fetch('/api/a2ui', {
*       method: 'POST',
*       body: JSON.stringify(message)
*     });
*     const newMessages = await response.json();
*   };
*
*   return (
*     <A2UIProvider onAction={handleAction}>
*       <A2UIRenderer surfaceId="main" />
*     </A2UIProvider>
*   );
* }
* ```
*/
function A2UIProvider({ onAction, theme, children }) {
	const processorRef = (0, react.useRef)(null);
	if (!processorRef.current) processorRef.current = _a2ui_lit_0_8.Data.createSignalA2uiMessageProcessor();
	const processor = processorRef.current;
	const [version, setVersion] = (0, react.useState)(0);
	const onActionRef = (0, react.useRef)(onAction ?? null);
	onActionRef.current = onAction ?? null;
	const actionsRef = (0, react.useRef)(null);
	if (!actionsRef.current) actionsRef.current = {
		processMessages: (messages) => {
			processor.processMessages(messages);
			setVersion((v) => v + 1);
		},
		setData: (node, path, value, surfaceId) => {
			processor.setData(node, path, value, surfaceId);
			setVersion((v) => v + 1);
		},
		dispatch: (message) => {
			if (onActionRef.current) onActionRef.current(message);
		},
		clearSurfaces: () => {
			processor.clearSurfaces();
			setVersion((v) => v + 1);
		},
		getSurface: (surfaceId) => {
			return processor.getSurfaces().get(surfaceId);
		},
		getSurfaces: () => {
			return processor.getSurfaces();
		},
		getData: (node, path, surfaceId) => {
			return processor.getData(node, path, surfaceId);
		},
		resolvePath: (path, dataContextPath) => {
			return processor.resolvePath(path, dataContextPath);
		}
	};
	const actions = actionsRef.current;
	const stateValue = (0, react.useMemo)(() => ({ version }), [version]);
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)(A2UIActionsContext.Provider, {
		value: actions,
		children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(A2UIStateContext.Provider, {
			value: stateValue,
			children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_ThemeContext.ThemeProvider, {
				theme,
				children
			})
		})
	});
}
/**
* Hook to access stable A2UI actions (won't cause re-renders).
* Use this when you only need to dispatch actions or read data.
*
* @returns Stable actions object
* @throws If used outside of an A2UIProvider
*/
function useA2UIActions() {
	const actions = (0, react.useContext)(A2UIActionsContext);
	if (!actions) throw new Error("useA2UIActions must be used within an A2UIProvider");
	return actions;
}
/**
* Hook to subscribe to A2UI state changes.
* Components using this will re-render when state changes.
*
* @returns Current version number
* @throws If used outside of an A2UIProvider
*/
function useA2UIState() {
	const state = (0, react.useContext)(A2UIStateContext);
	if (!state) throw new Error("useA2UIState must be used within an A2UIProvider");
	return state;
}
/**
* Hook to access the full A2UI context (actions + state).
* Components using this will re-render when state changes.
*
* @returns The A2UI context value
* @throws If used outside of an A2UIProvider
*/
function useA2UIContext() {
	const actions = useA2UIActions();
	const state = useA2UIState();
	return (0, react.useMemo)(() => ({
		...actions,
		processor: null,
		version: state.version,
		onAction: null
	}), [actions, state.version]);
}
/**
* @deprecated Use useA2UIContext instead. This alias exists for backward compatibility only.
*/
const useA2UIStore = useA2UIContext;
/**
* @deprecated This selector pattern does not provide performance benefits with React Context.
* Components will re-render on any context change regardless of what you select.
* Use useA2UIContext() or useA2UI() directly instead.
*
* @param selector - Function to select a slice of state
* @returns The selected state
*/
function useA2UIStoreSelector(selector) {
	return selector(useA2UIContext());
}

//#endregion
exports.A2UIProvider = A2UIProvider;
exports.useA2UIActions = useA2UIActions;
exports.useA2UIContext = useA2UIContext;
exports.useA2UIState = useA2UIState;
exports.useA2UIStore = useA2UIStore;
exports.useA2UIStoreSelector = useA2UIStoreSelector;
