// @copilotkit/a2ui-renderer/dist/react-renderer/core/ComponentNode.cjs
const require_runtime = require('../../_virtual/_rolldown/runtime.cjs');
const require_ComponentRegistry = require('../registry/ComponentRegistry.cjs');
let react = require("react");
let react_jsx_runtime = require("react/jsx-runtime");

//#region src/react-renderer/core/ComponentNode.tsx
/** Memoized loading fallback to avoid recreating on each render */
const LoadingFallback = (0, react.memo)(function LoadingFallback() {
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
		className: "a2ui-loading",
		style: {
			padding: "8px",
			opacity: .5
		},
		children: "Loading..."
	});
});
/**
* ComponentNode - dynamically renders an A2UI component based on its type.
*
* Looks up the component in the registry and renders it with the appropriate props.
* Supports lazy-loaded components via React.Suspense.
*
* No wrapper div is rendered - the component's root div (e.g., .a2ui-image) is the
* direct flex child, exactly matching Lit's structure where the :host element IS
* the flex item. Each component handles --weight CSS variable on its root div.
*
* Memoized to prevent unnecessary re-renders when parent updates but node hasn't changed.
*/
const ComponentNode = (0, react.memo)(function ComponentNode({ node, surfaceId, registry }) {
	const actualRegistry = registry ?? require_ComponentRegistry.ComponentRegistry.getInstance();
	const nodeType = node && typeof node === "object" && "type" in node ? node.type : null;
	const Component = (0, react.useMemo)(() => nodeType ? actualRegistry.get(nodeType) : null, [actualRegistry, nodeType]);
	if (!nodeType) {
		if (node) console.warn("[A2UI] Invalid component node (not resolved?):", node);
		return null;
	}
	if (!Component) {
		console.warn(`[A2UI] Unknown component type: ${nodeType}`);
		return null;
	}
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)(react.Suspense, {
		fallback: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(LoadingFallback, {}),
		children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(Component, {
			node,
			surfaceId
		})
	});
});

//#endregion
exports.default = ComponentNode;
