// @copilotkit/a2ui-renderer/dist/react-renderer/core/ComponentNode.mjs
import { ComponentRegistry } from "../registry/ComponentRegistry.mjs";
import { Suspense, memo, useMemo } from "react";
import { jsx } from "react/jsx-runtime";

//#region src/react-renderer/core/ComponentNode.tsx
/** Memoized loading fallback to avoid recreating on each render */
const LoadingFallback = memo(function LoadingFallback() {
	return /* @__PURE__ */ jsx("div", {
		className: "a2ui-loading",
		style: {
			padding: "8px",
			opacity: .5
		},
		children: "Loading..."
	});
});
/**
* ComponentNode - dynamically renders an A2UI component based on its type.
*
* Looks up the component in the registry and renders it with the appropriate props.
* Supports lazy-loaded components via React.Suspense.
*
* No wrapper div is rendered - the component's root div (e.g., .a2ui-image) is the
* direct flex child, exactly matching Lit's structure where the :host element IS
* the flex item. Each component handles --weight CSS variable on its root div.
*
* Memoized to prevent unnecessary re-renders when parent updates but node hasn't changed.
*/
const ComponentNode = memo(function ComponentNode({ node, surfaceId, registry }) {
	const actualRegistry = registry ?? ComponentRegistry.getInstance();
	const nodeType = node && typeof node === "object" && "type" in node ? node.type : null;
	const Component = useMemo(() => nodeType ? actualRegistry.get(nodeType) : null, [actualRegistry, nodeType]);
	if (!nodeType) {
		if (node) console.warn("[A2UI] Invalid component node (not resolved?):", node);
		return null;
	}
	if (!Component) {
		console.warn(`[A2UI] Unknown component type: ${nodeType}`);
		return null;
	}
	return /* @__PURE__ */ jsx(Suspense, {
		fallback: /* @__PURE__ */ jsx(LoadingFallback, {}),
		children: /* @__PURE__ */ jsx(Component, {
			node,
			surfaceId
		})
	});
});

//#endregion
export { ComponentNode as default };
