// @copilotkit/a2ui-renderer/dist/react-renderer/components/interactive/Button.cjs
const require_runtime = require('../../../_virtual/_rolldown/runtime.cjs');
const require_ComponentNode = require('../../core/ComponentNode.cjs');
const require_utils = require('../../theme/utils.cjs');
require('../../lib/utils.cjs');
const require_useA2UIComponent = require('../../hooks/useA2UIComponent.cjs');
let react = require("react");
let react_jsx_runtime = require("react/jsx-runtime");

//#region src/react-renderer/components/interactive/Button.tsx
/**
* Button component - a clickable element that triggers an action.
*
* Contains a child component (usually Text or Icon) and dispatches
* a user action when clicked.
*/
const Button = (0, react.memo)(function Button({ node, surfaceId }) {
	const { theme, sendAction } = require_useA2UIComponent.useA2UIComponent(node, surfaceId);
	const props = node.properties;
	const handleClick = (0, react.useCallback)(() => {
		if (props.action) sendAction(props.action);
	}, [props.action, sendAction]);
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
		className: "a2ui-button",
		style: node.weight !== void 0 ? { "--weight": node.weight } : {},
		children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)("button", {
			className: require_utils.classMapToString(theme.components.Button),
			style: require_utils.stylesToObject(theme.additionalStyles?.Button),
			onClick: handleClick,
			children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_ComponentNode.default, {
				node: props.child,
				surfaceId
			})
		})
	});
});

//#endregion
exports.default = Button;
