// @copilotkit/a2ui-renderer/dist/react-renderer/components/interactive/Button.mjs
import ComponentNode from "../../core/ComponentNode.mjs";
import { classMapToString, stylesToObject } from "../../theme/utils.mjs";
import "../../lib/utils.mjs";
import { useA2UIComponent } from "../../hooks/useA2UIComponent.mjs";
import { memo, useCallback } from "react";
import { jsx } from "react/jsx-runtime";

//#region src/react-renderer/components/interactive/Button.tsx
/**
* Button component - a clickable element that triggers an action.
*
* Contains a child component (usually Text or Icon) and dispatches
* a user action when clicked.
*/
const Button = memo(function Button({ node, surfaceId }) {
	const { theme, sendAction } = useA2UIComponent(node, surfaceId);
	const props = node.properties;
	const handleClick = useCallback(() => {
		if (props.action) sendAction(props.action);
	}, [props.action, sendAction]);
	return /* @__PURE__ */ jsx("div", {
		className: "a2ui-button",
		style: node.weight !== void 0 ? { "--weight": node.weight } : {},
		children: /* @__PURE__ */ jsx("button", {
			className: classMapToString(theme.components.Button),
			style: stylesToObject(theme.additionalStyles?.Button),
			onClick: handleClick,
			children: /* @__PURE__ */ jsx(ComponentNode, {
				node: props.child,
				surfaceId
			})
		})
	});
});

//#endregion
export { Button as default };
