// @copilotkit/a2ui-renderer/dist/react-renderer/components/interactive/CheckBox.mjs
import { classMapToString, stylesToObject } from "../../theme/utils.mjs";
import "../../lib/utils.mjs";
import { useA2UIComponent } from "../../hooks/useA2UIComponent.mjs";
import { memo, useCallback, useEffect, useId, useState } from "react";
import { jsx, jsxs } from "react/jsx-runtime";

//#region src/react-renderer/components/interactive/CheckBox.tsx
/**
* CheckBox component - a boolean toggle with a label.
*
* Supports two-way data binding for the checked state.
*/
const CheckBox = memo(function CheckBox({ node, surfaceId }) {
	const { theme, resolveString, resolveBoolean, setValue, getValue } = useA2UIComponent(node, surfaceId);
	const props = node.properties;
	const id = useId();
	const label = resolveString(props.label);
	const valuePath = props.value?.path;
	const [checked, setChecked] = useState(resolveBoolean(props.value) ?? false);
	useEffect(() => {
		if (valuePath) {
			const externalValue = getValue(valuePath);
			if (externalValue !== null && Boolean(externalValue) !== checked) setChecked(Boolean(externalValue));
		}
	}, [valuePath, getValue]);
	useEffect(() => {
		if (props.value?.literalBoolean !== void 0) setChecked(props.value.literalBoolean);
	}, [props.value?.literalBoolean]);
	const handleChange = useCallback((e) => {
		const newValue = e.target.checked;
		setChecked(newValue);
		if (valuePath) setValue(valuePath, newValue);
	}, [valuePath, setValue]);
	return /* @__PURE__ */ jsx("div", {
		className: "a2ui-checkbox",
		style: node.weight !== void 0 ? { "--weight": node.weight } : {},
		children: /* @__PURE__ */ jsxs("section", {
			className: classMapToString(theme.components.CheckBox.container),
			style: stylesToObject(theme.additionalStyles?.CheckBox),
			children: [/* @__PURE__ */ jsx("input", {
				type: "checkbox",
				id,
				checked,
				onChange: handleChange,
				className: classMapToString(theme.components.CheckBox.element)
			}), label && /* @__PURE__ */ jsx("label", {
				htmlFor: id,
				className: classMapToString(theme.components.CheckBox.label),
				children: label
			})]
		})
	});
});

//#endregion
export { CheckBox as default };
