// @copilotkit/a2ui-renderer/dist/react-renderer/components/interactive/CheckBox.cjs
const require_runtime = require('../../../_virtual/_rolldown/runtime.cjs');
const require_utils = require('../../theme/utils.cjs');
require('../../lib/utils.cjs');
const require_useA2UIComponent = require('../../hooks/useA2UIComponent.cjs');
let react = require("react");
let react_jsx_runtime = require("react/jsx-runtime");

//#region src/react-renderer/components/interactive/CheckBox.tsx
/**
* CheckBox component - a boolean toggle with a label.
*
* Supports two-way data binding for the checked state.
*/
const CheckBox = (0, react.memo)(function CheckBox({ node, surfaceId }) {
	const { theme, resolveString, resolveBoolean, setValue, getValue } = require_useA2UIComponent.useA2UIComponent(node, surfaceId);
	const props = node.properties;
	const id = (0, react.useId)();
	const label = resolveString(props.label);
	const valuePath = props.value?.path;
	const [checked, setChecked] = (0, react.useState)(resolveBoolean(props.value) ?? false);
	(0, react.useEffect)(() => {
		if (valuePath) {
			const externalValue = getValue(valuePath);
			if (externalValue !== null && Boolean(externalValue) !== checked) setChecked(Boolean(externalValue));
		}
	}, [valuePath, getValue]);
	(0, react.useEffect)(() => {
		if (props.value?.literalBoolean !== void 0) setChecked(props.value.literalBoolean);
	}, [props.value?.literalBoolean]);
	const handleChange = (0, react.useCallback)((e) => {
		const newValue = e.target.checked;
		setChecked(newValue);
		if (valuePath) setValue(valuePath, newValue);
	}, [valuePath, setValue]);
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
		className: "a2ui-checkbox",
		style: node.weight !== void 0 ? { "--weight": node.weight } : {},
		children: /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("section", {
			className: require_utils.classMapToString(theme.components.CheckBox.container),
			style: require_utils.stylesToObject(theme.additionalStyles?.CheckBox),
			children: [/* @__PURE__ */ (0, react_jsx_runtime.jsx)("input", {
				type: "checkbox",
				id,
				checked,
				onChange: handleChange,
				className: require_utils.classMapToString(theme.components.CheckBox.element)
			}), label && /* @__PURE__ */ (0, react_jsx_runtime.jsx)("label", {
				htmlFor: id,
				className: require_utils.classMapToString(theme.components.CheckBox.label),
				children: label
			})]
		})
	});
});

//#endregion
exports.default = CheckBox;
