// @copilotkit/a2ui-renderer/dist/react-renderer/components/interactive/DateTimeInput.mjs
import { classMapToString, stylesToObject } from "../../theme/utils.mjs";
import "../../lib/utils.mjs";
import { useA2UIComponent } from "../../hooks/useA2UIComponent.mjs";
import { memo, useCallback, useEffect, useId, useState } from "react";
import { jsx, jsxs } from "react/jsx-runtime";

//#region src/react-renderer/components/interactive/DateTimeInput.tsx
/**
* DateTimeInput component - a date and/or time picker.
*
* Supports enabling date, time, or both. Uses native HTML5 date/time inputs.
*/
const DateTimeInput = memo(function DateTimeInput({ node, surfaceId }) {
	const { theme, resolveString, setValue, getValue } = useA2UIComponent(node, surfaceId);
	const props = node.properties;
	const id = useId();
	const valuePath = props.value?.path;
	const initialValue = resolveString(props.value) ?? "";
	const enableDate = props.enableDate ?? true;
	const enableTime = props.enableTime ?? false;
	const [value, setLocalValue] = useState(initialValue);
	useEffect(() => {
		if (valuePath) {
			const externalValue = getValue(valuePath);
			if (externalValue !== null && String(externalValue) !== value) setLocalValue(String(externalValue));
		}
	}, [valuePath, getValue]);
	const handleChange = useCallback((e) => {
		const newValue = e.target.value;
		setLocalValue(newValue);
		if (valuePath) setValue(valuePath, newValue);
	}, [valuePath, setValue]);
	let inputType = "date";
	if (enableDate && enableTime) inputType = "datetime-local";
	else if (enableTime && !enableDate) inputType = "time";
	const getPlaceholderText = () => {
		if (enableDate && enableTime) return "Date & Time";
		else if (enableTime) return "Time";
		return "Date";
	};
	return /* @__PURE__ */ jsx("div", {
		className: "a2ui-datetime-input",
		style: node.weight !== void 0 ? { "--weight": node.weight } : {},
		children: /* @__PURE__ */ jsxs("section", {
			className: classMapToString(theme.components.DateTimeInput.container),
			children: [/* @__PURE__ */ jsx("label", {
				htmlFor: id,
				className: classMapToString(theme.components.DateTimeInput.label),
				children: getPlaceholderText()
			}), /* @__PURE__ */ jsx("input", {
				type: inputType,
				id,
				value,
				onChange: handleChange,
				className: classMapToString(theme.components.DateTimeInput.element),
				style: stylesToObject(theme.additionalStyles?.DateTimeInput)
			})]
		})
	});
});

//#endregion
export { DateTimeInput as default };
