// @copilotkit/a2ui-renderer/dist/react-renderer/components/interactive/DateTimeInput.cjs
const require_runtime = require('../../../_virtual/_rolldown/runtime.cjs');
const require_utils = require('../../theme/utils.cjs');
require('../../lib/utils.cjs');
const require_useA2UIComponent = require('../../hooks/useA2UIComponent.cjs');
let react = require("react");
let react_jsx_runtime = require("react/jsx-runtime");

//#region src/react-renderer/components/interactive/DateTimeInput.tsx
/**
* DateTimeInput component - a date and/or time picker.
*
* Supports enabling date, time, or both. Uses native HTML5 date/time inputs.
*/
const DateTimeInput = (0, react.memo)(function DateTimeInput({ node, surfaceId }) {
	const { theme, resolveString, setValue, getValue } = require_useA2UIComponent.useA2UIComponent(node, surfaceId);
	const props = node.properties;
	const id = (0, react.useId)();
	const valuePath = props.value?.path;
	const initialValue = resolveString(props.value) ?? "";
	const enableDate = props.enableDate ?? true;
	const enableTime = props.enableTime ?? false;
	const [value, setLocalValue] = (0, react.useState)(initialValue);
	(0, react.useEffect)(() => {
		if (valuePath) {
			const externalValue = getValue(valuePath);
			if (externalValue !== null && String(externalValue) !== value) setLocalValue(String(externalValue));
		}
	}, [valuePath, getValue]);
	const handleChange = (0, react.useCallback)((e) => {
		const newValue = e.target.value;
		setLocalValue(newValue);
		if (valuePath) setValue(valuePath, newValue);
	}, [valuePath, setValue]);
	let inputType = "date";
	if (enableDate && enableTime) inputType = "datetime-local";
	else if (enableTime && !enableDate) inputType = "time";
	const getPlaceholderText = () => {
		if (enableDate && enableTime) return "Date & Time";
		else if (enableTime) return "Time";
		return "Date";
	};
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
		className: "a2ui-datetime-input",
		style: node.weight !== void 0 ? { "--weight": node.weight } : {},
		children: /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("section", {
			className: require_utils.classMapToString(theme.components.DateTimeInput.container),
			children: [/* @__PURE__ */ (0, react_jsx_runtime.jsx)("label", {
				htmlFor: id,
				className: require_utils.classMapToString(theme.components.DateTimeInput.label),
				children: getPlaceholderText()
			}), /* @__PURE__ */ (0, react_jsx_runtime.jsx)("input", {
				type: inputType,
				id,
				value,
				onChange: handleChange,
				className: require_utils.classMapToString(theme.components.DateTimeInput.element),
				style: require_utils.stylesToObject(theme.additionalStyles?.DateTimeInput)
			})]
		})
	});
});

//#endregion
exports.default = DateTimeInput;
