// @copilotkit/a2ui-renderer/dist/react-renderer/components/interactive/MultipleChoice.cjs
const require_runtime = require('../../../_virtual/_rolldown/runtime.cjs');
const require_utils = require('../../theme/utils.cjs');
require('../../lib/utils.cjs');
const require_useA2UIComponent = require('../../hooks/useA2UIComponent.cjs');
let react = require("react");
let react_jsx_runtime = require("react/jsx-runtime");

//#region src/react-renderer/components/interactive/MultipleChoice.tsx
/**
* MultipleChoice component - a selection component using a dropdown.
*
* Renders a <select> element with options, matching the Lit renderer's behavior.
* Supports two-way data binding for the selected value.
*/
const MultipleChoice = (0, react.memo)(function MultipleChoice({ node, surfaceId }) {
	const { theme, resolveString, setValue } = require_useA2UIComponent.useA2UIComponent(node, surfaceId);
	const props = node.properties;
	const id = (0, react.useId)();
	const options = props.options ?? [];
	const selectionsPath = props.selections?.path;
	const description = resolveString(props.description) ?? "Select an item";
	const handleChange = (0, react.useCallback)((e) => {
		if (selectionsPath) setValue(selectionsPath, [e.target.value]);
	}, [selectionsPath, setValue]);
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
		className: "a2ui-multiplechoice",
		style: node.weight !== void 0 ? { "--weight": node.weight } : {},
		children: /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("section", {
			className: require_utils.classMapToString(theme.components.MultipleChoice.container),
			children: [/* @__PURE__ */ (0, react_jsx_runtime.jsx)("label", {
				htmlFor: id,
				className: require_utils.classMapToString(theme.components.MultipleChoice.label),
				children: description
			}), /* @__PURE__ */ (0, react_jsx_runtime.jsx)("select", {
				name: "data",
				id,
				className: require_utils.classMapToString(theme.components.MultipleChoice.element),
				style: require_utils.stylesToObject(theme.additionalStyles?.MultipleChoice),
				onChange: handleChange,
				children: options.map((option) => {
					const label = resolveString(option.label);
					return /* @__PURE__ */ (0, react_jsx_runtime.jsx)("option", {
						value: option.value,
						children: label
					}, option.value);
				})
			})]
		})
	});
});

//#endregion
exports.default = MultipleChoice;
