// @copilotkit/a2ui-renderer/dist/react-renderer/components/interactive/MultipleChoice.mjs
import { classMapToString, stylesToObject } from "../../theme/utils.mjs";
import "../../lib/utils.mjs";
import { useA2UIComponent } from "../../hooks/useA2UIComponent.mjs";
import { memo, useCallback, useId } from "react";
import { jsx, jsxs } from "react/jsx-runtime";

//#region src/react-renderer/components/interactive/MultipleChoice.tsx
/**
* MultipleChoice component - a selection component using a dropdown.
*
* Renders a <select> element with options, matching the Lit renderer's behavior.
* Supports two-way data binding for the selected value.
*/
const MultipleChoice = memo(function MultipleChoice({ node, surfaceId }) {
	const { theme, resolveString, setValue } = useA2UIComponent(node, surfaceId);
	const props = node.properties;
	const id = useId();
	const options = props.options ?? [];
	const selectionsPath = props.selections?.path;
	const description = resolveString(props.description) ?? "Select an item";
	const handleChange = useCallback((e) => {
		if (selectionsPath) setValue(selectionsPath, [e.target.value]);
	}, [selectionsPath, setValue]);
	return /* @__PURE__ */ jsx("div", {
		className: "a2ui-multiplechoice",
		style: node.weight !== void 0 ? { "--weight": node.weight } : {},
		children: /* @__PURE__ */ jsxs("section", {
			className: classMapToString(theme.components.MultipleChoice.container),
			children: [/* @__PURE__ */ jsx("label", {
				htmlFor: id,
				className: classMapToString(theme.components.MultipleChoice.label),
				children: description
			}), /* @__PURE__ */ jsx("select", {
				name: "data",
				id,
				className: classMapToString(theme.components.MultipleChoice.element),
				style: stylesToObject(theme.additionalStyles?.MultipleChoice),
				onChange: handleChange,
				children: options.map((option) => {
					const label = resolveString(option.label);
					return /* @__PURE__ */ jsx("option", {
						value: option.value,
						children: label
					}, option.value);
				})
			})]
		})
	});
});

//#endregion
export { MultipleChoice as default };
