// @copilotkit/a2ui-renderer/dist/react-renderer/components/interactive/Slider.mjs
import { classMapToString, stylesToObject } from "../../theme/utils.mjs";
import "../../lib/utils.mjs";
import { useA2UIComponent } from "../../hooks/useA2UIComponent.mjs";
import { memo, useCallback, useEffect, useId, useState } from "react";
import { jsx, jsxs } from "react/jsx-runtime";

//#region src/react-renderer/components/interactive/Slider.tsx
/**
* Slider component - a numeric value selector with a range.
*
* Supports two-way data binding for the value.
*/
const Slider = memo(function Slider({ node, surfaceId }) {
	const { theme, resolveNumber, resolveString, setValue, getValue } = useA2UIComponent(node, surfaceId);
	const props = node.properties;
	const id = useId();
	const valuePath = props.value?.path;
	const initialValue = resolveNumber(props.value) ?? 0;
	const minValue = props.minValue ?? 0;
	const maxValue = props.maxValue ?? 0;
	const [value, setLocalValue] = useState(initialValue);
	useEffect(() => {
		if (valuePath) {
			const externalValue = getValue(valuePath);
			if (externalValue !== null && Number(externalValue) !== value) setLocalValue(Number(externalValue));
		}
	}, [valuePath, getValue]);
	useEffect(() => {
		if (props.value?.literalNumber !== void 0) setLocalValue(props.value.literalNumber);
	}, [props.value?.literalNumber]);
	const handleChange = useCallback((e) => {
		const newValue = Number(e.target.value);
		setLocalValue(newValue);
		if (valuePath) setValue(valuePath, newValue);
	}, [valuePath, setValue]);
	const labelValue = props.label;
	const label = labelValue ? resolveString(labelValue) : "";
	return /* @__PURE__ */ jsx("div", {
		className: "a2ui-slider",
		style: node.weight !== void 0 ? { "--weight": node.weight } : {},
		children: /* @__PURE__ */ jsxs("section", {
			className: classMapToString(theme.components.Slider.container),
			children: [
				/* @__PURE__ */ jsx("label", {
					htmlFor: id,
					className: classMapToString(theme.components.Slider.label),
					children: label
				}),
				/* @__PURE__ */ jsx("input", {
					type: "range",
					id,
					name: "data",
					value,
					min: minValue,
					max: maxValue,
					onChange: handleChange,
					className: classMapToString(theme.components.Slider.element),
					style: stylesToObject(theme.additionalStyles?.Slider)
				}),
				/* @__PURE__ */ jsx("span", {
					className: classMapToString(theme.components.Slider.label),
					children: value
				})
			]
		})
	});
});

//#endregion
export { Slider as default };
