// @copilotkit/a2ui-renderer/dist/react-renderer/components/interactive/Slider.cjs
const require_runtime = require('../../../_virtual/_rolldown/runtime.cjs');
const require_utils = require('../../theme/utils.cjs');
require('../../lib/utils.cjs');
const require_useA2UIComponent = require('../../hooks/useA2UIComponent.cjs');
let react = require("react");
let react_jsx_runtime = require("react/jsx-runtime");

//#region src/react-renderer/components/interactive/Slider.tsx
/**
* Slider component - a numeric value selector with a range.
*
* Supports two-way data binding for the value.
*/
const Slider = (0, react.memo)(function Slider({ node, surfaceId }) {
	const { theme, resolveNumber, resolveString, setValue, getValue } = require_useA2UIComponent.useA2UIComponent(node, surfaceId);
	const props = node.properties;
	const id = (0, react.useId)();
	const valuePath = props.value?.path;
	const initialValue = resolveNumber(props.value) ?? 0;
	const minValue = props.minValue ?? 0;
	const maxValue = props.maxValue ?? 0;
	const [value, setLocalValue] = (0, react.useState)(initialValue);
	(0, react.useEffect)(() => {
		if (valuePath) {
			const externalValue = getValue(valuePath);
			if (externalValue !== null && Number(externalValue) !== value) setLocalValue(Number(externalValue));
		}
	}, [valuePath, getValue]);
	(0, react.useEffect)(() => {
		if (props.value?.literalNumber !== void 0) setLocalValue(props.value.literalNumber);
	}, [props.value?.literalNumber]);
	const handleChange = (0, react.useCallback)((e) => {
		const newValue = Number(e.target.value);
		setLocalValue(newValue);
		if (valuePath) setValue(valuePath, newValue);
	}, [valuePath, setValue]);
	const labelValue = props.label;
	const label = labelValue ? resolveString(labelValue) : "";
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
		className: "a2ui-slider",
		style: node.weight !== void 0 ? { "--weight": node.weight } : {},
		children: /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("section", {
			className: require_utils.classMapToString(theme.components.Slider.container),
			children: [
				/* @__PURE__ */ (0, react_jsx_runtime.jsx)("label", {
					htmlFor: id,
					className: require_utils.classMapToString(theme.components.Slider.label),
					children: label
				}),
				/* @__PURE__ */ (0, react_jsx_runtime.jsx)("input", {
					type: "range",
					id,
					name: "data",
					value,
					min: minValue,
					max: maxValue,
					onChange: handleChange,
					className: require_utils.classMapToString(theme.components.Slider.element),
					style: require_utils.stylesToObject(theme.additionalStyles?.Slider)
				}),
				/* @__PURE__ */ (0, react_jsx_runtime.jsx)("span", {
					className: require_utils.classMapToString(theme.components.Slider.label),
					children: value
				})
			]
		})
	});
});

//#endregion
exports.default = Slider;
