// @copilotkit/a2ui-renderer/dist/react-renderer/components/interactive/TextField.mjs
import { classMapToString, stylesToObject } from "../../theme/utils.mjs";
import "../../lib/utils.mjs";
import { useA2UIComponent } from "../../hooks/useA2UIComponent.mjs";
import { memo, useCallback, useEffect, useId, useState } from "react";
import { jsx, jsxs } from "react/jsx-runtime";

//#region src/react-renderer/components/interactive/TextField.tsx
/**
* TextField component - an input field for text entry.
*
* Supports various input types and two-way data binding.
*/
const TextField = memo(function TextField({ node, surfaceId }) {
	const { theme, resolveString, setValue, getValue } = useA2UIComponent(node, surfaceId);
	const props = node.properties;
	const id = useId();
	const label = resolveString(props.label);
	const textPath = props.text?.path;
	const initialValue = resolveString(props.text) ?? "";
	const fieldType = props.type;
	const validationRegexp = props.validationRegexp;
	const [value, setLocalValue] = useState(initialValue);
	const [_isValid, setIsValid] = useState(true);
	useEffect(() => {
		if (textPath) {
			const externalValue = getValue(textPath);
			if (externalValue !== null && String(externalValue) !== value) setLocalValue(String(externalValue));
		}
	}, [textPath, getValue]);
	const handleChange = useCallback((e) => {
		const newValue = e.target.value;
		setLocalValue(newValue);
		if (validationRegexp) setIsValid(new RegExp(validationRegexp).test(newValue));
		if (textPath) setValue(textPath, newValue);
	}, [
		validationRegexp,
		textPath,
		setValue
	]);
	const inputType = fieldType === "number" ? "number" : fieldType === "date" ? "date" : "text";
	const isTextArea = fieldType === "longText";
	return /* @__PURE__ */ jsx("div", {
		className: "a2ui-textfield",
		style: node.weight !== void 0 ? { "--weight": node.weight } : {},
		children: /* @__PURE__ */ jsxs("section", {
			className: classMapToString(theme.components.TextField.container),
			children: [label && /* @__PURE__ */ jsx("label", {
				htmlFor: id,
				className: classMapToString(theme.components.TextField.label),
				children: label
			}), isTextArea ? /* @__PURE__ */ jsx("textarea", {
				id,
				value,
				onChange: handleChange,
				placeholder: "Please enter a value",
				className: classMapToString(theme.components.TextField.element),
				style: stylesToObject(theme.additionalStyles?.TextField)
			}) : /* @__PURE__ */ jsx("input", {
				type: inputType,
				id,
				value,
				onChange: handleChange,
				placeholder: "Please enter a value",
				className: classMapToString(theme.components.TextField.element),
				style: stylesToObject(theme.additionalStyles?.TextField)
			})]
		})
	});
});

//#endregion
export { TextField as default };
