// @copilotkit/a2ui-renderer/dist/react-renderer/components/interactive/TextField.cjs
const require_runtime = require('../../../_virtual/_rolldown/runtime.cjs');
const require_utils = require('../../theme/utils.cjs');
require('../../lib/utils.cjs');
const require_useA2UIComponent = require('../../hooks/useA2UIComponent.cjs');
let react = require("react");
let react_jsx_runtime = require("react/jsx-runtime");

//#region src/react-renderer/components/interactive/TextField.tsx
/**
* TextField component - an input field for text entry.
*
* Supports various input types and two-way data binding.
*/
const TextField = (0, react.memo)(function TextField({ node, surfaceId }) {
	const { theme, resolveString, setValue, getValue } = require_useA2UIComponent.useA2UIComponent(node, surfaceId);
	const props = node.properties;
	const id = (0, react.useId)();
	const label = resolveString(props.label);
	const textPath = props.text?.path;
	const initialValue = resolveString(props.text) ?? "";
	const fieldType = props.type;
	const validationRegexp = props.validationRegexp;
	const [value, setLocalValue] = (0, react.useState)(initialValue);
	const [_isValid, setIsValid] = (0, react.useState)(true);
	(0, react.useEffect)(() => {
		if (textPath) {
			const externalValue = getValue(textPath);
			if (externalValue !== null && String(externalValue) !== value) setLocalValue(String(externalValue));
		}
	}, [textPath, getValue]);
	const handleChange = (0, react.useCallback)((e) => {
		const newValue = e.target.value;
		setLocalValue(newValue);
		if (validationRegexp) setIsValid(new RegExp(validationRegexp).test(newValue));
		if (textPath) setValue(textPath, newValue);
	}, [
		validationRegexp,
		textPath,
		setValue
	]);
	const inputType = fieldType === "number" ? "number" : fieldType === "date" ? "date" : "text";
	const isTextArea = fieldType === "longText";
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
		className: "a2ui-textfield",
		style: node.weight !== void 0 ? { "--weight": node.weight } : {},
		children: /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("section", {
			className: require_utils.classMapToString(theme.components.TextField.container),
			children: [label && /* @__PURE__ */ (0, react_jsx_runtime.jsx)("label", {
				htmlFor: id,
				className: require_utils.classMapToString(theme.components.TextField.label),
				children: label
			}), isTextArea ? /* @__PURE__ */ (0, react_jsx_runtime.jsx)("textarea", {
				id,
				value,
				onChange: handleChange,
				placeholder: "Please enter a value",
				className: require_utils.classMapToString(theme.components.TextField.element),
				style: require_utils.stylesToObject(theme.additionalStyles?.TextField)
			}) : /* @__PURE__ */ (0, react_jsx_runtime.jsx)("input", {
				type: inputType,
				id,
				value,
				onChange: handleChange,
				placeholder: "Please enter a value",
				className: require_utils.classMapToString(theme.components.TextField.element),
				style: require_utils.stylesToObject(theme.additionalStyles?.TextField)
			})]
		})
	});
});

//#endregion
exports.default = TextField;
