// @copilotkit/a2ui-renderer/dist/react-renderer/components/layout/Modal.mjs
import ComponentNode from "../../core/ComponentNode.mjs";
import { classMapToString, stylesToObject } from "../../theme/utils.mjs";
import "../../lib/utils.mjs";
import { useA2UIComponent } from "../../hooks/useA2UIComponent.mjs";
import { memo, useCallback, useEffect, useRef, useState } from "react";
import { jsx, jsxs } from "react/jsx-runtime";

//#region src/react-renderer/components/layout/Modal.tsx
/**
* Modal component - displays content in a dialog overlay.
*
* Matches Lit's rendering approach:
* - When closed: renders section with entry point child
* - When open: renders dialog with content child (entry point is replaced)
*
* The dialog is rendered in place (no portal) so it stays inside .a2ui-surface
* and CSS selectors work correctly. showModal() handles the top-layer overlay.
*/
const Modal = memo(function Modal({ node, surfaceId }) {
	const { theme } = useA2UIComponent(node, surfaceId);
	const props = node.properties;
	const [isOpen, setIsOpen] = useState(false);
	const dialogRef = useRef(null);
	const openModal = useCallback(() => {
		setIsOpen(true);
	}, []);
	const closeModal = useCallback(() => {
		setIsOpen(false);
	}, []);
	useEffect(() => {
		const dialog = dialogRef.current;
		if (!dialog) return;
		if (isOpen && !dialog.open) dialog.showModal();
		const handleClose = () => {
			setIsOpen(false);
		};
		dialog.addEventListener("close", handleClose);
		return () => dialog.removeEventListener("close", handleClose);
	}, [isOpen]);
	const handleBackdropClick = useCallback((e) => {
		if (e.target === e.currentTarget) closeModal();
	}, [closeModal]);
	const handleKeyDown = useCallback((e) => {
		if (e.key === "Escape") closeModal();
	}, [closeModal]);
	const hostStyle = node.weight !== void 0 ? { "--weight": node.weight } : {};
	if (!isOpen) return /* @__PURE__ */ jsx("div", {
		className: "a2ui-modal",
		style: hostStyle,
		children: /* @__PURE__ */ jsx("section", {
			onClick: openModal,
			style: { cursor: "pointer" },
			children: /* @__PURE__ */ jsx(ComponentNode, {
				node: props.entryPointChild,
				surfaceId
			})
		})
	});
	return /* @__PURE__ */ jsx("div", {
		className: "a2ui-modal",
		style: hostStyle,
		children: /* @__PURE__ */ jsx("dialog", {
			ref: dialogRef,
			className: classMapToString(theme.components.Modal.backdrop),
			onClick: handleBackdropClick,
			onKeyDown: handleKeyDown,
			children: /* @__PURE__ */ jsxs("section", {
				className: classMapToString(theme.components.Modal.element),
				style: stylesToObject(theme.additionalStyles?.Modal),
				children: [/* @__PURE__ */ jsx("div", {
					id: "controls",
					children: /* @__PURE__ */ jsx("button", {
						onClick: closeModal,
						"aria-label": "Close modal",
						children: /* @__PURE__ */ jsx("span", {
							className: "g-icon",
							children: "close"
						})
					})
				}), /* @__PURE__ */ jsx(ComponentNode, {
					node: props.contentChild,
					surfaceId
				})]
			})
		})
	});
});

//#endregion
export { Modal as default };
