// @copilotkit/a2ui-renderer/dist/react-renderer/components/layout/Modal.cjs
const require_runtime = require('../../../_virtual/_rolldown/runtime.cjs');
const require_ComponentNode = require('../../core/ComponentNode.cjs');
const require_utils = require('../../theme/utils.cjs');
require('../../lib/utils.cjs');
const require_useA2UIComponent = require('../../hooks/useA2UIComponent.cjs');
let react = require("react");
let react_jsx_runtime = require("react/jsx-runtime");

//#region src/react-renderer/components/layout/Modal.tsx
/**
* Modal component - displays content in a dialog overlay.
*
* Matches Lit's rendering approach:
* - When closed: renders section with entry point child
* - When open: renders dialog with content child (entry point is replaced)
*
* The dialog is rendered in place (no portal) so it stays inside .a2ui-surface
* and CSS selectors work correctly. showModal() handles the top-layer overlay.
*/
const Modal = (0, react.memo)(function Modal({ node, surfaceId }) {
	const { theme } = require_useA2UIComponent.useA2UIComponent(node, surfaceId);
	const props = node.properties;
	const [isOpen, setIsOpen] = (0, react.useState)(false);
	const dialogRef = (0, react.useRef)(null);
	const openModal = (0, react.useCallback)(() => {
		setIsOpen(true);
	}, []);
	const closeModal = (0, react.useCallback)(() => {
		setIsOpen(false);
	}, []);
	(0, react.useEffect)(() => {
		const dialog = dialogRef.current;
		if (!dialog) return;
		if (isOpen && !dialog.open) dialog.showModal();
		const handleClose = () => {
			setIsOpen(false);
		};
		dialog.addEventListener("close", handleClose);
		return () => dialog.removeEventListener("close", handleClose);
	}, [isOpen]);
	const handleBackdropClick = (0, react.useCallback)((e) => {
		if (e.target === e.currentTarget) closeModal();
	}, [closeModal]);
	const handleKeyDown = (0, react.useCallback)((e) => {
		if (e.key === "Escape") closeModal();
	}, [closeModal]);
	const hostStyle = node.weight !== void 0 ? { "--weight": node.weight } : {};
	if (!isOpen) return /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
		className: "a2ui-modal",
		style: hostStyle,
		children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)("section", {
			onClick: openModal,
			style: { cursor: "pointer" },
			children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_ComponentNode.default, {
				node: props.entryPointChild,
				surfaceId
			})
		})
	});
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
		className: "a2ui-modal",
		style: hostStyle,
		children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)("dialog", {
			ref: dialogRef,
			className: require_utils.classMapToString(theme.components.Modal.backdrop),
			onClick: handleBackdropClick,
			onKeyDown: handleKeyDown,
			children: /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("section", {
				className: require_utils.classMapToString(theme.components.Modal.element),
				style: require_utils.stylesToObject(theme.additionalStyles?.Modal),
				children: [/* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
					id: "controls",
					children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)("button", {
						onClick: closeModal,
						"aria-label": "Close modal",
						children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)("span", {
							className: "g-icon",
							children: "close"
						})
					})
				}), /* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_ComponentNode.default, {
					node: props.contentChild,
					surfaceId
				})]
			})
		})
	});
});

//#endregion
exports.default = Modal;
