// @copilotkit/a2ui-renderer/dist/react-renderer/components/layout/Tabs.mjs
import ComponentNode from "../../core/ComponentNode.mjs";
import { classMapToString, stylesToObject } from "../../theme/utils.mjs";
import { mergeClassMaps } from "../../lib/utils.mjs";
import { useA2UIComponent } from "../../hooks/useA2UIComponent.mjs";
import { memo, useState } from "react";
import { jsx, jsxs } from "react/jsx-runtime";

//#region src/react-renderer/components/layout/Tabs.tsx
/**
* Tabs component - displays content in switchable tabs.
*/
const Tabs = memo(function Tabs({ node, surfaceId }) {
	const { theme, resolveString } = useA2UIComponent(node, surfaceId);
	const props = node.properties;
	const [selectedIndex, setSelectedIndex] = useState(0);
	const tabItems = props.tabItems ?? [];
	return /* @__PURE__ */ jsx("div", {
		className: "a2ui-tabs",
		style: node.weight !== void 0 ? { "--weight": node.weight } : {},
		children: /* @__PURE__ */ jsxs("section", {
			className: classMapToString(theme.components.Tabs.container),
			style: stylesToObject(theme.additionalStyles?.Tabs),
			children: [/* @__PURE__ */ jsx("div", {
				id: "buttons",
				className: classMapToString(theme.components.Tabs.element),
				children: tabItems.map((tab, index) => {
					const title = resolveString(tab.title);
					const isSelected = index === selectedIndex;
					return /* @__PURE__ */ jsx("button", {
						disabled: isSelected,
						className: classMapToString(isSelected ? mergeClassMaps(theme.components.Tabs.controls.all, theme.components.Tabs.controls.selected) : theme.components.Tabs.controls.all),
						onClick: () => setSelectedIndex(index),
						children: title
					}, index);
				})
			}), tabItems[selectedIndex] && /* @__PURE__ */ jsx(ComponentNode, {
				node: tabItems[selectedIndex].child,
				surfaceId
			})]
		})
	});
});

//#endregion
export { Tabs as default };
