// @copilotkit/a2ui-renderer/dist/react-renderer/components/layout/Tabs.cjs
const require_runtime = require('../../../_virtual/_rolldown/runtime.cjs');
const require_ComponentNode = require('../../core/ComponentNode.cjs');
const require_utils = require('../../theme/utils.cjs');
const require_utils$1 = require('../../lib/utils.cjs');
const require_useA2UIComponent = require('../../hooks/useA2UIComponent.cjs');
let react = require("react");
let react_jsx_runtime = require("react/jsx-runtime");

//#region src/react-renderer/components/layout/Tabs.tsx
/**
* Tabs component - displays content in switchable tabs.
*/
const Tabs = (0, react.memo)(function Tabs({ node, surfaceId }) {
	const { theme, resolveString } = require_useA2UIComponent.useA2UIComponent(node, surfaceId);
	const props = node.properties;
	const [selectedIndex, setSelectedIndex] = (0, react.useState)(0);
	const tabItems = props.tabItems ?? [];
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
		className: "a2ui-tabs",
		style: node.weight !== void 0 ? { "--weight": node.weight } : {},
		children: /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("section", {
			className: require_utils.classMapToString(theme.components.Tabs.container),
			style: require_utils.stylesToObject(theme.additionalStyles?.Tabs),
			children: [/* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
				id: "buttons",
				className: require_utils.classMapToString(theme.components.Tabs.element),
				children: tabItems.map((tab, index) => {
					const title = resolveString(tab.title);
					const isSelected = index === selectedIndex;
					return /* @__PURE__ */ (0, react_jsx_runtime.jsx)("button", {
						disabled: isSelected,
						className: require_utils.classMapToString(isSelected ? require_utils$1.mergeClassMaps(theme.components.Tabs.controls.all, theme.components.Tabs.controls.selected) : theme.components.Tabs.controls.all),
						onClick: () => setSelectedIndex(index),
						children: title
					}, index);
				})
			}), tabItems[selectedIndex] && /* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_ComponentNode.default, {
				node: tabItems[selectedIndex].child,
				surfaceId
			})]
		})
	});
});

//#endregion
exports.default = Tabs;
