// @copilotkit/react-core/dist/index.umd.js
"use client";

(function(global, factory) {
  typeof exports === 'object' && typeof module !== 'undefined' ?  factory(exports, require('react'), require('@copilotkitnext/react'), require('@copilotkit/shared'), require('react-dom'), require('react/jsx-runtime'), require('react-markdown'), require('@copilotkit/runtime-client-gql'), require('@copilotkitnext/core'), require('@ag-ui/client')) :
  typeof define === 'function' && define.amd ? define(['exports', 'react', '@copilotkitnext/react', '@copilotkit/shared', 'react-dom', 'react/jsx-runtime', 'react-markdown', '@copilotkit/runtime-client-gql', '@copilotkitnext/core', '@ag-ui/client'], factory) :
  (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global.CopilotKitReactCore = {}), global.React,global.CopilotKitNextReact,global.CopilotKitShared,global.ReactDOM,global.ReactJsxRuntime,global.ReactMarkdown,global.CopilotKitRuntimeClientGQL,global.CopilotKitNextCore,global.AgUIClient));
})(this, function(exports, react, _copilotkitnext_react, _copilotkit_shared, react_dom, react_jsx_runtime, react_markdown, _copilotkit_runtime_client_gql, _copilotkitnext_core, _ag_ui_client) {
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
//#region \0rolldown/runtime.js
	var __create = Object.create;
	var __defProp = Object.defineProperty;
	var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
	var __getOwnPropNames = Object.getOwnPropertyNames;
	var __getProtoOf = Object.getPrototypeOf;
	var __hasOwnProp = Object.prototype.hasOwnProperty;
	var __copyProps = (to, from, except, desc) => {
		if (from && typeof from === "object" || typeof from === "function") {
			for (var keys = __getOwnPropNames(from), i = 0, n = keys.length, key; i < n; i++) {
				key = keys[i];
				if (!__hasOwnProp.call(to, key) && key !== except) {
					__defProp(to, key, {
						get: ((k) => from[k]).bind(null, key),
						enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
					});
				}
			}
		}
		return to;
	};
	var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
		value: mod,
		enumerable: true
	}) : target, mod));

//#endregion
react = __toESM(react);
react_markdown = __toESM(react_markdown);

//#region src/context/copilot-context.tsx
	const emptyCopilotContext$1 = {
		actions: {},
		setAction: () => {},
		removeAction: () => {},
		setRegisteredActions: () => "",
		removeRegisteredAction: () => {},
		chatComponentsCache: { current: {
			actions: {},
			coAgentStateRenders: {}
		} },
		getContextString: (documents, categories) => returnAndThrowInDebug(""),
		addContext: () => "",
		removeContext: () => {},
		getAllContext: () => [],
		getFunctionCallHandler: () => returnAndThrowInDebug(async () => {}),
		isLoading: false,
		setIsLoading: () => returnAndThrowInDebug(false),
		chatInstructions: "",
		setChatInstructions: () => returnAndThrowInDebug(""),
		additionalInstructions: [],
		setAdditionalInstructions: () => returnAndThrowInDebug([]),
		getDocumentsContext: (categories) => returnAndThrowInDebug([]),
		addDocumentContext: () => returnAndThrowInDebug(""),
		removeDocumentContext: () => {},
		copilotApiConfig: new class {
			get chatApiEndpoint() {
				throw new Error("Remember to wrap your app in a `<CopilotKit> {...} </CopilotKit>` !!!");
			}
			get headers() {
				return {};
			}
			get body() {
				return {};
			}
		}(),
		chatSuggestionConfiguration: {},
		addChatSuggestionConfiguration: () => {},
		removeChatSuggestionConfiguration: () => {},
		showDevConsole: false,
		coagentStates: {},
		setCoagentStates: () => {},
		coagentStatesRef: { current: {} },
		setCoagentStatesWithRef: () => {},
		agentSession: null,
		setAgentSession: () => {},
		forwardedParameters: {},
		agentLock: null,
		threadId: "",
		setThreadId: () => {},
		runId: null,
		setRunId: () => {},
		chatAbortControllerRef: { current: null },
		availableAgents: [],
		extensions: {},
		setExtensions: () => {},
		interruptActions: {},
		setInterruptAction: () => {},
		removeInterruptAction: () => {},
		interruptEventQueue: {},
		addInterruptEvent: () => {},
		resolveInterruptEvent: () => {},
		onError: () => {},
		bannerError: null,
		setBannerError: () => {},
		internalErrorHandlers: {},
		setInternalErrorHandler: () => {},
		removeInternalErrorHandler: () => {}
	};
	const CopilotContext = react.default.createContext(emptyCopilotContext$1);
	function useCopilotContext() {
		const context = react.default.useContext(CopilotContext);
		if (context === emptyCopilotContext$1) throw new Error("Remember to wrap your app in a `<CopilotKit> {...} </CopilotKit>` !!!");
		return context;
	}
	function returnAndThrowInDebug(_value) {
		throw new Error("Remember to wrap your app in a `<CopilotKit> {...} </CopilotKit>` !!!");
	}

//#endregion
//#region src/hooks/use-tree.ts
	const removeNode = (nodes, id) => {
		return nodes.reduce((result, node) => {
			if (node.id !== id) {
				const newNode = {
					...node,
					children: removeNode(node.children, id)
				};
				result.push(newNode);
			}
			return result;
		}, []);
	};
	const addNode = (nodes, newNode, parentId) => {
		if (!parentId) return [...nodes, newNode];
		return nodes.map((node) => {
			if (node.id === parentId) return {
				...node,
				children: [...node.children, newNode]
			};
			else if (node.children.length) return {
				...node,
				children: addNode(node.children, newNode, parentId)
			};
			return node;
		});
	};
	const treeIndentationRepresentation = (index, indentLevel) => {
		if (indentLevel === 0) return (index + 1).toString();
		else if (indentLevel === 1) return String.fromCharCode(65 + index);
		else if (indentLevel === 2) return String.fromCharCode(97 + index);
		else return "-";
	};
	const printNode = (node, prefix = "", indentLevel = 0) => {
		const indent = " ".repeat(3).repeat(indentLevel);
		const prefixPlusIndentLength = prefix.length + indent.length;
		const subsequentLinesPrefix = " ".repeat(prefixPlusIndentLength);
		const valueLines = node.value.split("\n");
		const outputFirstLine = `${indent}${prefix}${valueLines[0]}`;
		const outputSubsequentLines = valueLines.slice(1).map((line) => `${subsequentLinesPrefix}${line}`).join("\n");
		let output = `${outputFirstLine}\n`;
		if (outputSubsequentLines) output += `${outputSubsequentLines}\n`;
		const childPrePrefix = " ".repeat(prefix.length);
		node.children.forEach((child, index) => output += printNode(child, `${childPrePrefix}${treeIndentationRepresentation(index, indentLevel + 1)}. `, indentLevel + 1));
		return output;
	};
	function treeReducer(state, action) {
		switch (action.type) {
			case "ADD_NODE": {
				const { value, parentId, id: newNodeId } = action;
				const newNode = {
					id: newNodeId,
					value,
					children: [],
					categories: new Set(action.categories)
				};
				try {
					return addNode(state, newNode, parentId);
				} catch (error) {
					console.error(`Error while adding node with id ${newNodeId}: ${error}`);
					return state;
				}
			}
			case "REMOVE_NODE": return removeNode(state, action.id);
			default: return state;
		}
	}
	const useTree = () => {
		const [tree, dispatch] = (0, react.useReducer)(treeReducer, []);
		const addElement = (0, react.useCallback)((value, categories, parentId) => {
			const newNodeId = (0, _copilotkit_shared.randomId)();
			dispatch({
				type: "ADD_NODE",
				value,
				parentId,
				id: newNodeId,
				categories
			});
			return newNodeId;
		}, []);
		const removeElement = (0, react.useCallback)((id) => {
			dispatch({
				type: "REMOVE_NODE",
				id
			});
		}, []);
		const getAllElements = (0, react.useCallback)(() => {
			return tree;
		}, [tree]);
		return {
			tree,
			addElement,
			printTree: (0, react.useCallback)((categories) => {
				const categoriesSet = new Set(categories);
				let output = "";
				tree.forEach((node, index) => {
					if (!setsHaveIntersection$1(categoriesSet, node.categories)) return;
					if (index !== 0) output += "\n";
					output += printNode(node, `${treeIndentationRepresentation(index, 0)}. `);
				});
				return output;
			}, [tree]),
			removeElement,
			getAllElements
		};
	};
	function setsHaveIntersection$1(setA, setB) {
		const [smallerSet, largerSet] = setA.size <= setB.size ? [setA, setB] : [setB, setA];
		for (let item of smallerSet) if (largerSet.has(item)) return true;
		return false;
	}

//#endregion
//#region src/hooks/use-flat-category-store.ts
	const useFlatCategoryStore = () => {
		const [elements, dispatch] = (0, react.useReducer)(flatCategoryStoreReducer, /* @__PURE__ */ new Map());
		return {
			addElement: (0, react.useCallback)((value, categories) => {
				const newId = (0, _copilotkit_shared.randomId)();
				dispatch({
					type: "ADD_ELEMENT",
					value,
					id: newId,
					categories
				});
				return newId;
			}, []),
			removeElement: (0, react.useCallback)((id) => {
				dispatch({
					type: "REMOVE_ELEMENT",
					id
				});
			}, []),
			allElements: (0, react.useCallback)((categories) => {
				const categoriesSet = new Set(categories);
				const result = [];
				elements.forEach((element) => {
					if (setsHaveIntersection(categoriesSet, element.categories)) result.push(element.value);
				});
				return result;
			}, [elements])
		};
	};
	function flatCategoryStoreReducer(state, action) {
		switch (action.type) {
			case "ADD_ELEMENT": {
				const { value, id, categories } = action;
				const newElement = {
					id,
					value,
					categories: new Set(categories)
				};
				const newState = new Map(state);
				newState.set(id, newElement);
				return newState;
			}
			case "REMOVE_ELEMENT": {
				const newState = new Map(state);
				newState.delete(action.id);
				return newState;
			}
			default: return state;
		}
	}
	function setsHaveIntersection(setA, setB) {
		const [smallerSet, largerSet] = setA.size <= setB.size ? [setA, setB] : [setB, setA];
		for (let item of smallerSet) if (largerSet.has(item)) return true;
		return false;
	}

//#endregion
//#region src/context/copilot-messages-context.tsx
	const emptyCopilotContext = {
		messages: [],
		setMessages: () => [],
		suggestions: [],
		setSuggestions: () => []
	};
	const CopilotMessagesContext = react.default.createContext(emptyCopilotContext);
	function useCopilotMessagesContext() {
		const context = react.default.useContext(CopilotMessagesContext);
		if (context === emptyCopilotContext) throw new Error("A messages consuming component was not wrapped with `<CopilotMessages> {...} </CopilotMessages>`");
		return context;
	}

//#endregion
//#region src/components/toast/toast-provider.tsx
	const ToastContext = (0, react.createContext)(void 0);
	function getErrorSeverity(error) {
		if (error.severity) switch (error.severity) {
			case _copilotkit_shared.Severity.CRITICAL: return "critical";
			case _copilotkit_shared.Severity.WARNING: return "warning";
			case _copilotkit_shared.Severity.INFO: return "info";
			default: return "info";
		}
		const message = error.message.toLowerCase();
		if (message.includes("api key") || message.includes("401") || message.includes("unauthorized") || message.includes("authentication") || message.includes("incorrect api key")) return "critical";
		return "info";
	}
	function getErrorColors(severity) {
		switch (severity) {
			case "critical": return {
				background: "#fee2e2",
				border: "#dc2626",
				text: "#7f1d1d",
				icon: "#dc2626"
			};
			case "warning": return {
				background: "#fef3c7",
				border: "#d97706",
				text: "#78350f",
				icon: "#d97706"
			};
			case "info": return {
				background: "#dbeafe",
				border: "#2563eb",
				text: "#1e3a8a",
				icon: "#2563eb"
			};
		}
	}
	function useToast() {
		const context = (0, react.useContext)(ToastContext);
		if (!context) throw new Error("useToast must be used within a ToastProvider");
		return context;
	}
	function ToastProvider({ enabled, children }) {
		const [toasts, setToasts] = (0, react.useState)([]);
		const [bannerError, setBannerErrorState] = (0, react.useState)(null);
		const removeToast = (0, react.useCallback)((id) => {
			setToasts((prev) => prev.filter((toast) => toast.id !== id));
		}, []);
		const addToast = (0, react.useCallback)((toast) => {
			var _toast$id;
			if (!enabled) return;
			const id = (_toast$id = toast.id) !== null && _toast$id !== void 0 ? _toast$id : Math.random().toString(36).substring(2, 9);
			setToasts((currentToasts) => {
				if (currentToasts.find((toast) => toast.id === id)) return currentToasts;
				return [...currentToasts, {
					...toast,
					id
				}];
			});
			if (toast.duration) setTimeout(() => {
				removeToast(id);
			}, toast.duration);
		}, [enabled, removeToast]);
		const setBannerError = (0, react.useCallback)((error) => {
			if (!enabled && error !== null) return;
			setBannerErrorState(error);
		}, [enabled]);
		const value = {
			toasts,
			addToast,
			addGraphQLErrorsToast: (0, react.useCallback)((errors) => {
				console.warn("addGraphQLErrorsToast is deprecated. All errors now show as banners.");
			}, []),
			removeToast,
			enabled,
			bannerError,
			setBannerError
		};
		return /* @__PURE__ */ (0, react_jsx_runtime.jsxs)(ToastContext.Provider, {
			value,
			children: [bannerError && (() => {
				const colors = getErrorColors(getErrorSeverity(bannerError));
				return /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
					style: {
						position: "fixed",
						bottom: "20px",
						left: "50%",
						transform: "translateX(-50%)",
						zIndex: 9999,
						backgroundColor: colors.background,
						border: `1px solid ${colors.border}`,
						borderLeft: `4px solid ${colors.border}`,
						borderRadius: "8px",
						padding: "12px 16px",
						fontSize: "13px",
						boxShadow: "0 4px 12px rgba(0, 0, 0, 0.15)",
						backdropFilter: "blur(8px)",
						maxWidth: "min(90vw, 700px)",
						width: "100%",
						boxSizing: "border-box",
						overflow: "hidden"
					},
					children: /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
						style: {
							display: "flex",
							justifyContent: "space-between",
							alignItems: "center",
							gap: "10px"
						},
						children: [/* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
							style: {
								display: "flex",
								alignItems: "center",
								gap: "8px",
								flex: 1,
								minWidth: 0
							},
							children: [/* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", { style: {
								width: "12px",
								height: "12px",
								borderRadius: "50%",
								backgroundColor: colors.border,
								flexShrink: 0
							} }), /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
								style: {
									display: "flex",
									alignItems: "center",
									gap: "10px",
									flex: 1,
									minWidth: 0
								},
								children: [/* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
									style: {
										color: colors.text,
										lineHeight: "1.4",
										fontWeight: "400",
										fontSize: "13px",
										flex: 1,
										wordBreak: "break-all",
										overflowWrap: "break-word",
										maxWidth: "550px",
										overflow: "hidden",
										display: "-webkit-box",
										WebkitLineClamp: 10,
										WebkitBoxOrient: "vertical"
									},
									children: (() => {
										let message = bannerError.message;
										const jsonMatch = message.match(/'message':\s*'([^']+)'/);
										if (jsonMatch) return jsonMatch[1];
										message = message.split(" - ")[0];
										message = message.split(": Error code")[0];
										message = message.replace(/:\s*\d{3}$/, "");
										message = message.replace(/See more:.*$/g, "");
										message = message.trim();
										return message || "Configuration error occurred.";
									})()
								}), (() => {
									const message = bannerError.message;
									const markdownLinkRegex = /\[([^\]]+)\]\(([^)]+)\)/g;
									const plainUrlRegex = /(https?:\/\/[^\s)]+)/g;
									let url = null;
									let buttonText = "See More";
									const markdownMatch = markdownLinkRegex.exec(message);
									if (markdownMatch) {
										url = markdownMatch[2];
										buttonText = "See More";
									} else {
										const urlMatch = plainUrlRegex.exec(message);
										if (urlMatch) {
											url = urlMatch[0].replace(/[.,;:'"]*$/, "");
											buttonText = "See More";
										}
									}
									if (!url) return null;
									return /* @__PURE__ */ (0, react_jsx_runtime.jsx)("button", {
										onClick: () => window.open(url, "_blank", "noopener,noreferrer"),
										style: {
											background: colors.border,
											color: "white",
											border: "none",
											borderRadius: "5px",
											padding: "4px 10px",
											fontSize: "11px",
											fontWeight: "500",
											cursor: "pointer",
											transition: "all 0.2s ease",
											flexShrink: 0
										},
										onMouseEnter: (e) => {
											e.currentTarget.style.opacity = "0.9";
											e.currentTarget.style.transform = "translateY(-1px)";
										},
										onMouseLeave: (e) => {
											e.currentTarget.style.opacity = "1";
											e.currentTarget.style.transform = "translateY(0)";
										},
										children: buttonText
									});
								})()]
							})]
						}), /* @__PURE__ */ (0, react_jsx_runtime.jsx)("button", {
							onClick: () => setBannerError(null),
							style: {
								background: "transparent",
								border: "none",
								color: colors.text,
								cursor: "pointer",
								padding: "2px",
								borderRadius: "3px",
								fontSize: "14px",
								lineHeight: "1",
								opacity: .6,
								transition: "all 0.2s ease",
								flexShrink: 0
							},
							title: "Dismiss",
							onMouseEnter: (e) => {
								e.currentTarget.style.opacity = "1";
								e.currentTarget.style.background = "rgba(0, 0, 0, 0.05)";
							},
							onMouseLeave: (e) => {
								e.currentTarget.style.opacity = "0.6";
								e.currentTarget.style.background = "transparent";
							},
							children: "×"
						})]
					})
				});
			})(), children]
		});
	}

//#endregion
//#region src/utils/dev-console.ts
	function isLocalhost() {
		if (typeof window === "undefined") return false;
		return window.location.hostname === "localhost" || window.location.hostname === "127.0.0.1" || window.location.hostname === "0.0.0.0";
	}
	function shouldShowDevConsole(showDevConsole) {
		if (showDevConsole !== void 0) return showDevConsole;
		return isLocalhost();
	}

//#endregion
//#region src/components/copilot-provider/copilot-messages.tsx
/**
	* An internal context to separate the messages state (which is constantly changing) from the rest of CopilotKit context
	*/
	const MessagesTapContext = (0, react.createContext)(null);
	function useMessagesTap() {
		const tap = (0, react.useContext)(MessagesTapContext);
		if (!tap) throw new Error("useMessagesTap must be used inside <MessagesTapProvider>");
		return tap;
	}
	function MessagesTapProvider({ children }) {
		const messagesRef = (0, react.useRef)([]);
		const tapRef = (0, react.useRef)({
			getMessagesFromTap: () => messagesRef.current,
			updateTapMessages: (messages) => {
				messagesRef.current = messages;
			}
		});
		return /* @__PURE__ */ (0, react_jsx_runtime.jsx)(MessagesTapContext.Provider, {
			value: tapRef.current,
			children
		});
	}
	/**
	* CopilotKit messages context.
	*/
	function CopilotMessages({ children }) {
		const [messages, setMessages] = (0, react.useState)([]);
		(0, react.useRef)(void 0);
		(0, react.useRef)(void 0);
		(0, react.useRef)(void 0);
		const { updateTapMessages } = useMessagesTap();
		const { threadId, agentSession, showDevConsole, onError, copilotApiConfig } = useCopilotContext();
		const { setBannerError } = useToast();
		const traceUIError = (0, react.useCallback)(async (error, originalError) => {
			if (!onError || !copilotApiConfig.publicApiKey) return;
			try {
				await onError({
					type: "error",
					timestamp: Date.now(),
					context: {
						source: "ui",
						request: {
							operation: "loadAgentState",
							url: copilotApiConfig.chatApiEndpoint,
							startTime: Date.now()
						},
						technical: {
							environment: "browser",
							userAgent: typeof navigator !== "undefined" ? navigator.userAgent : void 0,
							stackTrace: originalError instanceof Error ? originalError.stack : void 0
						}
					},
					error
				});
			} catch (traceError) {
				console.error("Error in CopilotMessages onError handler:", traceError);
			}
		}, [
			onError,
			copilotApiConfig.publicApiKey,
			copilotApiConfig.chatApiEndpoint
		]);
		const createStructuredError = (gqlError) => {
			const extensions = gqlError.extensions;
			const originalError = extensions === null || extensions === void 0 ? void 0 : extensions.originalError;
			if (originalError === null || originalError === void 0 ? void 0 : originalError.stack) {
				if (originalError.stack.includes("CopilotApiDiscoveryError")) return new _copilotkit_shared.CopilotKitApiDiscoveryError({ message: originalError.message });
				if (originalError.stack.includes("CopilotKitRemoteEndpointDiscoveryError")) return new _copilotkit_shared.CopilotKitRemoteEndpointDiscoveryError({ message: originalError.message });
				if (originalError.stack.includes("CopilotKitAgentDiscoveryError")) return new _copilotkit_shared.CopilotKitAgentDiscoveryError({
					agentName: "",
					availableAgents: []
				});
			}
			const message = (originalError === null || originalError === void 0 ? void 0 : originalError.message) || gqlError.message;
			const code = extensions === null || extensions === void 0 ? void 0 : extensions.code;
			if (code) return new _copilotkit_shared.CopilotKitError({
				message,
				code
			});
			return null;
		};
		(0, react.useCallback)((error) => {
			var _error$graphQLErrors;
			if ((_error$graphQLErrors = error.graphQLErrors) === null || _error$graphQLErrors === void 0 ? void 0 : _error$graphQLErrors.length) {
				const graphQLErrors = error.graphQLErrors;
				const routeError = (gqlError) => {
					const extensions = gqlError.extensions;
					const visibility = extensions === null || extensions === void 0 ? void 0 : extensions.visibility;
					if (!shouldShowDevConsole(showDevConsole)) {
						console.error("CopilotKit Error (hidden in production):", gqlError.message);
						return;
					}
					if (visibility === _copilotkit_shared.ErrorVisibility.SILENT) {
						console.error("CopilotKit Silent Error:", gqlError.message);
						return;
					}
					const ckError = createStructuredError(gqlError);
					if (ckError) {
						setBannerError(ckError);
						traceUIError(ckError, gqlError);
					} else {
						const fallbackError = new _copilotkit_shared.CopilotKitError({
							message: gqlError.message,
							code: _copilotkit_shared.CopilotKitErrorCode.UNKNOWN
						});
						setBannerError(fallbackError);
						traceUIError(fallbackError, gqlError);
					}
				};
				graphQLErrors.forEach(routeError);
			} else if (!shouldShowDevConsole(showDevConsole)) console.error("CopilotKit Error (hidden in production):", error);
			else {
				const fallbackError = new _copilotkit_shared.CopilotKitError({
					message: (error === null || error === void 0 ? void 0 : error.message) || String(error),
					code: _copilotkit_shared.CopilotKitErrorCode.UNKNOWN
				});
				setBannerError(fallbackError);
				traceUIError(fallbackError, error);
			}
		}, [
			setBannerError,
			showDevConsole,
			traceUIError
		]);
		(0, react.useEffect)(() => {
			updateTapMessages(messages);
		}, [messages, updateTapMessages]);
		const memoizedChildren = (0, react.useMemo)(() => children, [children]);
		const [suggestions, setSuggestions] = (0, react.useState)([]);
		return /* @__PURE__ */ (0, react_jsx_runtime.jsx)(CopilotMessagesContext.Provider, {
			value: {
				messages,
				setMessages,
				suggestions,
				setSuggestions
			},
			children: memoizedChildren
		});
	}

//#endregion
//#region src/components/usage-banner.tsx
	function UsageBanner({ severity = _copilotkit_shared.Severity.CRITICAL, message = "", onClose, actions }) {
		if (!message || !severity) return null;
		const theme = {
			[_copilotkit_shared.Severity.INFO]: {
				bg: "#f8fafc",
				border: "#e2e8f0",
				text: "#475569",
				accent: "#3b82f6"
			},
			[_copilotkit_shared.Severity.WARNING]: {
				bg: "#fffbeb",
				border: "#fbbf24",
				text: "#92400e",
				accent: "#f59e0b"
			},
			[_copilotkit_shared.Severity.CRITICAL]: {
				bg: "#fef2f2",
				border: "#fecaca",
				text: "#dc2626",
				accent: "#ef4444"
			}
		}[severity];
		return /* @__PURE__ */ (0, react_jsx_runtime.jsxs)(react_jsx_runtime.Fragment, { children: [/* @__PURE__ */ (0, react_jsx_runtime.jsx)("style", { children: `
          @keyframes slideUp {
            from { opacity: 0; transform: translateX(-50%) translateY(8px); }
            to { opacity: 1; transform: translateX(-50%) translateY(0); }
          }
          
          .usage-banner {
            position: fixed;
            bottom: 24px;
            left: 50%;
            transform: translateX(-50%);
            width: min(600px, calc(100vw - 32px));
            z-index: 10000;
            animation: slideUp 0.2s cubic-bezier(0.16, 1, 0.3, 1);
          }
          
          .banner-content {
            background: linear-gradient(135deg, ${theme.bg} 0%, ${theme.bg}f5 100%);
            border: 1px solid ${theme.border};
            border-radius: 12px;
            padding: 18px 20px;
            box-shadow: 
              0 4px 24px rgba(0, 0, 0, 0.08),
              0 2px 8px rgba(0, 0, 0, 0.04),
              inset 0 1px 0 rgba(255, 255, 255, 0.7);
            display: flex;
            align-items: center;
            gap: 16px;
            font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', system-ui, sans-serif;
            backdrop-filter: blur(12px);
            position: relative;
            overflow: hidden;
          }
          
          .banner-content::before {
            content: '';
            position: absolute;
            top: 0;
            left: 0;
            right: 0;
            height: 1px;
            background: linear-gradient(90deg, transparent, ${theme.accent}40, transparent);
          }
          
          .banner-message {
            color: ${theme.text};
            font-size: 14px;
            line-height: 1.5;
            font-weight: 500;
            flex: 1;
            letter-spacing: -0.01em;
          }
          
          .close-btn {
            background: rgba(0, 0, 0, 0.05);
            border: none;
            color: ${theme.text};
            cursor: pointer;
            padding: 0;
            border-radius: 6px;
            opacity: 0.6;
            transition: all 0.15s cubic-bezier(0.16, 1, 0.3, 1);
            font-size: 14px;
            line-height: 1;
            flex-shrink: 0;
            width: 24px;
            height: 24px;
            display: flex;
            align-items: center;
            justify-content: center;
          }
          
          .close-btn:hover {
            opacity: 1;
            background: rgba(0, 0, 0, 0.08);
            transform: scale(1.05);
          }
          
          .btn-primary {
            background: linear-gradient(135deg, ${theme.accent} 0%, ${theme.accent}e6 100%);
            color: white;
            border: none;
            border-radius: 8px;
            padding: 10px 18px;
            font-size: 13px;
            font-weight: 600;
            cursor: pointer;
            transition: all 0.15s cubic-bezier(0.16, 1, 0.3, 1);
            font-family: inherit;
            flex-shrink: 0;
            box-shadow: 
              0 2px 8px ${theme.accent}30,
              inset 0 1px 0 rgba(255, 255, 255, 0.2);
            letter-spacing: -0.01em;
          }
          
          .btn-primary:hover {
            transform: translateY(-1px) scale(1.02);
            box-shadow: 
              0 4px 12px ${theme.accent}40,
              inset 0 1px 0 rgba(255, 255, 255, 0.25);
          }
          
          .btn-primary:active {
            transform: translateY(0) scale(0.98);
            transition: all 0.08s cubic-bezier(0.16, 1, 0.3, 1);
          }
          
          @media (max-width: 640px) {
            .usage-banner {
              width: calc(100vw - 24px);
            }
            
            .banner-content {
              padding: 16px;
              gap: 12px;
            }
            
            .banner-message {
              font-size: 13px;
              line-height: 1.45;
            }
            
            .btn-primary {
              padding: 8px 14px;
              font-size: 12px;
            }
            
            .close-btn {
              width: 22px;
              height: 22px;
              font-size: 12px;
            }
          }
        ` }), /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
			className: "usage-banner",
			children: /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
				className: "banner-content",
				children: [
					/* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
						className: "banner-message",
						children: message
					}),
					(actions === null || actions === void 0 ? void 0 : actions.primary) && /* @__PURE__ */ (0, react_jsx_runtime.jsx)("button", {
						className: "btn-primary",
						onClick: actions.primary.onClick,
						children: actions.primary.label
					}),
					onClose && /* @__PURE__ */ (0, react_jsx_runtime.jsx)("button", {
						className: "close-btn",
						onClick: onClose,
						title: "Close",
						children: "×"
					})
				]
			})
		})] });
	}
	const getErrorActions = (error) => {
		switch (error.code) {
			case _copilotkit_shared.CopilotKitErrorCode.MISSING_PUBLIC_API_KEY_ERROR: return { primary: {
				label: "Show me how",
				onClick: () => window.open("https://docs.copilotkit.ai/premium#how-do-i-get-access-to-premium-features", "_blank", "noopener,noreferrer")
			} };
			case _copilotkit_shared.CopilotKitErrorCode.UPGRADE_REQUIRED_ERROR: return { primary: {
				label: "Upgrade",
				onClick: () => window.open("https://cloud.copilotkit.ai", "_blank", "noopener,noreferrer")
			} };
			default: return;
		}
	};

//#endregion
//#region src/utils/suggestions-constants.ts
/**
	* Constants for suggestions retry logic
	*/
	const SUGGESTION_RETRY_CONFIG = {
		MAX_RETRIES: 3,
		COOLDOWN_MS: 5e3
	};

//#endregion
//#region src/lib/status-checker.ts
	const STATUS_CHECK_INTERVAL = 1e3 * 60 * 5;
	var StatusChecker = class {
		constructor() {
			this.activeKey = null;
			this.intervalId = null;
			this.instanceCount = 0;
			this.lastResponse = null;
		}
		async start(publicApiKey, onUpdate) {
			this.instanceCount++;
			if (this.activeKey === publicApiKey) return;
			if (this.intervalId) clearInterval(this.intervalId);
			const checkStatus = async () => {
				try {
					const response = await fetch(`${_copilotkit_shared.COPILOT_CLOUD_API_URL}/ciu`, {
						method: "GET",
						headers: { [_copilotkit_shared.COPILOT_CLOUD_PUBLIC_API_KEY_HEADER]: publicApiKey }
					}).then((response) => response.json());
					this.lastResponse = response;
					onUpdate === null || onUpdate === void 0 || onUpdate(response);
					return response;
				} catch (error) {
					return null;
				}
			};
			const initialResponse = await checkStatus();
			this.intervalId = setInterval(checkStatus, STATUS_CHECK_INTERVAL);
			this.activeKey = publicApiKey;
			return initialResponse;
		}
		getLastResponse() {
			return this.lastResponse;
		}
		stop() {
			this.instanceCount--;
			if (this.instanceCount === 0) {
				if (this.intervalId) {
					clearInterval(this.intervalId);
					this.intervalId = null;
					this.activeKey = null;
					this.lastResponse = null;
				}
			}
		}
	};

//#endregion
//#region src/components/toast/exclamation-mark-icon.tsx
	const ExclamationMarkIcon = ({ className, style }) => /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("svg", {
		xmlns: "http://www.w3.org/2000/svg",
		width: "24",
		height: "24",
		viewBox: "0 0 24 24",
		fill: "none",
		stroke: "currentColor",
		strokeWidth: "2",
		strokeLinecap: "round",
		strokeLinejoin: "round",
		className: `lucide lucide-circle-alert ${className ? className : ""}`,
		style,
		children: [
			/* @__PURE__ */ (0, react_jsx_runtime.jsx)("circle", {
				cx: "12",
				cy: "12",
				r: "10"
			}),
			/* @__PURE__ */ (0, react_jsx_runtime.jsx)("line", {
				x1: "12",
				x2: "12",
				y1: "8",
				y2: "12"
			}),
			/* @__PURE__ */ (0, react_jsx_runtime.jsx)("line", {
				x1: "12",
				x2: "12.01",
				y1: "16",
				y2: "16"
			})
		]
	});

//#endregion
//#region src/components/error-boundary/error-utils.tsx
	function ErrorToast({ errors }) {
		return /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
			style: {
				fontSize: "13px",
				maxWidth: "600px"
			},
			children: [errors.map((error, idx) => {
				var _error$extensions, _originalError$messag, _error$extensions2;
				const originalError = "extensions" in error ? (_error$extensions = error.extensions) === null || _error$extensions === void 0 ? void 0 : _error$extensions.originalError : {};
				const message = (_originalError$messag = originalError === null || originalError === void 0 ? void 0 : originalError.message) !== null && _originalError$messag !== void 0 ? _originalError$messag : error.message;
				const code = "extensions" in error ? (_error$extensions2 = error.extensions) === null || _error$extensions2 === void 0 ? void 0 : _error$extensions2.code : null;
				return /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
					style: {
						marginTop: idx === 0 ? 0 : 10,
						marginBottom: 14
					},
					children: [
						/* @__PURE__ */ (0, react_jsx_runtime.jsx)(ExclamationMarkIcon, { style: { marginBottom: 4 } }),
						code && /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
							style: {
								fontWeight: "600",
								marginBottom: 4
							},
							children: [
								"Copilot Runtime Error:",
								" ",
								/* @__PURE__ */ (0, react_jsx_runtime.jsx)("span", {
									style: {
										fontFamily: "monospace",
										fontWeight: "normal"
									},
									children: code
								})
							]
						}),
						/* @__PURE__ */ (0, react_jsx_runtime.jsx)(react_markdown.default, { children: message })
					]
				}, idx);
			}), /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
				style: {
					fontSize: "11px",
					opacity: .75
				},
				children: "NOTE: This error only displays during local development."
			})]
		});
	}
	function useErrorToast() {
		const { addToast } = useToast();
		return (0, react.useCallback)((errors) => {
			addToast({
				type: "error",
				id: errors.map((err) => {
					var _err$extensions;
					const message = "extensions" in err ? ((_err$extensions = err.extensions) === null || _err$extensions === void 0 || (_err$extensions = _err$extensions.originalError) === null || _err$extensions === void 0 ? void 0 : _err$extensions.message) || err.message : err.message;
					const stack = err.stack || "";
					return btoa(message + stack).slice(0, 32);
				}).join("|"),
				message: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(ErrorToast, { errors })
			});
		}, [addToast]);
	}
	function useAsyncCallback(callback, deps) {
		const addErrorToast = useErrorToast();
		return (0, react.useCallback)(async (...args) => {
			try {
				return await callback(...args);
			} catch (error) {
				console.error("Error in async callback:", error);
				addErrorToast([error]);
				throw error;
			}
		}, deps);
	}

//#endregion
//#region src/components/error-boundary/error-boundary.tsx
	const statusChecker = new StatusChecker();
	var CopilotErrorBoundary = class extends react.default.Component {
		constructor(props) {
			super(props);
			this.state = { hasError: false };
		}
		static getDerivedStateFromError(error) {
			return {
				hasError: true,
				error
			};
		}
		componentDidMount() {
			if (this.props.publicApiKey) statusChecker.start(this.props.publicApiKey, (newStatus) => {
				this.setState((prevState) => {
					var _prevState$status;
					if ((newStatus === null || newStatus === void 0 ? void 0 : newStatus.severity) !== ((_prevState$status = prevState.status) === null || _prevState$status === void 0 ? void 0 : _prevState$status.severity)) return { status: newStatus !== null && newStatus !== void 0 ? newStatus : void 0 };
					return null;
				});
			});
		}
		componentWillUnmount() {
			statusChecker.stop();
		}
		componentDidCatch(error, errorInfo) {
			console.error("CopilotKit Error:", error, errorInfo);
		}
		render() {
			if (this.state.hasError) {
				if (this.state.error instanceof _copilotkit_shared.CopilotKitError) {
					var _this$state$status$se, _this$state$status, _this$state$status$me, _this$state$status2;
					return /* @__PURE__ */ (0, react_jsx_runtime.jsxs)(react_jsx_runtime.Fragment, { children: [this.props.children, this.props.showUsageBanner && /* @__PURE__ */ (0, react_jsx_runtime.jsx)(UsageBanner, {
						severity: (_this$state$status$se = (_this$state$status = this.state.status) === null || _this$state$status === void 0 ? void 0 : _this$state$status.severity) !== null && _this$state$status$se !== void 0 ? _this$state$status$se : this.state.error.severity,
						message: (_this$state$status$me = (_this$state$status2 = this.state.status) === null || _this$state$status2 === void 0 ? void 0 : _this$state$status2.message) !== null && _this$state$status$me !== void 0 ? _this$state$status$me : this.state.error.message,
						actions: getErrorActions(this.state.error)
					})] });
				}
				throw this.state.error;
			}
			return this.props.children;
		}
	};

//#endregion
//#region src/context/coagent-state-renders-context.tsx
	const CoAgentStateRendersContext = (0, react.createContext)(void 0);
	function CoAgentStateRendersProvider({ children }) {
		const [coAgentStateRenders, setCoAgentStateRenders] = (0, react.useState)({});
		const setCoAgentStateRender = (0, react.useCallback)((id, stateRender) => {
			setCoAgentStateRenders((prevPoints) => ({
				...prevPoints,
				[id]: stateRender
			}));
		}, []);
		const removeCoAgentStateRender = (0, react.useCallback)((id) => {
			setCoAgentStateRenders((prevPoints) => {
				const newPoints = { ...prevPoints };
				delete newPoints[id];
				return newPoints;
			});
		}, []);
		const claimsRef = (0, react.useRef)({});
		return /* @__PURE__ */ (0, react_jsx_runtime.jsx)(CoAgentStateRendersContext.Provider, {
			value: {
				coAgentStateRenders,
				setCoAgentStateRender,
				removeCoAgentStateRender,
				claimsRef
			},
			children
		});
	}
	function useCoAgentStateRenders() {
		const context = (0, react.useContext)(CoAgentStateRendersContext);
		if (!context) throw new Error("useCoAgentStateRenders must be used within CoAgentStateRendersProvider");
		return context;
	}

//#endregion
//#region src/context/threads-context.tsx
	const ThreadsContext = (0, react.createContext)(void 0);
	function ThreadsProvider({ children, threadId: explicitThreadId }) {
		const [internalThreadId, setThreadId] = (0, react.useState)(() => (0, _copilotkit_shared.randomUUID)());
		const threadId = explicitThreadId !== null && explicitThreadId !== void 0 ? explicitThreadId : internalThreadId;
		return /* @__PURE__ */ (0, react_jsx_runtime.jsx)(ThreadsContext.Provider, {
			value: {
				threadId,
				setThreadId
			},
			children
		});
	}
	function useThreads() {
		const context = (0, react.useContext)(ThreadsContext);
		if (!context) throw new Error("useThreads must be used within ThreadsProvider");
		return context;
	}

//#endregion
//#region src/hooks/use-coagent-state-render-bridge.helpers.ts
	let RenderStatus = /* @__PURE__ */ function(RenderStatus) {
		RenderStatus["InProgress"] = "inProgress";
		RenderStatus["Complete"] = "complete";
		return RenderStatus;
	}({});
	let ClaimAction = /* @__PURE__ */ function(ClaimAction) {
		ClaimAction["Create"] = "create";
		ClaimAction["Override"] = "override";
		ClaimAction["Existing"] = "existing";
		ClaimAction["Block"] = "block";
		return ClaimAction;
	}({});
	function getStateWithoutConstantKeys(state) {
		if (!state) return {};
		const { messages, tools, copilotkit, ...stateWithoutConstantKeys } = state;
		return stateWithoutConstantKeys;
	}
	function areStatesEquals(a, b) {
		if (a && !b || !a && b) return false;
		const { messages, tools, copilotkit, ...aWithoutConstantKeys } = a;
		const { messages: bMessages, tools: bTools, copilotkit: bCopilotkit, ...bWithoutConstantKeys } = b;
		return JSON.stringify(aWithoutConstantKeys) === JSON.stringify(bWithoutConstantKeys);
	}
	function isPlaceholderMessageId(messageId) {
		return !!messageId && messageId.startsWith("coagent-state-render-");
	}
	function isPlaceholderMessageName(messageName) {
		return messageName === "coagent-state-render";
	}
	function readCachedMessageEntry(entry) {
		if (!entry || typeof entry !== "object") return {
			snapshot: entry,
			runId: void 0
		};
		return {
			snapshot: "snapshot" in entry ? entry.snapshot : entry,
			runId: "runId" in entry ? entry.runId : void 0
		};
	}
	function getEffectiveRunId({ existingClaimRunId, cachedMessageRunId, runId }) {
		return existingClaimRunId || cachedMessageRunId || runId || "pending";
	}
	/**
	* Resolve whether a message can claim a render slot.
	* This is a pure decision function; the caller applies claim mutations.
	*/
	function resolveClaim({ claims, context, stateSnapshot }) {
		const { messageId, stateRenderId, runId, messageIndex } = context;
		const existing = claims[messageId];
		if (existing) {
			const canRender = existing.stateRenderId === stateRenderId;
			const shouldUpdateRunId = canRender && runId && (!existing.runId || existing.runId === "pending");
			return {
				canRender,
				action: canRender ? ClaimAction.Existing : ClaimAction.Block,
				updateRunId: shouldUpdateRunId ? runId : void 0
			};
		}
		const normalizedRunId = runId !== null && runId !== void 0 ? runId : "pending";
		const renderClaimedByOtherMessageEntry = Object.entries(claims).find(([, claim]) => {
			var _claim$runId;
			return claim.stateRenderId === stateRenderId && ((_claim$runId = claim.runId) !== null && _claim$runId !== void 0 ? _claim$runId : "pending") === normalizedRunId && (0, _copilotkit_shared.dataToUUID)(getStateWithoutConstantKeys(claim.stateSnapshot)) === (0, _copilotkit_shared.dataToUUID)(getStateWithoutConstantKeys(stateSnapshot));
		});
		const renderClaimedByOtherMessage = renderClaimedByOtherMessageEntry === null || renderClaimedByOtherMessageEntry === void 0 ? void 0 : renderClaimedByOtherMessageEntry[1];
		const claimedMessageId = renderClaimedByOtherMessageEntry === null || renderClaimedByOtherMessageEntry === void 0 ? void 0 : renderClaimedByOtherMessageEntry[0];
		if (renderClaimedByOtherMessage) {
			if (messageIndex !== void 0 && renderClaimedByOtherMessage.messageIndex !== void 0 && messageIndex > renderClaimedByOtherMessage.messageIndex) return {
				canRender: true,
				action: ClaimAction.Override,
				nextClaim: {
					stateRenderId,
					runId,
					messageIndex
				},
				lockOthers: runId === renderClaimedByOtherMessage.runId || isPlaceholderMessageId(claimedMessageId)
			};
			if (runId && renderClaimedByOtherMessage.runId && runId !== renderClaimedByOtherMessage.runId) return {
				canRender: true,
				action: ClaimAction.Override,
				nextClaim: {
					stateRenderId,
					runId,
					messageIndex
				},
				lockOthers: isPlaceholderMessageId(claimedMessageId)
			};
			if (isPlaceholderMessageId(claimedMessageId)) return {
				canRender: true,
				action: ClaimAction.Override,
				nextClaim: {
					stateRenderId,
					runId,
					messageIndex
				},
				lockOthers: true
			};
			if (stateSnapshot && renderClaimedByOtherMessage.stateSnapshot && !areStatesEquals(renderClaimedByOtherMessage.stateSnapshot, stateSnapshot)) return {
				canRender: true,
				action: ClaimAction.Override,
				nextClaim: {
					stateRenderId,
					runId
				}
			};
			return {
				canRender: false,
				action: ClaimAction.Block
			};
		}
		if (!runId) return {
			canRender: false,
			action: ClaimAction.Block
		};
		return {
			canRender: true,
			action: ClaimAction.Create,
			nextClaim: {
				stateRenderId,
				runId,
				messageIndex
			}
		};
	}
	/**
	* Select the best snapshot to render for this message.
	* Priority order is:
	* 1) explicit message snapshot
	* 2) live agent state (latest assistant only)
	* 3) cached snapshot for message
	* 4) cached snapshot for stateRenderId+runId
	* 5) last cached snapshot for stateRenderId
	*/
	function selectSnapshot({ messageId, messageName, allowLiveState, skipLatestCache, stateRenderId, effectiveRunId, stateSnapshotProp, agentState, agentMessages, existingClaim, caches }) {
		var _reverse$find, _agentMessages$messag, _existingClaim$stateS, _existingClaim$stateS2;
		const lastAssistantId = agentMessages ? (_reverse$find = [...agentMessages].reverse().find((msg) => msg.role === "assistant")) === null || _reverse$find === void 0 ? void 0 : _reverse$find.id : void 0;
		const latestSnapshot = stateRenderId !== void 0 ? caches.byStateRenderAndRun[`${stateRenderId}::latest`] : void 0;
		const messageIndex = agentMessages ? agentMessages.findIndex((msg) => msg.id === messageId) : -1;
		const messageRole = messageIndex >= 0 && agentMessages ? (_agentMessages$messag = agentMessages[messageIndex]) === null || _agentMessages$messag === void 0 ? void 0 : _agentMessages$messag.role : void 0;
		let previousUserMessageId;
		if (messageIndex > 0 && agentMessages) for (let i = messageIndex - 1; i >= 0; i -= 1) {
			var _agentMessages$i;
			if (((_agentMessages$i = agentMessages[i]) === null || _agentMessages$i === void 0 ? void 0 : _agentMessages$i.role) === "user") {
				var _agentMessages$i2;
				previousUserMessageId = (_agentMessages$i2 = agentMessages[i]) === null || _agentMessages$i2 === void 0 ? void 0 : _agentMessages$i2.id;
				break;
			}
		}
		const liveStateIsStale = stateSnapshotProp === void 0 && latestSnapshot !== void 0 && agentState !== void 0 && areStatesEquals(latestSnapshot, agentState);
		const shouldUseLiveState = (Boolean(allowLiveState) || !lastAssistantId || messageId === lastAssistantId) && !liveStateIsStale;
		const snapshot = stateSnapshotProp ? (0, _copilotkit_shared.parseJson)(stateSnapshotProp, stateSnapshotProp) : shouldUseLiveState ? agentState : void 0;
		const hasSnapshotKeys = !!(snapshot && Object.keys(snapshot).length > 0);
		const allowEmptySnapshot = snapshot !== void 0 && !hasSnapshotKeys && (stateSnapshotProp !== void 0 || shouldUseLiveState);
		const messageCacheEntry = caches.byMessageId[messageId];
		const cachedMessageSnapshot = readCachedMessageEntry(messageCacheEntry).snapshot;
		const cacheKey = stateRenderId !== void 0 ? `${stateRenderId}::${effectiveRunId}` : void 0;
		let cachedSnapshot = cachedMessageSnapshot !== null && cachedMessageSnapshot !== void 0 ? cachedMessageSnapshot : caches.byMessageId[messageId];
		if (cachedSnapshot === void 0 && cacheKey && caches.byStateRenderAndRun[cacheKey] !== void 0) cachedSnapshot = caches.byStateRenderAndRun[cacheKey];
		if (cachedSnapshot === void 0 && stateRenderId && previousUserMessageId && caches.byStateRenderAndRun[`${stateRenderId}::pending:${previousUserMessageId}`] !== void 0) cachedSnapshot = caches.byStateRenderAndRun[`${stateRenderId}::pending:${previousUserMessageId}`];
		if (cachedSnapshot === void 0 && !skipLatestCache && stateRenderId && messageRole !== "assistant" && (stateSnapshotProp !== void 0 || agentState && Object.keys(agentState).length > 0)) cachedSnapshot = caches.byStateRenderAndRun[`${stateRenderId}::latest`];
		const snapshotForClaim = (existingClaim === null || existingClaim === void 0 ? void 0 : existingClaim.locked) ? (_existingClaim$stateS = existingClaim.stateSnapshot) !== null && _existingClaim$stateS !== void 0 ? _existingClaim$stateS : cachedSnapshot : hasSnapshotKeys ? snapshot : (_existingClaim$stateS2 = existingClaim === null || existingClaim === void 0 ? void 0 : existingClaim.stateSnapshot) !== null && _existingClaim$stateS2 !== void 0 ? _existingClaim$stateS2 : cachedSnapshot;
		return {
			snapshot,
			hasSnapshotKeys,
			cachedSnapshot,
			allowEmptySnapshot,
			snapshotForClaim
		};
	}

//#endregion
//#region src/hooks/use-coagent-state-render-registry.ts
	const LAST_SNAPSHOTS_BY_RENDER_AND_RUN = "__lastSnapshotsByStateRenderIdAndRun";
	const LAST_SNAPSHOTS_BY_MESSAGE = "__lastSnapshotsByMessageId";
	function getClaimsStore(claimsRef) {
		return claimsRef.current;
	}
	function getSnapshotCaches(claimsRef) {
		var _store$LAST_SNAPSHOTS, _store$LAST_SNAPSHOTS2;
		const store = getClaimsStore(claimsRef);
		return {
			byStateRenderAndRun: (_store$LAST_SNAPSHOTS = store[LAST_SNAPSHOTS_BY_RENDER_AND_RUN]) !== null && _store$LAST_SNAPSHOTS !== void 0 ? _store$LAST_SNAPSHOTS : {},
			byMessageId: (_store$LAST_SNAPSHOTS2 = store[LAST_SNAPSHOTS_BY_MESSAGE]) !== null && _store$LAST_SNAPSHOTS2 !== void 0 ? _store$LAST_SNAPSHOTS2 : {}
		};
	}
	function useStateRenderRegistry({ agentId, stateRenderId, message, messageIndex, stateSnapshot, agentState, agentMessages, claimsRef }) {
		var _store$LAST_SNAPSHOTS3, _claimsRef$current$me;
		const store = getClaimsStore(claimsRef);
		const runId = message.runId;
		const { runId: cachedMessageRunId } = readCachedMessageEntry((_store$LAST_SNAPSHOTS3 = store[LAST_SNAPSHOTS_BY_MESSAGE]) === null || _store$LAST_SNAPSHOTS3 === void 0 ? void 0 : _store$LAST_SNAPSHOTS3[message.id]);
		const effectiveRunId = getEffectiveRunId({
			existingClaimRunId: (_claimsRef$current$me = claimsRef.current[message.id]) === null || _claimsRef$current$me === void 0 ? void 0 : _claimsRef$current$me.runId,
			cachedMessageRunId,
			runId
		});
		(0, react.useEffect)(() => {
			return () => {
				const existingClaim = claimsRef.current[message.id];
				if ((existingClaim === null || existingClaim === void 0 ? void 0 : existingClaim.stateSnapshot) && Object.keys(existingClaim.stateSnapshot).length > 0) {
					var _store$LAST_SNAPSHOTS4, _existingClaim$runId, _store$LAST_SNAPSHOTS5, _existingClaim$runId2;
					const snapshotCache = { ...(_store$LAST_SNAPSHOTS4 = store[LAST_SNAPSHOTS_BY_RENDER_AND_RUN]) !== null && _store$LAST_SNAPSHOTS4 !== void 0 ? _store$LAST_SNAPSHOTS4 : {} };
					const cacheKey = `${existingClaim.stateRenderId}::${(_existingClaim$runId = existingClaim.runId) !== null && _existingClaim$runId !== void 0 ? _existingClaim$runId : "pending"}`;
					snapshotCache[cacheKey] = existingClaim.stateSnapshot;
					snapshotCache[`${existingClaim.stateRenderId}::latest`] = existingClaim.stateSnapshot;
					store[LAST_SNAPSHOTS_BY_RENDER_AND_RUN] = snapshotCache;
					const messageCache = { ...(_store$LAST_SNAPSHOTS5 = store[LAST_SNAPSHOTS_BY_MESSAGE]) !== null && _store$LAST_SNAPSHOTS5 !== void 0 ? _store$LAST_SNAPSHOTS5 : {} };
					messageCache[message.id] = {
						snapshot: existingClaim.stateSnapshot,
						runId: (_existingClaim$runId2 = existingClaim.runId) !== null && _existingClaim$runId2 !== void 0 ? _existingClaim$runId2 : effectiveRunId
					};
					store[LAST_SNAPSHOTS_BY_MESSAGE] = messageCache;
				}
				delete claimsRef.current[message.id];
			};
		}, [
			claimsRef,
			effectiveRunId,
			message.id
		]);
		if (!stateRenderId) return { canRender: false };
		const caches = getSnapshotCaches(claimsRef);
		const existingClaim = claimsRef.current[message.id];
		const { snapshot, hasSnapshotKeys, allowEmptySnapshot, snapshotForClaim } = selectSnapshot({
			messageId: message.id,
			messageName: message.name,
			allowLiveState: isPlaceholderMessageName(message.name) || isPlaceholderMessageId(message.id),
			skipLatestCache: isPlaceholderMessageName(message.name) || isPlaceholderMessageId(message.id),
			stateRenderId,
			effectiveRunId,
			stateSnapshotProp: stateSnapshot,
			agentState,
			agentMessages,
			existingClaim,
			caches
		});
		const resolution = resolveClaim({
			claims: claimsRef.current,
			context: {
				agentId,
				messageId: message.id,
				stateRenderId,
				runId: effectiveRunId,
				messageIndex
			},
			stateSnapshot: snapshotForClaim
		});
		if (resolution.action === ClaimAction.Block) return { canRender: false };
		if (resolution.updateRunId && claimsRef.current[message.id]) claimsRef.current[message.id].runId = resolution.updateRunId;
		if (resolution.nextClaim) claimsRef.current[message.id] = resolution.nextClaim;
		if (resolution.lockOthers) Object.entries(claimsRef.current).forEach(([id, claim]) => {
			if (id !== message.id && claim.stateRenderId === stateRenderId) claim.locked = true;
		});
		if (existingClaim && !existingClaim.locked && (agentMessages === null || agentMessages === void 0 ? void 0 : agentMessages.length)) {
			const indexInAgentMessages = agentMessages.findIndex((msg) => msg.id === message.id);
			if (indexInAgentMessages >= 0 && indexInAgentMessages < agentMessages.length - 1) existingClaim.locked = true;
		}
		const existingSnapshot = claimsRef.current[message.id].stateSnapshot;
		const snapshotChanged = stateSnapshot && existingSnapshot !== void 0 && !areStatesEquals(existingSnapshot, snapshot);
		if (snapshot && (stateSnapshot || hasSnapshotKeys || allowEmptySnapshot) && (!claimsRef.current[message.id].locked || snapshotChanged)) {
			if (!claimsRef.current[message.id].locked || snapshotChanged) {
				var _store$LAST_SNAPSHOTS6, _store$LAST_SNAPSHOTS7;
				claimsRef.current[message.id].stateSnapshot = snapshot;
				const snapshotCache = { ...(_store$LAST_SNAPSHOTS6 = store[LAST_SNAPSHOTS_BY_RENDER_AND_RUN]) !== null && _store$LAST_SNAPSHOTS6 !== void 0 ? _store$LAST_SNAPSHOTS6 : {} };
				const cacheKey = `${stateRenderId}::${effectiveRunId}`;
				snapshotCache[cacheKey] = snapshot;
				snapshotCache[`${stateRenderId}::latest`] = snapshot;
				store[LAST_SNAPSHOTS_BY_RENDER_AND_RUN] = snapshotCache;
				const messageCache = { ...(_store$LAST_SNAPSHOTS7 = store[LAST_SNAPSHOTS_BY_MESSAGE]) !== null && _store$LAST_SNAPSHOTS7 !== void 0 ? _store$LAST_SNAPSHOTS7 : {} };
				messageCache[message.id] = {
					snapshot,
					runId: effectiveRunId
				};
				store[LAST_SNAPSHOTS_BY_MESSAGE] = messageCache;
				if (stateSnapshot) claimsRef.current[message.id].locked = true;
			}
		} else if (snapshotForClaim) {
			if (!claimsRef.current[message.id].stateSnapshot) {
				var _store$LAST_SNAPSHOTS8, _store$LAST_SNAPSHOTS9;
				claimsRef.current[message.id].stateSnapshot = snapshotForClaim;
				const snapshotCache = { ...(_store$LAST_SNAPSHOTS8 = store[LAST_SNAPSHOTS_BY_RENDER_AND_RUN]) !== null && _store$LAST_SNAPSHOTS8 !== void 0 ? _store$LAST_SNAPSHOTS8 : {} };
				const cacheKey = `${stateRenderId}::${effectiveRunId}`;
				snapshotCache[cacheKey] = snapshotForClaim;
				snapshotCache[`${stateRenderId}::latest`] = snapshotForClaim;
				store[LAST_SNAPSHOTS_BY_RENDER_AND_RUN] = snapshotCache;
				const messageCache = { ...(_store$LAST_SNAPSHOTS9 = store[LAST_SNAPSHOTS_BY_MESSAGE]) !== null && _store$LAST_SNAPSHOTS9 !== void 0 ? _store$LAST_SNAPSHOTS9 : {} };
				messageCache[message.id] = {
					snapshot: snapshotForClaim,
					runId: effectiveRunId
				};
				store[LAST_SNAPSHOTS_BY_MESSAGE] = messageCache;
			}
		}
		return { canRender: true };
	}

//#endregion
//#region src/hooks/use-coagent-state-render-bridge.tsx
	function useCoagentStateRenderBridge(agentId, props) {
		var _props$runId;
		const { stateSnapshot, message } = props;
		const { coAgentStateRenders, claimsRef } = useCoAgentStateRenders();
		const { agent } = (0, _copilotkitnext_react.useAgent)({ agentId });
		const [nodeName, setNodeName] = (0, react.useState)(void 0);
		const [, forceUpdate] = (0, react.useState)(0);
		(0, react.useEffect)(() => {
			if (!agent) return;
			const { unsubscribe } = agent.subscribe({
				onStateChanged: () => {
					forceUpdate((value) => value + 1);
				},
				onStepStartedEvent: ({ event }) => {
					if (event.stepName !== nodeName) setNodeName(event.stepName);
				},
				onStepFinishedEvent: ({ event }) => {
					if (event.stepName === nodeName) setNodeName(void 0);
				}
			});
			return () => {
				unsubscribe();
			};
		}, [agentId, nodeName]);
		const getStateRender = (0, react.useCallback)((messageId) => {
			return Object.entries(coAgentStateRenders).find(([stateRenderId, stateRender]) => {
				if (claimsRef.current[messageId]) return stateRenderId === claimsRef.current[messageId].stateRenderId;
				const matchingAgentName = stateRender.name === agentId;
				const matchesNodeContext = stateRender.nodeName ? stateRender.nodeName === nodeName : true;
				return matchingAgentName && matchesNodeContext;
			});
		}, [
			coAgentStateRenders,
			nodeName,
			agentId
		]);
		const stateRenderEntry = (0, react.useMemo)(() => getStateRender(message.id), [getStateRender, message.id]);
		const stateRenderId = stateRenderEntry === null || stateRenderEntry === void 0 ? void 0 : stateRenderEntry[0];
		const stateRender = stateRenderEntry === null || stateRenderEntry === void 0 ? void 0 : stateRenderEntry[1];
		const { canRender } = useStateRenderRegistry({
			agentId,
			stateRenderId,
			message: {
				...message,
				runId: (_props$runId = props.runId) !== null && _props$runId !== void 0 ? _props$runId : message.runId
			},
			messageIndex: props.messageIndex,
			stateSnapshot,
			agentState: agent === null || agent === void 0 ? void 0 : agent.state,
			agentMessages: agent === null || agent === void 0 ? void 0 : agent.messages,
			claimsRef
		});
		return (0, react.useMemo)(() => {
			if (!stateRender || !stateRenderId) return null;
			if (!canRender) return null;
			if (stateRender.handler) {
				var _agent$state;
				stateRender.handler({
					state: stateSnapshot ? (0, _copilotkit_shared.parseJson)(stateSnapshot, stateSnapshot) : (_agent$state = agent === null || agent === void 0 ? void 0 : agent.state) !== null && _agent$state !== void 0 ? _agent$state : {},
					nodeName: nodeName !== null && nodeName !== void 0 ? nodeName : ""
				});
			}
			if (stateRender.render) {
				var _claimsRef$current$me;
				const status = (agent === null || agent === void 0 ? void 0 : agent.isRunning) ? RenderStatus.InProgress : RenderStatus.Complete;
				if (typeof stateRender.render === "string") return stateRender.render;
				return stateRender.render({
					status,
					state: (_claimsRef$current$me = claimsRef.current[message.id].stateSnapshot) !== null && _claimsRef$current$me !== void 0 ? _claimsRef$current$me : {},
					nodeName: nodeName !== null && nodeName !== void 0 ? nodeName : ""
				});
			}
		}, [
			stateRender,
			stateRenderId,
			agent === null || agent === void 0 ? void 0 : agent.state,
			agent === null || agent === void 0 ? void 0 : agent.isRunning,
			nodeName,
			message.id,
			stateSnapshot,
			canRender
		]);
	}
	function CoAgentStateRenderBridge(props) {
		return useCoagentStateRenderBridge(props.agentId, props);
	}

//#endregion
//#region src/components/CopilotListeners.tsx
	const usePredictStateSubscription = (agent) => {
		const predictStateToolsRef = (0, react.useRef)([]);
		const getSubscriber = (0, react.useCallback)((agent) => ({
			onCustomEvent: ({ event }) => {
				if (event.name === "PredictState") predictStateToolsRef.current = event.value;
			},
			onToolCallArgsEvent: ({ partialToolCallArgs, toolCallName }) => {
				predictStateToolsRef.current.forEach((t) => {
					if ((t === null || t === void 0 ? void 0 : t.tool) !== toolCallName) return;
					const emittedState = typeof partialToolCallArgs === "string" ? (0, _copilotkit_shared.parseJson)(partialToolCallArgs, partialToolCallArgs) : partialToolCallArgs;
					agent.setState({ [t.state_key]: emittedState[t.state_key] });
				});
			}
		}), []);
		(0, react.useEffect)(() => {
			if (!agent) return;
			const subscriber = getSubscriber(agent);
			const { unsubscribe } = agent.subscribe(subscriber);
			return () => {
				unsubscribe();
			};
		}, [agent, getSubscriber]);
	};
	function CopilotListeners() {
		const { copilotkit } = (0, _copilotkitnext_react.useCopilotKit)();
		const existingConfig = (0, _copilotkitnext_react.useCopilotChatConfiguration)();
		const resolvedAgentId = existingConfig === null || existingConfig === void 0 ? void 0 : existingConfig.agentId;
		const { setBannerError } = useToast();
		const { agent } = (0, _copilotkitnext_react.useAgent)({ agentId: resolvedAgentId });
		usePredictStateSubscription(agent);
		(0, react.useEffect)(() => {
			const subscription = copilotkit.subscribe({ onError: ({ error }) => {
				setBannerError(new _copilotkit_shared.CopilotKitLowLevelError({
					error,
					message: error.message,
					url: typeof window !== "undefined" ? window.location.href : ""
				}));
			} });
			return () => {
				subscription.unsubscribe();
			};
		}, [copilotkit === null || copilotkit === void 0 ? void 0 : copilotkit.subscribe]);
		return null;
	}

//#endregion
//#region src/components/copilot-provider/copilotkit.tsx
/**
	* This component will typically wrap your entire application (or a sub-tree of your application where you want to have a copilot). It provides the copilot context to all other components and hooks.
	*
	* ## Example
	*
	* You can find more information about self-hosting CopilotKit [here](/guides/self-hosting).
	*
	* ```tsx
	* import { CopilotKit } from "@copilotkit/react-core";
	*
	* <CopilotKit runtimeUrl="<your-runtime-url>">
	*   // ... your app ...
	* </CopilotKit>
	* ```
	*/
	function CopilotKit({ children, ...props }) {
		var _props$useSingleEndpo;
		const enabled = shouldShowDevConsole(props.showDevConsole);
		const showInspector = shouldShowDevConsole(props.enableInspector);
		const publicApiKey = props.publicApiKey || props.publicLicenseKey;
		const renderArr = (0, react.useMemo)(() => [{ render: CoAgentStateRenderBridge }], []);
		return /* @__PURE__ */ (0, react_jsx_runtime.jsx)(ToastProvider, {
			enabled,
			children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(CopilotErrorBoundary, {
				publicApiKey,
				showUsageBanner: enabled,
				children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(ThreadsProvider, {
					threadId: props.threadId,
					children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(_copilotkitnext_react.CopilotKitProvider, {
						...props,
						showDevConsole: showInspector,
						renderCustomMessages: renderArr,
						useSingleEndpoint: (_props$useSingleEndpo = props.useSingleEndpoint) !== null && _props$useSingleEndpo !== void 0 ? _props$useSingleEndpo : true,
						children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(CopilotKitInternal, {
							...props,
							children
						})
					})
				})
			})
		});
	}
	/**
	* Bridge component that subscribes to v2.x copilotkit core error events
	* and forwards them to v1.x error handling system.
	* This ensures only ONE subscription exists regardless of how many times
	* Chat components are rendered.
	*/
	function CopilotKitErrorBridge() {
		const { copilotkit } = (0, _copilotkitnext_react.useCopilotKit)();
		const { onError, copilotApiConfig } = useCopilotContext();
		(0, react.useEffect)(() => {
			if (!copilotkit) return;
			const subscription = copilotkit.subscribe({ onError: async (event) => {
				const errorEvent = {
					type: "error",
					timestamp: Date.now(),
					context: {
						source: "agent",
						request: {
							operation: event.code || "unknown",
							url: copilotApiConfig === null || copilotApiConfig === void 0 ? void 0 : copilotApiConfig.chatApiEndpoint,
							startTime: Date.now()
						},
						technical: {
							environment: "browser",
							userAgent: typeof navigator !== "undefined" ? navigator.userAgent : void 0,
							stackTrace: event.error.stack
						},
						...event.context
					},
					error: event.error
				};
				try {
					await onError(errorEvent);
				} catch (handlerError) {
					console.error("Error in onError handler:", handlerError);
				}
			} });
			return () => {
				subscription.unsubscribe();
			};
		}, [
			copilotkit,
			onError,
			copilotApiConfig
		]);
		return null;
	}
	function CopilotKitInternal(cpkProps) {
		var _props$agent2;
		const { children, ...props } = cpkProps;
		/**
		* This will throw an error if the props are invalid.
		*/
		validateProps(cpkProps);
		const publicApiKey = props.publicLicenseKey || props.publicApiKey;
		const chatApiEndpoint = props.runtimeUrl || _copilotkit_shared.COPILOT_CLOUD_CHAT_URL;
		const [actions, setActions] = (0, react.useState)({});
		const [registeredActionConfigs, setRegisteredActionConfigs] = (0, react.useState)(/* @__PURE__ */ new Map());
		const chatComponentsCache = (0, react.useRef)({
			actions: {},
			coAgentStateRenders: {}
		});
		const { addElement, removeElement, printTree, getAllElements } = useTree();
		const [isLoading, setIsLoading] = (0, react.useState)(false);
		const [chatInstructions, setChatInstructions] = (0, react.useState)("");
		const [authStates, setAuthStates] = (0, react.useState)({});
		const [extensions, setExtensions] = (0, react.useState)({});
		const [additionalInstructions, setAdditionalInstructions] = (0, react.useState)([]);
		const { addElement: addDocument, removeElement: removeDocument, allElements: allDocuments } = useFlatCategoryStore();
		const setAction = (0, react.useCallback)((id, action) => {
			setActions((prevPoints) => {
				return {
					...prevPoints,
					[id]: action
				};
			});
		}, []);
		const removeAction = (0, react.useCallback)((id) => {
			setActions((prevPoints) => {
				const newPoints = { ...prevPoints };
				delete newPoints[id];
				return newPoints;
			});
		}, []);
		const getContextString = (0, react.useCallback)((documents, categories) => {
			return `${documents.map((document) => {
				return `${document.name} (${document.sourceApplication}):\n${document.getContents()}`;
			}).join("\n\n")}\n\n${printTree(categories)}`;
		}, [printTree]);
		const addContext = (0, react.useCallback)((context, parentId, categories = defaultCopilotContextCategories) => {
			return addElement(context, categories, parentId);
		}, [addElement]);
		const removeContext = (0, react.useCallback)((id) => {
			removeElement(id);
		}, [removeElement]);
		const getAllContext = (0, react.useCallback)(() => {
			return getAllElements();
		}, [getAllElements]);
		const getFunctionCallHandler = (0, react.useCallback)((customEntryPoints) => {
			return entryPointsToFunctionCallHandler(Object.values(customEntryPoints || actions));
		}, [actions]);
		const getDocumentsContext = (0, react.useCallback)((categories) => {
			return allDocuments(categories);
		}, [allDocuments]);
		const addDocumentContext = (0, react.useCallback)((documentPointer, categories = defaultCopilotContextCategories) => {
			return addDocument(documentPointer, categories);
		}, [addDocument]);
		const removeDocumentContext = (0, react.useCallback)((documentId) => {
			removeDocument(documentId);
		}, [removeDocument]);
		const copilotApiConfig = (0, react.useMemo)(() => {
			let cloud = void 0;
			if (publicApiKey) {
				var _props$guardrails_c, _props$guardrails_c2;
				cloud = { guardrails: { input: { restrictToTopic: {
					enabled: Boolean(props.guardrails_c),
					validTopics: ((_props$guardrails_c = props.guardrails_c) === null || _props$guardrails_c === void 0 ? void 0 : _props$guardrails_c.validTopics) || [],
					invalidTopics: ((_props$guardrails_c2 = props.guardrails_c) === null || _props$guardrails_c2 === void 0 ? void 0 : _props$guardrails_c2.invalidTopics) || []
				} } } };
			}
			return {
				publicApiKey,
				...cloud ? { cloud } : {},
				chatApiEndpoint,
				headers: props.headers || {},
				properties: props.properties || {},
				transcribeAudioUrl: props.transcribeAudioUrl,
				textToSpeechUrl: props.textToSpeechUrl,
				credentials: props.credentials
			};
		}, [
			publicApiKey,
			props.headers,
			props.properties,
			props.transcribeAudioUrl,
			props.textToSpeechUrl,
			props.credentials,
			props.cloudRestrictToTopic,
			props.guardrails_c
		]);
		(0, react.useMemo)(() => {
			const authHeaders = Object.values(authStates || {}).reduce((acc, state) => {
				if (state.status === "authenticated" && state.authHeaders) return {
					...acc,
					...Object.entries(state.authHeaders).reduce((headers, [key, value]) => ({
						...headers,
						[key.startsWith("X-Custom-") ? key : `X-Custom-${key}`]: value
					}), {})
				};
				return acc;
			}, {});
			return {
				...copilotApiConfig.headers || {},
				...copilotApiConfig.publicApiKey ? { [_copilotkit_shared.COPILOT_CLOUD_PUBLIC_API_KEY_HEADER]: copilotApiConfig.publicApiKey } : {},
				...authHeaders
			};
		}, [
			copilotApiConfig.headers,
			copilotApiConfig.publicApiKey,
			authStates
		]);
		const [internalErrorHandlers, _setInternalErrorHandler] = (0, react.useState)({});
		const setInternalErrorHandler = (0, react.useCallback)((handler) => {
			_setInternalErrorHandler((prev) => ({
				...prev,
				...handler
			}));
		}, []);
		const removeInternalErrorHandler = (0, react.useCallback)((key) => {
			_setInternalErrorHandler((prev) => {
				const { [key]: _removed, ...rest } = prev;
				return rest;
			});
		}, []);
		const onErrorRef = (0, react.useRef)(props.onError);
		(0, react.useEffect)(() => {
			onErrorRef.current = props.onError;
		}, [props.onError]);
		const internalHandlersRef = (0, react.useRef)({});
		(0, react.useEffect)(() => {
			internalHandlersRef.current = internalErrorHandlers;
		}, [internalErrorHandlers]);
		const handleErrors = (0, react.useCallback)(async (error) => {
			if (copilotApiConfig.publicApiKey && onErrorRef.current) try {
				await onErrorRef.current(error);
			} catch (e) {
				console.error("Error in public onError handler:", e);
			}
			const handlers = Object.values(internalHandlersRef.current);
			await Promise.all(handlers.map((h) => Promise.resolve(h(error)).catch((e) => console.error("Error in internal error handler:", e))));
		}, [copilotApiConfig.publicApiKey]);
		const [chatSuggestionConfiguration, setChatSuggestionConfiguration] = (0, react.useState)({});
		const addChatSuggestionConfiguration = (0, react.useCallback)((id, suggestion) => {
			setChatSuggestionConfiguration((prev) => ({
				...prev,
				[id]: suggestion
			}));
		}, [setChatSuggestionConfiguration]);
		const removeChatSuggestionConfiguration = (0, react.useCallback)((id) => {
			setChatSuggestionConfiguration((prev) => {
				const { [id]: _, ...rest } = prev;
				return rest;
			});
		}, [setChatSuggestionConfiguration]);
		const [availableAgents, setAvailableAgents] = (0, react.useState)([]);
		const [coagentStates, setCoagentStates] = (0, react.useState)({});
		const coagentStatesRef = (0, react.useRef)({});
		const setCoagentStatesWithRef = (0, react.useCallback)((value) => {
			const newValue = typeof value === "function" ? value(coagentStatesRef.current) : value;
			coagentStatesRef.current = newValue;
			setCoagentStates((prev) => {
				return newValue;
			});
		}, []);
		let initialAgentSession = null;
		if (props.agent) initialAgentSession = { agentName: props.agent };
		const [agentSession, setAgentSession] = (0, react.useState)(initialAgentSession);
		(0, react.useEffect)(() => {
			if (props.agent) setAgentSession({ agentName: props.agent });
			else setAgentSession(null);
		}, [props.agent]);
		const { threadId, setThreadId: setInternalThreadId } = useThreads();
		const setThreadId = (0, react.useCallback)((value) => {
			if (props.threadId) throw new Error("Cannot call setThreadId() when threadId is provided via props.");
			setInternalThreadId(value);
		}, [props.threadId]);
		const [runId, setRunId] = (0, react.useState)(null);
		const chatAbortControllerRef = (0, react.useRef)(null);
		const showDevConsole = shouldShowDevConsole(props.showDevConsole);
		const [interruptActions, _setInterruptActions] = (0, react.useState)({});
		const setInterruptAction = (0, react.useCallback)((action) => {
			_setInterruptActions((prev) => {
				var _prev$action$id;
				if (action == null || !action.id) return prev;
				return {
					...prev,
					[action.id]: {
						...(_prev$action$id = prev[action.id]) !== null && _prev$action$id !== void 0 ? _prev$action$id : {},
						...action
					}
				};
			});
		}, []);
		const removeInterruptAction = (0, react.useCallback)((actionId) => {
			_setInterruptActions((prev) => {
				const { [actionId]: _, ...rest } = prev;
				return rest;
			});
		}, []);
		const [interruptEventQueue, setInterruptEventQueue] = (0, react.useState)({});
		const addInterruptEvent = (0, react.useCallback)((queuedEvent) => {
			setInterruptEventQueue((prev) => {
				const threadQueue = prev[queuedEvent.threadId] || [];
				return {
					...prev,
					[queuedEvent.threadId]: [...threadQueue, queuedEvent]
				};
			});
		}, []);
		const resolveInterruptEvent = (0, react.useCallback)((threadId, eventId, response) => {
			setInterruptEventQueue((prev) => {
				const threadQueue = prev[threadId] || [];
				return {
					...prev,
					[threadId]: threadQueue.map((queuedEvent) => queuedEvent.eventId === eventId ? {
						...queuedEvent,
						event: {
							...queuedEvent.event,
							response
						}
					} : queuedEvent)
				};
			});
		}, []);
		const memoizedChildren = (0, react.useMemo)(() => children, [children]);
		const [bannerError, setBannerError] = (0, react.useState)(null);
		const agentLock = (0, react.useMemo)(() => {
			var _props$agent;
			return (_props$agent = props.agent) !== null && _props$agent !== void 0 ? _props$agent : null;
		}, [props.agent]);
		const forwardedParameters = (0, react.useMemo)(() => {
			var _props$forwardedParam;
			return (_props$forwardedParam = props.forwardedParameters) !== null && _props$forwardedParam !== void 0 ? _props$forwardedParam : {};
		}, [props.forwardedParameters]);
		const updateExtensions = (0, react.useCallback)((newExtensions) => {
			setExtensions((prev) => {
				const resolved = typeof newExtensions === "function" ? newExtensions(prev) : newExtensions;
				return Object.keys(resolved).length === Object.keys(prev).length && Object.entries(resolved).every(([key, value]) => prev[key] === value) ? prev : resolved;
			});
		}, [setExtensions]);
		const updateAuthStates = (0, react.useCallback)((newAuthStates) => {
			setAuthStates((prev) => {
				const resolved = typeof newAuthStates === "function" ? newAuthStates(prev) : newAuthStates;
				return Object.keys(resolved).length === Object.keys(prev).length && Object.entries(resolved).every(([key, value]) => prev[key] === value) ? prev : resolved;
			});
		}, [setAuthStates]);
		const handleSetRegisteredActions = (0, react.useCallback)((actionConfig) => {
			const key = actionConfig.action.name || (0, _copilotkit_shared.randomUUID)();
			setRegisteredActionConfigs((prev) => {
				const newMap = new Map(prev);
				newMap.set(key, actionConfig);
				return newMap;
			});
			return key;
		}, []);
		const handleRemoveRegisteredAction = (0, react.useCallback)((actionKey) => {
			setRegisteredActionConfigs((prev) => {
				const newMap = new Map(prev);
				newMap.delete(actionKey);
				return newMap;
			});
		}, []);
		const RegisteredActionsRenderer = (0, react.useMemo)(() => {
			return () => /* @__PURE__ */ (0, react_jsx_runtime.jsx)(react_jsx_runtime.Fragment, { children: Array.from(registeredActionConfigs.entries()).map(([key, config]) => {
				const Component = config.component;
				return /* @__PURE__ */ (0, react_jsx_runtime.jsx)(Component, { action: config.action }, key);
			}) });
		}, [registeredActionConfigs]);
		const copilotContextValue = (0, react.useMemo)(() => ({
			actions,
			chatComponentsCache,
			getFunctionCallHandler,
			setAction,
			removeAction,
			setRegisteredActions: handleSetRegisteredActions,
			removeRegisteredAction: handleRemoveRegisteredAction,
			getContextString,
			addContext,
			removeContext,
			getAllContext,
			getDocumentsContext,
			addDocumentContext,
			removeDocumentContext,
			copilotApiConfig,
			isLoading,
			setIsLoading,
			chatSuggestionConfiguration,
			addChatSuggestionConfiguration,
			removeChatSuggestionConfiguration,
			chatInstructions,
			setChatInstructions,
			additionalInstructions,
			setAdditionalInstructions,
			showDevConsole,
			coagentStates,
			setCoagentStates,
			coagentStatesRef,
			setCoagentStatesWithRef,
			agentSession,
			setAgentSession,
			forwardedParameters,
			agentLock,
			threadId,
			setThreadId,
			runId,
			setRunId,
			chatAbortControllerRef,
			availableAgents,
			authConfig_c: props.authConfig_c,
			authStates_c: authStates,
			setAuthStates_c: updateAuthStates,
			extensions,
			setExtensions: updateExtensions,
			interruptActions,
			setInterruptAction,
			removeInterruptAction,
			interruptEventQueue,
			addInterruptEvent,
			resolveInterruptEvent,
			bannerError,
			setBannerError,
			onError: handleErrors,
			internalErrorHandlers,
			setInternalErrorHandler,
			removeInternalErrorHandler
		}), [
			actions,
			chatComponentsCache,
			getFunctionCallHandler,
			setAction,
			removeAction,
			handleSetRegisteredActions,
			handleRemoveRegisteredAction,
			getContextString,
			addContext,
			removeContext,
			getAllContext,
			getDocumentsContext,
			addDocumentContext,
			removeDocumentContext,
			copilotApiConfig,
			isLoading,
			chatSuggestionConfiguration,
			addChatSuggestionConfiguration,
			removeChatSuggestionConfiguration,
			chatInstructions,
			additionalInstructions,
			showDevConsole,
			coagentStates,
			setCoagentStatesWithRef,
			agentSession,
			setAgentSession,
			forwardedParameters,
			agentLock,
			threadId,
			setThreadId,
			runId,
			availableAgents,
			props.authConfig_c,
			authStates,
			updateAuthStates,
			extensions,
			updateExtensions,
			interruptActions,
			setInterruptAction,
			removeInterruptAction,
			interruptEventQueue,
			addInterruptEvent,
			resolveInterruptEvent,
			bannerError,
			handleErrors,
			internalErrorHandlers,
			setInternalErrorHandler,
			removeInternalErrorHandler
		]);
		return /* @__PURE__ */ (0, react_jsx_runtime.jsx)(_copilotkitnext_react.CopilotChatConfigurationProvider, {
			agentId: (_props$agent2 = props.agent) !== null && _props$agent2 !== void 0 ? _props$agent2 : "default",
			threadId,
			children: /* @__PURE__ */ (0, react_jsx_runtime.jsxs)(CopilotContext.Provider, {
				value: copilotContextValue,
				children: [
					/* @__PURE__ */ (0, react_jsx_runtime.jsx)(CopilotListeners, {}),
					/* @__PURE__ */ (0, react_jsx_runtime.jsx)(CopilotKitErrorBridge, {}),
					/* @__PURE__ */ (0, react_jsx_runtime.jsxs)(CoAgentStateRendersProvider, { children: [/* @__PURE__ */ (0, react_jsx_runtime.jsx)(MessagesTapProvider, { children: /* @__PURE__ */ (0, react_jsx_runtime.jsxs)(CopilotMessages, { children: [memoizedChildren, /* @__PURE__ */ (0, react_jsx_runtime.jsx)(RegisteredActionsRenderer, {})] }) }), bannerError && showDevConsole && /* @__PURE__ */ (0, react_jsx_runtime.jsx)(UsageBanner, {
						severity: bannerError.severity,
						message: bannerError.message,
						onClose: () => setBannerError(null),
						actions: getErrorActions(bannerError)
					})] })
				]
			})
		});
	}
	const defaultCopilotContextCategories = ["global"];
	function entryPointsToFunctionCallHandler(actions) {
		return async ({ name, args }) => {
			let actionsByFunctionName = {};
			for (let action of actions) actionsByFunctionName[action.name] = action;
			const action = actionsByFunctionName[name];
			let result = void 0;
			if (action) {
				await new Promise((resolve, reject) => {
					(0, react_dom.flushSync)(async () => {
						try {
							var _action$handler;
							result = await ((_action$handler = action.handler) === null || _action$handler === void 0 ? void 0 : _action$handler.call(action, args));
							resolve();
						} catch (error) {
							reject(error);
						}
					});
				});
				await new Promise((resolve) => setTimeout(resolve, 20));
			}
			return result;
		};
	}
	function formatFeatureName(featureName) {
		return featureName.replace(/_c$/, "").split("_").map((word) => word.charAt(0).toUpperCase() + word.slice(1).toLowerCase()).join(" ");
	}
	function validateProps(props) {
		const cloudFeatures = Object.keys(props).filter((key) => key.endsWith("_c"));
		const hasApiKey = props.publicApiKey || props.publicLicenseKey;
		if (!props.runtimeUrl && !hasApiKey) throw new _copilotkit_shared.ConfigurationError("Missing required prop: 'runtimeUrl' or 'publicApiKey' or 'publicLicenseKey'");
		if (cloudFeatures.length > 0 && !hasApiKey) throw new _copilotkit_shared.MissingPublicApiKeyError(`Missing required prop: 'publicApiKey' or 'publicLicenseKey' to use cloud features: ${cloudFeatures.map(formatFeatureName).join(", ")}`);
	}

//#endregion
//#region src/hooks/use-lazy-tool-renderer.tsx
	function useLazyToolRenderer() {
		const renderToolCall = (0, _copilotkitnext_react.useRenderToolCall)();
		return (0, react.useCallback)((message, messages) => {
			var _message$toolCalls;
			if (!(message === null || message === void 0 || (_message$toolCalls = message.toolCalls) === null || _message$toolCalls === void 0 ? void 0 : _message$toolCalls.length)) return null;
			const toolCall = message.toolCalls[0];
			if (!toolCall) return null;
			const toolMessage = messages === null || messages === void 0 ? void 0 : messages.find((m) => m.role === "tool" && m.toolCallId === toolCall.id);
			return () => renderToolCall({
				toolCall,
				toolMessage
			});
		}, [renderToolCall]);
	}

//#endregion
//#region src/hooks/use-copilot-chat_internal.ts
	function useCopilotChatInternal({ suggestions, onInProgress, onSubmitMessage, onStopGeneration, onReloadMessages } = {}) {
		var _existingConfig$agent, _existingConfig$threa, _agent$messages5;
		const { copilotkit } = (0, _copilotkitnext_react.useCopilotKit)();
		const { threadId, agentSession } = useCopilotContext();
		const existingConfig = (0, _copilotkitnext_react.useCopilotChatConfiguration)();
		const [agentAvailable, setAgentAvailable] = (0, react.useState)(false);
		const resolvedAgentId = (_existingConfig$agent = existingConfig === null || existingConfig === void 0 ? void 0 : existingConfig.agentId) !== null && _existingConfig$agent !== void 0 ? _existingConfig$agent : "default";
		const { agent } = (0, _copilotkitnext_react.useAgent)({ agentId: resolvedAgentId });
		(0, react.useEffect)(() => {
			const connect = async (agent) => {
				setAgentAvailable(false);
				try {
					await copilotkit.connectAgent({ agent });
					setAgentAvailable(true);
				} catch (error) {
					if (error instanceof _ag_ui_client.AGUIConnectNotImplementedError) {} else console.error("CopilotChat: connectAgent failed", error);
				}
			};
			if (agent && (existingConfig === null || existingConfig === void 0 ? void 0 : existingConfig.threadId) && agent.threadId !== existingConfig.threadId && copilotkit.runtimeConnectionStatus === _copilotkitnext_core.CopilotKitCoreRuntimeConnectionStatus.Connected) {
				agent.threadId = existingConfig.threadId;
				connect(agent);
			}
			return () => {};
		}, [
			existingConfig === null || existingConfig === void 0 ? void 0 : existingConfig.threadId,
			agent,
			copilotkit,
			copilotkit.runtimeConnectionStatus,
			resolvedAgentId
		]);
		(0, react.useEffect)(() => {
			onInProgress === null || onInProgress === void 0 || onInProgress(Boolean(agent === null || agent === void 0 ? void 0 : agent.isRunning));
		}, [agent === null || agent === void 0 ? void 0 : agent.isRunning, onInProgress]);
		const [interrupt, setInterrupt] = (0, react.useState)(null);
		(0, react.useEffect)(() => {
			setInterrupt(copilotkit.interruptElement);
			const subscription = copilotkit.subscribe({ onInterruptElementChanged: ({ interruptElement }) => {
				setInterrupt(interruptElement);
			} });
			return () => subscription.unsubscribe();
		}, [copilotkit]);
		const reset = () => {
			agent === null || agent === void 0 || agent.setMessages([]);
			agent === null || agent === void 0 || agent.setState(null);
		};
		const latestDelete = useUpdatedRef((0, react.useCallback)((messageId) => {
			var _agent$messages;
			const filteredMessages = ((_agent$messages = agent === null || agent === void 0 ? void 0 : agent.messages) !== null && _agent$messages !== void 0 ? _agent$messages : []).filter((message) => message.id !== messageId);
			agent === null || agent === void 0 || agent.setMessages(filteredMessages);
		}, [agent === null || agent === void 0 ? void 0 : agent.setMessages, agent === null || agent === void 0 ? void 0 : agent.messages]));
		const latestDeleteFunc = (0, react.useCallback)((messageId) => {
			return latestDelete.current(messageId);
		}, [latestDelete]);
		const currentSuggestions = (0, _copilotkitnext_react.useSuggestions)({ agentId: resolvedAgentId });
		const reload = useAsyncCallback(async (reloadMessageId) => {
			var _agent$messages2;
			if (!agent) return;
			const messages = (_agent$messages2 = agent === null || agent === void 0 ? void 0 : agent.messages) !== null && _agent$messages2 !== void 0 ? _agent$messages2 : [];
			if (agent.isRunning || messages.length === 0) return;
			const reloadMessageIndex = messages.findIndex((msg) => msg.id === reloadMessageId);
			if (reloadMessageIndex === -1) {
				console.warn(`Message with id ${reloadMessageId} not found`);
				return;
			}
			const reloadMessageRole = messages[reloadMessageIndex].role;
			if (reloadMessageRole !== "assistant") {
				console.warn(`Regenerate cannot be performed on ${reloadMessageRole} role`);
				return;
			}
			let historyCutoff = [messages[0]];
			if (messages.length > 2 && reloadMessageIndex !== 0) {
				const lastUserMessageBeforeRegenerate = messages.slice(0, reloadMessageIndex).reverse().find((msg) => msg.role === "user");
				if (!lastUserMessageBeforeRegenerate) historyCutoff = [messages[0]];
				else {
					const indexOfLastUserMessageBeforeRegenerate = messages.findIndex((msg) => msg.id === lastUserMessageBeforeRegenerate.id);
					historyCutoff = messages.slice(0, indexOfLastUserMessageBeforeRegenerate + 1);
				}
			} else if (messages.length > 2 && reloadMessageIndex === 0) historyCutoff = [messages[0], messages[1]];
			agent === null || agent === void 0 || agent.setMessages(historyCutoff);
			if (agent) try {
				await copilotkit.runAgent({ agent });
			} catch (error) {
				console.error("CopilotChat: runAgent failed during reload", error);
			}
		}, [
			agent === null || agent === void 0 ? void 0 : agent.messages.length,
			agent === null || agent === void 0 ? void 0 : agent.isRunning,
			agent === null || agent === void 0 ? void 0 : agent.setMessages,
			copilotkit === null || copilotkit === void 0 ? void 0 : copilotkit.runAgent
		]);
		const latestSendMessageFunc = useAsyncCallback(async (message, options) => {
			var _options$followUp;
			if (!agent) return;
			const followUp = (_options$followUp = options === null || options === void 0 ? void 0 : options.followUp) !== null && _options$followUp !== void 0 ? _options$followUp : true;
			if (options === null || options === void 0 ? void 0 : options.clearSuggestions) copilotkit.clearSuggestions(resolvedAgentId);
			if (onSubmitMessage) {
				const content = typeof message.content === "string" ? message.content : message.content && "text" in message.content ? message.content.text : message.content && "filename" in message.content ? message.content.filename : "";
				try {
					await onSubmitMessage(content);
				} catch (error) {
					console.error("Error in onSubmitMessage:", error);
				}
			}
			agent === null || agent === void 0 || agent.addMessage(message);
			if (followUp) try {
				await copilotkit.runAgent({ agent });
			} catch (error) {
				console.error("CopilotChat: runAgent failed", error);
			}
		}, [
			agent,
			copilotkit,
			resolvedAgentId,
			onSubmitMessage
		]);
		const latestAppendFunc = useAsyncCallback(async (message, options) => {
			return latestSendMessageFunc((0, _copilotkit_runtime_client_gql.gqlToAGUI)([message])[0], options);
		}, [latestSendMessageFunc]);
		const latestSetMessagesFunc = (0, react.useCallback)((messages) => {
			var _agent$setMessages2;
			if (messages.every((message) => message instanceof _copilotkit_runtime_client_gql.Message)) {
				var _agent$setMessages;
				return agent === null || agent === void 0 || (_agent$setMessages = agent.setMessages) === null || _agent$setMessages === void 0 ? void 0 : _agent$setMessages.call(agent, (0, _copilotkit_runtime_client_gql.gqlToAGUI)(messages));
			}
			return agent === null || agent === void 0 || (_agent$setMessages2 = agent.setMessages) === null || _agent$setMessages2 === void 0 ? void 0 : _agent$setMessages2.call(agent, messages);
		}, [agent === null || agent === void 0 ? void 0 : agent.setMessages, agent]);
		const latestReload = useUpdatedRef(reload);
		const latestReloadFunc = useAsyncCallback(async (messageId) => {
			var _agent$messages3;
			onReloadMessages === null || onReloadMessages === void 0 || onReloadMessages({
				messageId,
				currentAgentName: agent === null || agent === void 0 ? void 0 : agent.agentId,
				messages: (_agent$messages3 = agent === null || agent === void 0 ? void 0 : agent.messages) !== null && _agent$messages3 !== void 0 ? _agent$messages3 : []
			});
			return await latestReload.current(messageId);
		}, [
			latestReload,
			agent,
			onReloadMessages
		]);
		const latestStopFunc = (0, react.useCallback)(() => {
			var _agent$messages4, _agent$abortRun;
			onStopGeneration === null || onStopGeneration === void 0 || onStopGeneration({
				currentAgentName: agent === null || agent === void 0 ? void 0 : agent.agentId,
				messages: (_agent$messages4 = agent === null || agent === void 0 ? void 0 : agent.messages) !== null && _agent$messages4 !== void 0 ? _agent$messages4 : []
			});
			return agent === null || agent === void 0 || (_agent$abortRun = agent.abortRun) === null || _agent$abortRun === void 0 ? void 0 : _agent$abortRun.call(agent);
		}, [onStopGeneration, agent]);
		const latestReset = useUpdatedRef(reset);
		const latestResetFunc = (0, react.useCallback)(() => {
			return latestReset.current();
		}, [latestReset]);
		const lazyToolRendered = useLazyToolRenderer();
		const renderCustomMessage = (0, _copilotkitnext_react.useRenderCustomMessages)();
		const legacyCustomMessageRenderer = useLegacyCoagentRenderer({
			copilotkit,
			agent,
			agentId: resolvedAgentId,
			threadId: (_existingConfig$threa = existingConfig === null || existingConfig === void 0 ? void 0 : existingConfig.threadId) !== null && _existingConfig$threa !== void 0 ? _existingConfig$threa : threadId
		});
		const allMessages = (_agent$messages5 = agent === null || agent === void 0 ? void 0 : agent.messages) !== null && _agent$messages5 !== void 0 ? _agent$messages5 : [];
		const resolvedMessages = (0, react.useMemo)(() => {
			var _copilotkit$getAgent, _ref;
			let processedMessages = allMessages.map((message) => {
				if (message.role !== "assistant") return message;
				const lazyRendered = lazyToolRendered(message, allMessages);
				if (lazyRendered) {
					const renderedGenUi = lazyRendered();
					if (renderedGenUi) return {
						...message,
						generativeUI: () => renderedGenUi
					};
				}
				const bridgeRenderer = legacyCustomMessageRenderer || renderCustomMessage ? () => {
					if (legacyCustomMessageRenderer) return legacyCustomMessageRenderer({
						message,
						position: "before"
					});
					try {
						var _renderCustomMessage;
						return (_renderCustomMessage = renderCustomMessage === null || renderCustomMessage === void 0 ? void 0 : renderCustomMessage({
							message,
							position: "before"
						})) !== null && _renderCustomMessage !== void 0 ? _renderCustomMessage : null;
					} catch (error) {
						console.warn("[CopilotKit] renderCustomMessages failed, falling back to legacy renderer", error);
						return null;
					}
				} : null;
				if (bridgeRenderer) return {
					...message,
					generativeUI: bridgeRenderer,
					generativeUIPosition: "before"
				};
				return message;
			});
			const hasAssistantMessages = processedMessages.some((msg) => msg.role === "assistant");
			const canUseCustomRenderer = Boolean(renderCustomMessage && (copilotkit === null || copilotkit === void 0 || (_copilotkit$getAgent = copilotkit.getAgent) === null || _copilotkit$getAgent === void 0 ? void 0 : _copilotkit$getAgent.call(copilotkit, resolvedAgentId)));
			const placeholderRenderer = legacyCustomMessageRenderer ? legacyCustomMessageRenderer : canUseCustomRenderer ? renderCustomMessage : null;
			const shouldRenderPlaceholder = Boolean(agent === null || agent === void 0 ? void 0 : agent.isRunning) || Boolean((agent === null || agent === void 0 ? void 0 : agent.state) && Object.keys(agent.state).length);
			const effectiveThreadId = (_ref = threadId !== null && threadId !== void 0 ? threadId : agent === null || agent === void 0 ? void 0 : agent.threadId) !== null && _ref !== void 0 ? _ref : "default";
			let latestUserIndex = -1;
			for (let i = processedMessages.length - 1; i >= 0; i -= 1) if (processedMessages[i].role === "user") {
				latestUserIndex = i;
				break;
			}
			const latestUserMessageId = latestUserIndex >= 0 ? processedMessages[latestUserIndex].id : void 0;
			const currentRunId = latestUserMessageId ? copilotkit.getRunIdForMessage(resolvedAgentId, effectiveThreadId, latestUserMessageId) || `pending:${latestUserMessageId}` : void 0;
			const hasAssistantForCurrentRun = latestUserIndex >= 0 ? processedMessages.slice(latestUserIndex + 1).some((msg) => msg.role === "assistant") : hasAssistantMessages;
			if (placeholderRenderer && shouldRenderPlaceholder && !hasAssistantForCurrentRun) {
				const placeholderMessage = {
					id: currentRunId ? `coagent-state-render-${resolvedAgentId}-${currentRunId}` : `coagent-state-render-${resolvedAgentId}`,
					role: "assistant",
					content: "",
					name: "coagent-state-render",
					runId: currentRunId
				};
				processedMessages = [...processedMessages, {
					...placeholderMessage,
					generativeUIPosition: "before",
					generativeUI: () => placeholderRenderer({
						message: placeholderMessage,
						position: "before"
					})
				}];
			}
			return processedMessages;
		}, [
			agent === null || agent === void 0 ? void 0 : agent.messages,
			lazyToolRendered,
			allMessages,
			renderCustomMessage,
			legacyCustomMessageRenderer,
			resolvedAgentId,
			copilotkit,
			agent === null || agent === void 0 ? void 0 : agent.isRunning,
			agent === null || agent === void 0 ? void 0 : agent.state
		]);
		const renderedSuggestions = (0, react.useMemo)(() => {
			if (Array.isArray(suggestions)) return {
				suggestions: suggestions.map((s) => ({
					...s,
					isLoading: false
				})),
				isLoading: false
			};
			return currentSuggestions;
		}, [suggestions, currentSuggestions]);
		return {
			messages: resolvedMessages,
			sendMessage: latestSendMessageFunc,
			appendMessage: latestAppendFunc,
			setMessages: latestSetMessagesFunc,
			reloadMessages: latestReloadFunc,
			stopGeneration: latestStopFunc,
			reset: latestResetFunc,
			deleteMessage: latestDeleteFunc,
			isAvailable: agentAvailable,
			isLoading: Boolean(agent === null || agent === void 0 ? void 0 : agent.isRunning),
			suggestions: renderedSuggestions.suggestions,
			setSuggestions: (suggestions) => copilotkit.addSuggestionsConfig({ suggestions }),
			generateSuggestions: async () => copilotkit.reloadSuggestions(resolvedAgentId),
			resetSuggestions: () => copilotkit.clearSuggestions(resolvedAgentId),
			isLoadingSuggestions: renderedSuggestions.isLoading,
			interrupt,
			agent,
			threadId
		};
	}
	function useUpdatedRef(value) {
		const ref = (0, react.useRef)(value);
		(0, react.useEffect)(() => {
			ref.current = value;
		}, [value]);
		return ref;
	}
	function useLegacyCoagentRenderer({ copilotkit, agent, agentId, threadId }) {
		return (0, react.useMemo)(() => {
			if (!copilotkit || !agent) return null;
			return ({ message, position }) => {
				var _ref2;
				const effectiveThreadId = (_ref2 = threadId !== null && threadId !== void 0 ? threadId : agent.threadId) !== null && _ref2 !== void 0 ? _ref2 : "default";
				const providedRunId = message.runId;
				return (0, react.createElement)(CoAgentStateRenderBridge, {
					message,
					position,
					runId: (providedRunId ? providedRunId : copilotkit.getRunIdForMessage(agentId, effectiveThreadId, message.id)) || `pending:${message.id}`,
					messageIndex: Math.max(agent.messages.findIndex((msg) => msg.id === message.id), 0),
					messageIndexInRun: 0,
					numberOfMessagesInRun: 1,
					agentId,
					stateSnapshot: message.state
				});
			};
		}, [
			agent,
			agentId,
			copilotkit,
			threadId
		]);
	}

//#endregion
//#region src/hooks/use-copilot-chat.ts
/**
	* `useCopilotChat` is a lightweight React hook for headless chat interactions.
	* Perfect for controlling the prebuilt chat components programmatically.
	*
	* **Open Source Friendly** - Works without requiring a free public license key.
	*
	* <Callout title="Looking for fully headless UI?">
	* Get started with [useCopilotChatHeadless_c](https://docs.copilotkit.ai/reference/v1/hooks/useCopilotChatHeadless_c).
	* </Callout>
	*
	* ## Use Cases
	*
	* - **Programmatic Messaging**: Send messages without displaying chat UI
	* - **Programmatic control**: Control prebuilt component programmatically
	* - **Background Operations**: Trigger AI interactions in the background
	* - **Fire-and-Forget**: Send messages without needing to read responses
	*
	* ## Usage
	*
	* ```tsx
	* import { TextMessage, MessageRole } from "@copilotkit/runtime-client-gql";
	*
	* const { appendMessage } = useCopilotChat();
	*
	* // Example usage without naming conflicts
	* const handleSendMessage = async (content: string) => {
	*   await appendMessage(
	*     new TextMessage({
	*       role: MessageRole.User,
	*       content,
	*     })
	*   );
	* };
	* ```
	*
	* ## Return Values
	* The following properties are returned from the hook:
	*
	* <PropertyReference name="visibleMessages" type="DeprecatedGqlMessage[]" deprecated>
	* Array of messages in old non-AG-UI format, use for compatibility only
	* </PropertyReference>
	*
	* <PropertyReference name="appendMessage" type="(message: DeprecatedGqlMessage, options?) => Promise<void>" deprecated>
	* Append message using old format, use `sendMessage` instead
	* </PropertyReference>
	*
	* <PropertyReference name="reloadMessages" type="(messageId: string) => Promise<void>">
	* Regenerate the response for a specific message by ID
	* </PropertyReference>
	*
	* <PropertyReference name="stopGeneration" type="() => void">
	* Stop the current message generation process
	* </PropertyReference>
	*
	* <PropertyReference name="reset" type="() => void">
	* Clear all messages and reset chat state completely
	* </PropertyReference>
	*
	* <PropertyReference name="isLoading" type="boolean">
	* Whether the chat is currently generating a response
	* </PropertyReference>
	*
	* <PropertyReference name="runChatCompletion" type="() => Promise<Message[]>">
	* Manually trigger chat completion for advanced usage
	* </PropertyReference>
	*
	* <PropertyReference name="mcpServers" type="MCPServerConfig[]">
	* Array of Model Context Protocol server configurations
	* </PropertyReference>
	*
	* <PropertyReference name="setMcpServers" type="(servers: MCPServerConfig[]) => void">
	* Update MCP server configurations for enhanced context
	* </PropertyReference>
	*/
	/**
	* A lightweight React hook for headless chat interactions.
	* Perfect for programmatic messaging, background operations, and custom UI implementations.
	*
	* **Open Source Friendly** - Works without requiring a `publicApiKey`.
	*/
	function useCopilotChat(options = {}) {
		const { visibleMessages, appendMessage, reloadMessages, stopGeneration, reset, isLoading, isAvailable, runChatCompletion, mcpServers, setMcpServers } = useCopilotChatInternal(options);
		return {
			visibleMessages,
			appendMessage,
			reloadMessages,
			stopGeneration,
			reset,
			isLoading,
			isAvailable,
			runChatCompletion,
			mcpServers,
			setMcpServers
		};
	}

//#endregion
//#region src/hooks/use-copilot-chat-headless_c.ts
/**
	* `useCopilotChatHeadless_c` is for building fully custom UI (headless UI) implementations.
	*
	* <Callout title="This is a premium-only feature">
	* Sign up for free on [Copilot Cloud](https://cloud.copilotkit.ai) to get your public license key or read more about <a href="/premium/overview">premium features</a>.
	*
	* Usage is generous, **free** to get started, and works with **either self-hosted or Copilot Cloud** environments.
	* </Callout>
	*
	* ## Key Features
	*
	* - **Fully headless**: Build your own fully custom UI's for your agentic applications.
	* - **Advanced Suggestions**: Direct access to suggestions array with full control
	* - **Interrupt Handling**: Support for advanced interrupt functionality
	* - **MCP Server Support**: Model Context Protocol server configurations
	* - **Chat Controls**: Complete set of chat management functions
	* - **Loading States**: Comprehensive loading state management
	*
	*
	* ## Usage
	*
	* ### Basic Setup
	*
	* ```tsx
	* import { CopilotKit } from "@copilotkit/react-core";
	* import { useCopilotChatHeadless_c } from "@copilotkit/react-core";
	*
	* export function App() {
	*   return (
	*     <CopilotKit publicApiKey="your-free-public-license-key">
	*       <YourComponent />
	*     </CopilotKit>
	*   );
	* }
	*
	* export function YourComponent() {
	*   const { messages, sendMessage, isLoading } = useCopilotChatHeadless_c();
	*
	*   const handleSendMessage = async () => {
	*     await sendMessage({
	*       id: "123",
	*       role: "user",
	*       content: "Hello World",
	*     });
	*   };
	*
	*   return (
	*     <div>
	*       {messages.map(msg => <div key={msg.id}>{msg.content}</div>)}
	*       <button onClick={handleSendMessage} disabled={isLoading}>
	*         Send Message
	*       </button>
	*     </div>
	*   );
	* }
	* ```
	*
	* ### Working with Suggestions
	*
	* ```tsx
	* import { useCopilotChatHeadless_c, useCopilotChatSuggestions } from "@copilotkit/react-core";
	*
	* export function SuggestionExample() {
	*   const {
	*     suggestions,
	*     setSuggestions,
	*     generateSuggestions,
	*     isLoadingSuggestions
	*   } = useCopilotChatHeadless_c();
	*
	*   // Configure AI suggestion generation
	*   useCopilotChatSuggestions({
	*     instructions: "Suggest helpful actions based on the current context",
	*     maxSuggestions: 3
	*   });
	*
	*   return (
	*     <div>
	*       {suggestions.map(suggestion => (
	*         <button key={suggestion.title}>{suggestion.title}</button>
	*       ))}
	*       <button onClick={generateSuggestions} disabled={isLoadingSuggestions}>
	*         Generate Suggestions
	*       </button>
	*     </div>
	*   );
	* }
	* ```
	*
	* ## Return Values
	* The following properties are returned from the hook:
	*
	* <PropertyReference name="messages" type="Message[]">
	* The messages currently in the chat in AG-UI format
	* </PropertyReference>
	*
	* <PropertyReference name="sendMessage" type="(message: Message, options?) => Promise<void>">
	* Send a new message to the chat and trigger AI response
	* </PropertyReference>
	*
	* <PropertyReference name="setMessages" type="(messages: Message[] | DeprecatedGqlMessage[]) => void">
	* Replace all messages in the chat with new array
	* </PropertyReference>
	*
	* <PropertyReference name="deleteMessage" type="(messageId: string) => void">
	* Remove a specific message by ID from the chat
	* </PropertyReference>
	*
	* <PropertyReference name="reloadMessages" type="(messageId: string) => Promise<void>">
	* Regenerate the response for a specific message by ID
	* </PropertyReference>
	*
	* <PropertyReference name="stopGeneration" type="() => void">
	* Stop the current message generation process
	* </PropertyReference>
	*
	* <PropertyReference name="reset" type="() => void">
	* Clear all messages and reset chat state completely
	* </PropertyReference>
	*
	* <PropertyReference name="isLoading" type="boolean">
	* Whether the chat is currently generating a response
	* </PropertyReference>
	*
	* <PropertyReference name="runChatCompletion" type="() => Promise<Message[]>">
	* Manually trigger chat completion for advanced usage
	* </PropertyReference>
	*
	* <PropertyReference name="mcpServers" type="MCPServerConfig[]">
	* Array of Model Context Protocol server configurations
	* </PropertyReference>
	*
	* <PropertyReference name="setMcpServers" type="(servers: MCPServerConfig[]) => void">
	* Update MCP server configurations for enhanced context
	* </PropertyReference>
	*
	* <PropertyReference name="suggestions" type="SuggestionItem[]">
	* Current suggestions array for reading or manual control
	* </PropertyReference>
	*
	* <PropertyReference name="setSuggestions" type="(suggestions: SuggestionItem[]) => void">
	* Manually set suggestions for custom workflows
	* </PropertyReference>
	*
	* <PropertyReference name="generateSuggestions" type="() => Promise<void>">
	* Trigger AI-powered suggestion generation using configured settings
	* </PropertyReference>
	*
	* <PropertyReference name="resetSuggestions" type="() => void">
	* Clear all current suggestions and reset generation state
	* </PropertyReference>
	*
	* <PropertyReference name="isLoadingSuggestions" type="boolean">
	* Whether suggestions are currently being generated
	* </PropertyReference>
	*
	* <PropertyReference name="interrupt" type="string | React.ReactElement | null">
	* Interrupt content for human-in-the-loop workflows
	* </PropertyReference>
	*/
	const createNonFunctionalReturn = () => ({
		visibleMessages: [],
		messages: [],
		sendMessage: async () => {},
		appendMessage: async () => {},
		setMessages: () => {},
		deleteMessage: () => {},
		reloadMessages: async () => {},
		stopGeneration: () => {},
		reset: () => {},
		isLoading: false,
		isAvailable: false,
		runChatCompletion: async () => [],
		mcpServers: [],
		setMcpServers: () => {},
		suggestions: [],
		setSuggestions: () => {},
		generateSuggestions: async () => {},
		resetSuggestions: () => {},
		isLoadingSuggestions: false,
		interrupt: null
	});
	/**
	* Enterprise React hook that provides complete chat functionality for fully custom UI implementations.
	* Includes all advanced features like direct message access, suggestions array, interrupt handling, and MCP support.
	*
	* **Requires a publicApiKey** - Sign up for free at https://cloud.copilotkit.ai/
	*
	* @param options - Configuration options for the chat
	* @returns Complete chat interface with all enterprise features
	*
	* @example
	* ```tsx
	* const { messages, sendMessage, suggestions, interrupt } = useCopilotChatHeadless_c();
	* ```
	*/
	function useCopilotChatHeadless_c(options = {}) {
		const { copilotApiConfig, setBannerError } = useCopilotContext();
		const hasPublicApiKey = Boolean(copilotApiConfig.publicApiKey);
		const internalResult = useCopilotChatInternal(options);
		(0, react.useEffect)(() => {
			if (!hasPublicApiKey) {
				setBannerError(new _copilotkit_shared.CopilotKitError({
					message: "You're using useCopilotChatHeadless_c, a premium-only feature, which offers extensive headless chat capabilities. To continue, you'll need to provide a free public license key.",
					code: _copilotkit_shared.CopilotKitErrorCode.MISSING_PUBLIC_API_KEY_ERROR,
					severity: _copilotkit_shared.Severity.WARNING,
					visibility: _copilotkit_shared.ErrorVisibility.BANNER
				}));
				_copilotkit_shared.styledConsole.logCopilotKitPlatformMessage();
			} else setBannerError(null);
		}, [hasPublicApiKey]);
		if (hasPublicApiKey) return internalResult;
		return createNonFunctionalReturn();
	}

//#endregion
//#region src/hooks/use-frontend-tool.ts
	function useFrontendTool(tool, dependencies) {
		const { name, description, parameters, render, followUp, available } = tool;
		const zodParameters = (0, _copilotkit_shared.getZodParameters)(parameters);
		const renderRef = (0, react.useRef)(render);
		(0, react.useEffect)(() => {
			renderRef.current = render;
		}, [render, ...dependencies !== null && dependencies !== void 0 ? dependencies : []]);
		const normalizedRender = (0, react.useMemo)(() => {
			if (typeof render === "undefined") return;
			return ((args) => {
				const currentRender = renderRef.current;
				if (typeof currentRender === "undefined") return null;
				if (typeof currentRender === "string") return react.default.createElement(react.default.Fragment, null, currentRender);
				const rendered = currentRender({
					...args,
					result: typeof args.result === "string" ? (0, _copilotkit_shared.parseJson)(args.result, args.result) : args.result
				});
				if (typeof rendered === "string") return react.default.createElement(react.default.Fragment, null, rendered);
				return rendered !== null && rendered !== void 0 ? rendered : null;
			});
		}, []);
		const handlerRef = (0, react.useRef)(tool.handler);
		(0, react.useEffect)(() => {
			handlerRef.current = tool.handler;
		}, [tool.handler, ...dependencies !== null && dependencies !== void 0 ? dependencies : []]);
		(0, _copilotkitnext_react.useFrontendTool)({
			name,
			description,
			parameters: zodParameters,
			handler: tool.handler ? (args) => {
				var _handlerRef$current;
				return (_handlerRef$current = handlerRef.current) === null || _handlerRef$current === void 0 ? void 0 : _handlerRef$current.call(handlerRef, args);
			} : void 0,
			followUp,
			render: normalizedRender,
			available: available === void 0 ? void 0 : available !== "disabled"
		});
	}

//#endregion
//#region src/hooks/use-render-tool-call.ts
	function useRenderToolCall(tool, dependencies) {
		const { copilotkit } = (0, _copilotkitnext_react.useCopilotKit)();
		const hasAddedRef = (0, react.useRef)(false);
		(0, react.useEffect)(() => {
			const { name, parameters, render } = tool;
			const zodParameters = (0, _copilotkit_shared.getZodParameters)(parameters);
			const renderToolCall = name === "*" ? (0, _copilotkitnext_react.defineToolCallRenderer)({
				name: "*",
				render: ((args) => {
					return render({
						...args,
						result: args.result ? (0, _copilotkit_shared.parseJson)(args.result, args.result) : args.result
					});
				})
			}) : (0, _copilotkitnext_react.defineToolCallRenderer)({
				name,
				args: zodParameters,
				render: ((args) => {
					return render({
						...args,
						result: args.result ? (0, _copilotkit_shared.parseJson)(args.result, args.result) : args.result
					});
				})
			});
			const existingIndex = copilotkit.renderToolCalls.findIndex((r) => r.name === name);
			if (existingIndex !== -1) copilotkit.renderToolCalls.splice(existingIndex, 1);
			copilotkit.renderToolCalls.push(renderToolCall);
			hasAddedRef.current = true;
			return () => {
				if (hasAddedRef.current) {
					const index = copilotkit.renderToolCalls.findIndex((r) => r.name === name);
					if (index !== -1) copilotkit.renderToolCalls.splice(index, 1);
					hasAddedRef.current = false;
				}
			};
		}, [tool, ...dependencies !== null && dependencies !== void 0 ? dependencies : []]);
	}

//#endregion
//#region src/hooks/use-human-in-the-loop.ts
	function useHumanInTheLoop(tool, dependencies) {
		const { render, ...toolRest } = tool;
		const { name, description, parameters, followUp } = toolRest;
		const zodParameters = (0, _copilotkit_shared.getZodParameters)(parameters);
		const renderRef = (0, react.useRef)(null);
		(0, react.useEffect)(() => {
			renderRef.current = (args) => {
				if (typeof render === "string") return react.default.createElement(react.default.Fragment, null, render);
				if (!render) return null;
				const rendered = render((() => {
					const mappedArgs = args.args;
					switch (args.status) {
						case _copilotkitnext_core.ToolCallStatus.InProgress: return {
							args: mappedArgs,
							respond: args.respond,
							status: args.status,
							handler: void 0
						};
						case _copilotkitnext_core.ToolCallStatus.Executing: return {
							args: mappedArgs,
							respond: args.respond,
							status: args.status,
							handler: () => {}
						};
						case _copilotkitnext_core.ToolCallStatus.Complete: return {
							args: mappedArgs,
							respond: args.respond,
							status: args.status,
							result: args.result ? (0, _copilotkit_shared.parseJson)(args.result, args.result) : args.result,
							handler: void 0
						};
						default: throw new _copilotkit_shared.CopilotKitError({
							code: _copilotkit_shared.CopilotKitErrorCode.UNKNOWN,
							message: `Invalid tool call status: ${args.status}`
						});
					}
				})());
				if (typeof rendered === "string") return react.default.createElement(react.default.Fragment, null, rendered);
				return rendered !== null && rendered !== void 0 ? rendered : null;
			};
		}, [render, ...dependencies !== null && dependencies !== void 0 ? dependencies : []]);
		(0, _copilotkitnext_react.useHumanInTheLoop)({
			name,
			description,
			followUp,
			parameters: zodParameters,
			render: ((args) => {
				var _renderRef$current, _renderRef$current2;
				return (_renderRef$current = (_renderRef$current2 = renderRef.current) === null || _renderRef$current2 === void 0 ? void 0 : _renderRef$current2.call(renderRef, args)) !== null && _renderRef$current !== void 0 ? _renderRef$current : null;
			})
		});
	}

//#endregion
//#region src/hooks/use-copilot-action.ts
/**
	* Example usage of useCopilotAction with complex parameters:
	*
	* @example
	* useCopilotAction({
	*   name: "myAction",
	*   parameters: [
	*     { name: "arg1", type: "string", enum: ["option1", "option2", "option3"], required: false },
	*     { name: "arg2", type: "number" },
	*     {
	*       name: "arg3",
	*       type: "object",
	*       attributes: [
	*         { name: "nestedArg1", type: "boolean" },
	*         { name: "xyz", required: false },
	*       ],
	*     },
	*     { name: "arg4", type: "number[]" },
	*   ],
	*   handler: ({ arg1, arg2, arg3, arg4 }) => {
	*     const x = arg3.nestedArg1;
	*     const z = arg3.xyz;
	*     console.log(arg1, arg2, arg3);
	*   },
	* });
	*
	* @example
	* // Simple action without parameters
	* useCopilotAction({
	*   name: "myAction",
	*   handler: () => {
	*     console.log("No parameters provided.");
	*   },
	* });
	*
	* @example
	* // Interactive action with UI rendering and response handling
	* useCopilotAction({
	*   name: "handleMeeting",
	*   description: "Handle a meeting by booking or canceling",
	*   parameters: [
	*     {
	*       name: "meeting",
	*       type: "string",
	*       description: "The meeting to handle",
	*       required: true,
	*     },
	*     {
	*       name: "date",
	*       type: "string",
	*       description: "The date of the meeting",
	*       required: true,
	*     },
	*     {
	*       name: "title",
	*       type: "string",
	*       description: "The title of the meeting",
	*       required: true,
	*     },
	*   ],
	*   renderAndWaitForResponse: ({ args, respond, status }) => {
	*     const { meeting, date, title } = args;
	*     return (
	*       <MeetingConfirmationDialog
	*         meeting={meeting}
	*         date={date}
	*         title={title}
	*         onConfirm={() => respond('meeting confirmed')}
	*         onCancel={() => respond('meeting canceled')}
	*       />
	*     );
	*   },
	* });
	*
	* @example
	* // Catch all action allows you to render actions that are not defined in the frontend
	* useCopilotAction({
	*   name: "*",
	*   render: ({ name, args, status, result, handler, respond }) => {
	*     return <div>Rendering action: {name}</div>;
	*   },
	* });
	*/
	/**
	* <img src="https://cdn.copilotkit.ai/docs/copilotkit/images/use-copilot-action/useCopilotAction.gif" width="500" />
	* `useCopilotAction` is a React hook that you can use in your application to provide
	* custom actions that can be called by the AI. Essentially, it allows the Copilot to
	* execute these actions contextually during a chat, based on the user's interactions
	* and needs.
	*
	* Here's how it works:
	*
	* Use `useCopilotAction` to set up actions that the Copilot can call. To provide
	* more context to the Copilot, you can provide it with a `description` (for example to explain
	* what the action does, under which conditions it can be called, etc.).
	*
	* Then you define the parameters of the action, which can be simple, e.g. primitives like strings or numbers,
	* or complex, e.g. objects or arrays.
	*
	* Finally, you provide a `handler` function that receives the parameters and returns a result.
	* CopilotKit takes care of automatically inferring the parameter types, so you get type safety
	* and autocompletion for free.
	*
	* To render a custom UI for the action, you can provide a `render()` function. This function
	* lets you render a custom component or return a string to display.
	*
	* ## Usage
	*
	* ### Simple Usage
	*
	* ```tsx
	* useCopilotAction({
	*   name: "sayHello",
	*   description: "Say hello to someone.",
	*   parameters: [
	*     {
	*       name: "name",
	*       type: "string",
	*       description: "name of the person to say greet",
	*     },
	*   ],
	*   handler: async ({ name }) => {
	*     alert(`Hello, ${name}!`);
	*   },
	* });
	* ```
	*
	* ## Generative UI
	*
	* This hooks enables you to dynamically generate UI elements and render them in the copilot chat. For more information, check out the [Generative UI](/guides/generative-ui) page.
	*/
	function getActionConfig(action) {
		if (action.name === "*") return {
			type: "render",
			action
		};
		if ("renderAndWaitForResponse" in action || "renderAndWait" in action) {
			let render = action.render;
			if (!render && "renderAndWaitForResponse" in action) render = action.renderAndWaitForResponse;
			if (!render && "renderAndWait" in action) render = action.renderAndWait;
			return {
				type: "hitl",
				action: {
					...action,
					render
				}
			};
		}
		if ("available" in action) {
			if (action.available === "enabled" || action.available === "remote") return {
				type: "frontend",
				action
			};
			if (action.available === "frontend" || action.available === "disabled") return {
				type: "render",
				action
			};
		}
		if ("handler" in action) return {
			type: "frontend",
			action
		};
		throw new Error("Invalid action configuration");
	}
	/**
	* useCopilotAction is a legacy hook maintained for backwards compatibility.
	*
	* To avoid violating React's Rules of Hooks (which prohibit conditional hook calls),
	* we use a registration pattern:
	* 1. This hook registers the action configuration with the CopilotContext
	* 2. A renderer component in CopilotKit actually renders the appropriate hook wrapper
	* 3. React properly manages hook state since components are rendered, not conditionally called
	*
	* This allows action types to change between renders without corrupting React's hook state.
	*/
	function useCopilotAction(action, dependencies) {
		const [initialActionConfig] = (0, react.useState)(getActionConfig(action));
		const currentActionConfig = getActionConfig(action);
		/**
		* Calling hooks conditionally violates React's Rules of Hooks. This rule exists because
		* React maintains the call stack for hooks like useEffect or useState, and conditionally
		* calling a hook would result in inconsistent call stacks between renders.
		*
		* Unfortunately, useCopilotAction _has_ to conditionally call a hook based on the
		* supplied parameters. In order to avoid breaking React's call stack tracking, while
		* breaking the Rule of Hooks, we use a ref to store the initial action configuration
		* and throw an error if the _configuration_ changes such that we would call a different hook.
		*/
		if (initialActionConfig.type !== currentActionConfig.type) throw new Error("Action configuration changed between renders");
		switch (currentActionConfig.type) {
			case "render": return useRenderToolCall(currentActionConfig.action, dependencies);
			case "hitl": return useHumanInTheLoop(currentActionConfig.action, dependencies);
			case "frontend": return useFrontendTool(currentActionConfig.action, dependencies);
			default: throw new Error("Invalid action configuration");
		}
	}

//#endregion
//#region src/hooks/use-coagent-state-render.ts
/**
	* The useCoAgentStateRender hook allows you to render UI or text based components on a Agentic Copilot's state in the chat.
	* This is particularly useful for showing intermediate state or progress during Agentic Copilot operations.
	*
	* ## Usage
	*
	* ### Simple Usage
	*
	* ```tsx
	* import { useCoAgentStateRender } from "@copilotkit/react-core";
	*
	* type YourAgentState = {
	*   agent_state_property: string;
	* }
	*
	* useCoAgentStateRender<YourAgentState>({
	*   name: "basic_agent",
	*   nodeName: "optionally_specify_a_specific_node",
	*   render: ({ status, state, nodeName }) => {
	*     return (
	*       <YourComponent
	*         agentStateProperty={state.agent_state_property}
	*         status={status}
	*         nodeName={nodeName}
	*       />
	*     );
	*   },
	* });
	* ```
	*
	* This allows for you to render UI components or text based on what is happening within the agent.
	*
	* ### Example
	* A great example of this is in our Perplexity Clone where we render the progress of an agent's internet search as it is happening.
	* You can play around with it below or learn how to build it with its [demo](/coagents/videos/perplexity-clone).
	*
	* <Callout type="info">
	*   This example is hosted on Vercel and may take a few seconds to load.
	* </Callout>
	*
	* <iframe src="https://examples-coagents-ai-researcher-ui.vercel.app/" className="w-full rounded-lg border h-[700px] my-4" />
	*/
	/**
	* This hook is used to render agent state with custom UI components or text. This is particularly
	* useful for showing intermediate state or progress during Agentic Copilot operations.
	* To get started using rendering intermediate state through this hook, checkout the documentation.
	*
	* https://docs.copilotkit.ai/coagents/shared-state/predictive-state-updates
	*/
	function useCoAgentStateRender(action, dependencies) {
		const { chatComponentsCache, availableAgents } = (0, react.useContext)(CopilotContext);
		const { setCoAgentStateRender, removeCoAgentStateRender, coAgentStateRenders } = useCoAgentStateRenders();
		const idRef = (0, react.useRef)((0, _copilotkit_shared.randomId)());
		const { setBannerError, addToast } = useToast();
		(0, react.useEffect)(() => {
			if ((availableAgents === null || availableAgents === void 0 ? void 0 : availableAgents.length) && !availableAgents.some((a) => a.name === action.name)) {
				`${action.name}`;
				setBannerError(new _copilotkit_shared.CopilotKitAgentDiscoveryError({
					agentName: action.name,
					availableAgents: availableAgents.map((a) => ({
						name: a.name,
						id: a.id
					}))
				}));
			}
		}, [availableAgents]);
		const key = `${action.name}-${action.nodeName || "global"}`;
		if (dependencies === void 0) {
			if (coAgentStateRenders[idRef.current]) {
				coAgentStateRenders[idRef.current].handler = action.handler;
				if (typeof action.render === "function") {
					if (chatComponentsCache.current !== null) chatComponentsCache.current.coAgentStateRenders[key] = action.render;
				}
			}
		}
		(0, react.useEffect)(() => {
			const currentId = idRef.current;
			if (Object.entries(coAgentStateRenders).some(([id, otherAction]) => {
				if (id === currentId) return false;
				if (otherAction.name !== action.name) return false;
				const hasNodeName = !!action.nodeName;
				const hasOtherNodeName = !!otherAction.nodeName;
				if (!hasNodeName && !hasOtherNodeName) return true;
				if (hasNodeName !== hasOtherNodeName) return false;
				return action.nodeName === otherAction.nodeName;
			})) addToast({
				type: "warning",
				message: action.nodeName ? `Found multiple state renders for agent ${action.name} and node ${action.nodeName}. State renders might get overridden` : `Found multiple state renders for agent ${action.name}. State renders might get overridden`,
				id: `dup-action-${action.name}`
			});
		}, [coAgentStateRenders]);
		(0, react.useEffect)(() => {
			setCoAgentStateRender(idRef.current, action);
			if (chatComponentsCache.current !== null && action.render !== void 0) chatComponentsCache.current.coAgentStateRenders[key] = action.render;
			return () => {
				removeCoAgentStateRender(idRef.current);
			};
		}, [
			setCoAgentStateRender,
			removeCoAgentStateRender,
			action.name,
			typeof action.render === "string" ? action.render : void 0,
			...dependencies || []
		]);
	}

//#endregion
//#region src/hooks/use-make-copilot-document-readable.ts
/**
	* Makes a document readable by Copilot.
	* @param document The document to make readable.
	* @param categories The categories to associate with the document.
	* @param dependencies The dependencies to use for the effect.
	* @returns The id of the document.
	*/
	function useMakeCopilotDocumentReadable(document, categories, dependencies = []) {
		const { addDocumentContext, removeDocumentContext } = useCopilotContext();
		const idRef = (0, react.useRef)(void 0);
		(0, react.useEffect)(() => {
			const id = addDocumentContext(document, categories);
			idRef.current = id;
			return () => {
				removeDocumentContext(id);
			};
		}, [
			addDocumentContext,
			removeDocumentContext,
			...dependencies
		]);
		return idRef.current;
	}

//#endregion
//#region src/hooks/use-copilot-readable.ts
/**
	* `useCopilotReadable` is a React hook that provides app-state and other information
	* to the Copilot. Optionally, the hook can also handle hierarchical state within your
	* application, passing these parent-child relationships to the Copilot.
	*
	* ## Usage
	*
	* ### Simple Usage
	*
	* In its most basic usage, useCopilotReadable accepts a single string argument
	* representing any piece of app state, making it available for the Copilot to use
	* as context when responding to user input.
	*
	* ```tsx
	* import { useCopilotReadable } from "@copilotkit/react-core";
	*
	* export function MyComponent() {
	*   const [employees, setEmployees] = useState([]);
	*
	*   useCopilotReadable({
	*     description: "The list of employees",
	*     value: employees,
	*   });
	* }
	* ```
	*
	* ### Nested Components
	*
	* Optionally, you can maintain the hierarchical structure of information by passing
	* `parentId`. This allows you to use `useCopilotReadable` in nested components:
	*
	* ```tsx /employeeContextId/1 {17,23}
	* import { useCopilotReadable } from "@copilotkit/react-core";
	*
	* function Employee(props: EmployeeProps) {
	*   const { employeeName, workProfile, metadata } = props;
	*
	*   // propagate any information to copilot
	*   const employeeContextId = useCopilotReadable({
	*     description: "Employee name",
	*     value: employeeName
	*   });
	*
	*   // Pass a parentID to maintain a hierarchical structure.
	*   // Especially useful with child React components, list elements, etc.
	*   useCopilotReadable({
	*     description: "Work profile",
	*     value: workProfile.description(),
	*     parentId: employeeContextId
	*   });
	*
	*   useCopilotReadable({
	*     description: "Employee metadata",
	*     value: metadata.description(),
	*     parentId: employeeContextId
	*   });
	*
	*   return (
	*     // Render as usual...
	*   );
	* }
	* ```
	*/
	/**
	* Adds the given information to the Copilot context to make it readable by Copilot.
	*/
	function useCopilotReadable({ description, value, convert, available }, dependencies) {
		const { copilotkit } = (0, _copilotkitnext_react.useCopilotKit)();
		const ctxIdRef = (0, react.useRef)(void 0);
		(0, react.useEffect)(() => {
			if (!copilotkit) return;
			const found = Object.entries(copilotkit.context).find(([id, ctxItem]) => {
				return JSON.stringify({
					description,
					value
				}) == JSON.stringify(ctxItem);
			});
			if (found) {
				ctxIdRef.current = found[0];
				if (available === "disabled") copilotkit.removeContext(ctxIdRef.current);
				return;
			}
			if (!found && available === "disabled") return;
			ctxIdRef.current = copilotkit.addContext({
				description,
				value: (convert !== null && convert !== void 0 ? convert : JSON.stringify)(value)
			});
			return () => {
				if (!ctxIdRef.current) return;
				copilotkit.removeContext(ctxIdRef.current);
			};
		}, [
			description,
			value,
			convert
		]);
		return ctxIdRef.current;
	}

//#endregion
//#region src/hooks/use-agent-nodename.ts
	function useAgentNodeName(agentName) {
		const { agent } = (0, _copilotkitnext_react.useAgent)({ agentId: agentName });
		const nodeNameRef = (0, react.useRef)("start");
		(0, react.useEffect)(() => {
			if (!agent) return;
			const subscription = agent.subscribe({
				onStepStartedEvent: ({ event }) => {
					nodeNameRef.current = event.stepName;
				},
				onRunStartedEvent: () => {
					nodeNameRef.current = "start";
				},
				onRunFinishedEvent: () => {
					nodeNameRef.current = "end";
				}
			});
			return () => {
				subscription.unsubscribe();
			};
		}, [agent]);
		return nodeNameRef.current;
	}

//#endregion
//#region src/hooks/use-coagent.ts
/**
	* <Callout type="info">
	*   Usage of this hook assumes some additional setup in your application, for more information
	*   on that see the CoAgents <span className="text-blue-500">[getting started guide](/coagents/quickstart/langgraph)</span>.
	* </Callout>
	* <Frame className="my-12">
	*   <img
	*     src="https://cdn.copilotkit.ai/docs/copilotkit/images/coagents/SharedStateCoAgents.gif"
	*     alt="CoAgents demonstration"
	*     className="w-auto"
	*   />
	* </Frame>
	*
	* This hook is used to integrate an agent into your application. With its use, you can
	* render and update the state of an agent, allowing for a dynamic and interactive experience.
	* We call these shared state experiences agentic copilots, or CoAgents for short.
	*
	* ## Usage
	*
	* ### Simple Usage
	*
	* ```tsx
	* import { useCoAgent } from "@copilotkit/react-core";
	*
	* type AgentState = {
	*   count: number;
	* }
	*
	* const agent = useCoAgent<AgentState>({
	*   name: "my-agent",
	*   initialState: {
	*     count: 0,
	*   },
	* });
	*
	* ```
	*
	* `useCoAgent` returns an object with the following properties:
	*
	* ```tsx
	* const {
	*   name,     // The name of the agent currently being used.
	*   nodeName, // The name of the current LangGraph node.
	*   state,    // The current state of the agent.
	*   setState, // A function to update the state of the agent.
	*   running,  // A boolean indicating if the agent is currently running.
	*   start,    // A function to start the agent.
	*   stop,     // A function to stop the agent.
	*   run,      // A function to re-run the agent. Takes a HintFunction to inform the agent why it is being re-run.
	* } = agent;
	* ```
	*
	* Finally we can leverage these properties to create reactive experiences with the agent!
	*
	* ```tsx
	* const { state, setState } = useCoAgent<AgentState>({
	*   name: "my-agent",
	*   initialState: {
	*     count: 0,
	*   },
	* });
	*
	* return (
	*   <div>
	*     <p>Count: {state.count}</p>
	*     <button onClick={() => setState({ count: state.count + 1 })}>Increment</button>
	*   </div>
	* );
	* ```
	*
	* This reactivity is bidirectional, meaning that changes to the state from the agent will be reflected in the UI and vice versa.
	*
	* ## Parameters
	* <PropertyReference name="options" type="UseCoagentOptions<T>" required>
	*   The options to use when creating the coagent.
	*   <PropertyReference name="name" type="string" required>
	*     The name of the agent to use.
	*   </PropertyReference>
	*   <PropertyReference name="initialState" type="T | any">
	*     The initial state of the agent.
	*   </PropertyReference>
	*   <PropertyReference name="state" type="T | any">
	*     State to manage externally if you are using this hook with external state management.
	*   </PropertyReference>
	*   <PropertyReference name="setState" type="(newState: T | ((prevState: T | undefined) => T)) => void">
	*     A function to update the state of the agent if you are using this hook with external state management.
	*   </PropertyReference>
	* </PropertyReference>
	*/
	/**
	* This hook is used to integrate an agent into your application. With its use, you can
	* render and update the state of the agent, allowing for a dynamic and interactive experience.
	* We call these shared state experiences "agentic copilots". To get started using agentic copilots, which
	* we refer to as CoAgents, checkout the documentation at https://docs.copilotkit.ai/coagents/quickstart/langgraph.
	*/
	function useCoAgent(options) {
		const { agent } = (0, _copilotkitnext_react.useAgent)({ agentId: options.name });
		const { copilotkit } = (0, _copilotkitnext_react.useCopilotKit)();
		const nodeName = useAgentNodeName(options.name);
		const handleStateUpdate = (0, react.useCallback)((newState) => {
			if (!agent) return;
			if (typeof newState === "function") {
				const updater = newState;
				agent.setState(updater(agent.state));
			} else agent.setState({
				...agent.state,
				...newState
			});
		}, [agent === null || agent === void 0 ? void 0 : agent.state, agent === null || agent === void 0 ? void 0 : agent.setState]);
		(0, react.useEffect)(() => {
			var _options$config;
			if (!options.config && !options.configurable) return;
			let config = (_options$config = options.config) !== null && _options$config !== void 0 ? _options$config : {};
			if (options.configurable) config = {
				...config,
				configurable: {
					...options.configurable,
					...config.configurable
				}
			};
			copilotkit.setProperties(config);
		}, [options.config, options.configurable]);
		(0, react.useEffect)(() => {
			if ((agent === null || agent === void 0 ? void 0 : agent.state) && isExternalStateManagement(options) && JSON.stringify(options.state) !== JSON.stringify(agent.state)) handleStateUpdate(options.state);
		}, [
			agent,
			(0, react.useMemo)(() => isExternalStateManagement(options) ? JSON.stringify(options.state) : void 0, [isExternalStateManagement(options) ? JSON.stringify(options.state) : void 0]),
			handleStateUpdate
		]);
		const hasStateValues = (0, react.useCallback)((value) => {
			return Boolean(value && Object.keys(value).length);
		}, []);
		const initialStateRef = (0, react.useRef)(isExternalStateManagement(options) ? options.state : "initialState" in options ? options.initialState : void 0);
		(0, react.useEffect)(() => {
			if (isExternalStateManagement(options)) initialStateRef.current = options.state;
			else if ("initialState" in options) initialStateRef.current = options.initialState;
		}, [isExternalStateManagement(options) ? JSON.stringify(options.state) : "initialState" in options ? JSON.stringify(options.initialState) : void 0]);
		(0, react.useEffect)(() => {
			if (!agent) return;
			const subscription = agent.subscribe({
				onStateChanged: (args) => {
					if (isExternalStateManagement(options)) options.setState(args.state);
				},
				onRunInitialized: (args) => {
					if (hasStateValues(args.state)) {
						handleStateUpdate(args.state);
						return;
					}
					if (hasStateValues(agent.state)) return;
					if (initialStateRef.current !== void 0) handleStateUpdate(initialStateRef.current);
				}
			});
			return () => {
				subscription.unsubscribe();
			};
		}, [
			agent,
			handleStateUpdate,
			hasStateValues
		]);
		return (0, react.useMemo)(() => {
			var _agent$agentId;
			if (!agent) {
				var _ref, _ref2;
				const noop = () => {};
				const noopAsync = async () => {};
				const initialState = (_ref = (_ref2 = "state" in options && options.state) !== null && _ref2 !== void 0 ? _ref2 : "initialState" in options && options.initialState) !== null && _ref !== void 0 ? _ref : {};
				return {
					name: options.name,
					nodeName,
					threadId: void 0,
					running: false,
					state: initialState,
					setState: noop,
					start: noop,
					stop: noop,
					run: noopAsync
				};
			}
			return {
				name: (_agent$agentId = agent === null || agent === void 0 ? void 0 : agent.agentId) !== null && _agent$agentId !== void 0 ? _agent$agentId : options.name,
				nodeName,
				threadId: agent.threadId,
				running: agent.isRunning,
				state: agent.state,
				setState: handleStateUpdate,
				start: agent.runAgent,
				stop: agent.abortRun,
				run: agent.runAgent
			};
		}, [
			agent === null || agent === void 0 ? void 0 : agent.state,
			agent === null || agent === void 0 ? void 0 : agent.runAgent,
			agent === null || agent === void 0 ? void 0 : agent.abortRun,
			agent === null || agent === void 0 ? void 0 : agent.runAgent,
			agent === null || agent === void 0 ? void 0 : agent.threadId,
			agent === null || agent === void 0 ? void 0 : agent.isRunning,
			agent === null || agent === void 0 ? void 0 : agent.agentId,
			handleStateUpdate,
			options.name
		]);
	}
	const isExternalStateManagement = (options) => {
		return "state" in options && "setState" in options;
	};

//#endregion
//#region src/hooks/use-copilot-runtime-client.ts
	const useCopilotRuntimeClient = (options) => {
		const { setBannerError } = useToast();
		const { showDevConsole, onError, ...runtimeOptions } = options;
		const lastStructuredErrorRef = (0, react.useRef)(null);
		const traceUIError = async (error, originalError) => {
			try {
				await onError({
					type: "error",
					timestamp: Date.now(),
					context: {
						source: "ui",
						request: {
							operation: "runtimeClient",
							url: runtimeOptions.url,
							startTime: Date.now()
						},
						technical: {
							environment: "browser",
							userAgent: typeof navigator !== "undefined" ? navigator.userAgent : void 0,
							stackTrace: originalError instanceof Error ? originalError.stack : void 0
						}
					},
					error
				});
			} catch (error) {
				console.error("Error in onError handler:", error);
			}
		};
		return (0, react.useMemo)(() => {
			return new _copilotkit_runtime_client_gql.CopilotRuntimeClient({
				...runtimeOptions,
				handleGQLErrors: (error) => {
					var _graphQLErrors;
					if ((_graphQLErrors = error.graphQLErrors) === null || _graphQLErrors === void 0 ? void 0 : _graphQLErrors.length) {
						const graphQLErrors = error.graphQLErrors;
						const routeError = (gqlError) => {
							const extensions = gqlError.extensions;
							if ((extensions === null || extensions === void 0 ? void 0 : extensions.visibility) === _copilotkit_shared.ErrorVisibility.SILENT) {
								console.error("CopilotKit Silent Error:", gqlError.message);
								return;
							}
							const now = Date.now();
							const errorMessage = gqlError.message;
							if (lastStructuredErrorRef.current && lastStructuredErrorRef.current.message === errorMessage && now - lastStructuredErrorRef.current.timestamp < 150) return;
							lastStructuredErrorRef.current = {
								message: errorMessage,
								timestamp: now
							};
							const ckError = createStructuredError(gqlError);
							if (ckError) {
								setBannerError(ckError);
								traceUIError(ckError, gqlError);
							} else {
								const fallbackError = new _copilotkit_shared.CopilotKitError({
									message: gqlError.message,
									code: _copilotkit_shared.CopilotKitErrorCode.UNKNOWN
								});
								setBannerError(fallbackError);
								traceUIError(fallbackError, gqlError);
							}
						};
						graphQLErrors.forEach(routeError);
					} else {
						const fallbackError = new _copilotkit_shared.CopilotKitError({
							message: (error === null || error === void 0 ? void 0 : error.message) || String(error),
							code: _copilotkit_shared.CopilotKitErrorCode.UNKNOWN
						});
						setBannerError(fallbackError);
						traceUIError(fallbackError, error);
					}
				},
				handleGQLWarning: (message) => {
					console.warn(message);
					setBannerError(new _copilotkit_shared.CopilotKitError({
						message,
						code: _copilotkit_shared.CopilotKitErrorCode.UNKNOWN
					}));
				}
			});
		}, [
			runtimeOptions,
			setBannerError,
			onError
		]);
	};
	function createStructuredError(gqlError) {
		var _originalError$stack, _originalError$stack2, _originalError$stack3;
		const extensions = gqlError.extensions;
		const originalError = extensions === null || extensions === void 0 ? void 0 : extensions.originalError;
		const message = (originalError === null || originalError === void 0 ? void 0 : originalError.message) || gqlError.message;
		const code = extensions === null || extensions === void 0 ? void 0 : extensions.code;
		if (code) return new _copilotkit_shared.CopilotKitError({
			message,
			code
		});
		if (originalError === null || originalError === void 0 || (_originalError$stack = originalError.stack) === null || _originalError$stack === void 0 ? void 0 : _originalError$stack.includes("CopilotApiDiscoveryError")) return new _copilotkit_shared.CopilotKitApiDiscoveryError({ message });
		if (originalError === null || originalError === void 0 || (_originalError$stack2 = originalError.stack) === null || _originalError$stack2 === void 0 ? void 0 : _originalError$stack2.includes("CopilotKitRemoteEndpointDiscoveryError")) return new _copilotkit_shared.CopilotKitRemoteEndpointDiscoveryError({ message });
		if (originalError === null || originalError === void 0 || (_originalError$stack3 = originalError.stack) === null || _originalError$stack3 === void 0 ? void 0 : _originalError$stack3.includes("CopilotKitAgentDiscoveryError")) return new _copilotkit_shared.CopilotKitAgentDiscoveryError({
			agentName: "",
			availableAgents: []
		});
		return null;
	}

//#endregion
//#region src/hooks/use-copilot-authenticated-action.ts
/**
	* Hook to create an authenticated action that requires user sign-in before execution.
	*
	* @remarks
	* This feature is only available when using CopilotKit's hosted cloud service.
	* To use this feature, sign up at https://cloud.copilotkit.ai to get your publicApiKey.
	*
	* @param action - The frontend action to be wrapped with authentication
	* @param dependencies - Optional array of dependencies that will trigger recreation of the action when changed
	*/
	function useCopilotAuthenticatedAction_c(action, dependencies) {
		const { authConfig_c, authStates_c, setAuthStates_c } = useCopilotContext();
		const pendingActionRef = (0, react.useRef)(null);
		const executeAction = (0, react.useCallback)((props) => {
			if (typeof action.render === "function") return action.render(props);
			return action.render || react.default.createElement(react.Fragment);
		}, [action]);
		const wrappedRender = (0, react.useCallback)((props) => {
			if (!Object.values(authStates_c || {}).some((state) => state.status === "authenticated")) {
				pendingActionRef.current = props;
				return (authConfig_c === null || authConfig_c === void 0 ? void 0 : authConfig_c.SignInComponent) ? react.default.createElement(authConfig_c.SignInComponent, { onSignInComplete: (authState) => {
					setAuthStates_c === null || setAuthStates_c === void 0 || setAuthStates_c((prev) => ({
						...prev,
						[action.name]: authState
					}));
					if (pendingActionRef.current) {
						executeAction(pendingActionRef.current);
						pendingActionRef.current = null;
					}
				} }) : react.default.createElement(react.Fragment);
			}
			return executeAction(props);
		}, [
			action,
			authStates_c,
			setAuthStates_c
		]);
		useCopilotAction({
			...action,
			render: wrappedRender
		}, dependencies);
	}

//#endregion
//#region src/hooks/use-langgraph-interrupt.ts
/**
	* Transforms a v2 InterruptEvent into the v1 LangGraphInterruptEvent shape
	* expected by existing useLangGraphInterrupt callbacks.
	*/
	function toV1Event(event) {
		const value = typeof event.value === "string" ? (0, _copilotkit_shared.parseJson)(event.value, event.value) : event.value;
		return {
			name: _copilotkit_runtime_client_gql.MetaEventName.LangGraphInterruptEvent,
			type: "MetaEvent",
			value
		};
	}
	function useLangGraphInterrupt(action, _dependencies) {
		var _ref, _action$agentId;
		const actionRef = (0, react.useRef)(action);
		actionRef.current = action;
		const existingConfig = (0, _copilotkitnext_react.useCopilotChatConfiguration)();
		const resolvedAgentId = (_ref = (_action$agentId = action.agentId) !== null && _action$agentId !== void 0 ? _action$agentId : existingConfig === null || existingConfig === void 0 ? void 0 : existingConfig.agentId) !== null && _ref !== void 0 ? _ref : "default";
		const threadId = existingConfig === null || existingConfig === void 0 ? void 0 : existingConfig.threadId;
		const nodeName = useAgentNodeName(resolvedAgentId);
		const metadataRef = (0, react.useRef)({
			agentName: resolvedAgentId,
			threadId,
			nodeName
		});
		metadataRef.current = {
			agentName: resolvedAgentId,
			threadId,
			nodeName
		};
		(0, _copilotkitnext_react.useInterrupt)({
			render: (0, react.useCallback)(({ event, result, resolve }) => {
				const renderFn = actionRef.current.render;
				if (!renderFn) return react.default.createElement(react.default.Fragment);
				const rendered = renderFn({
					event: toV1Event(event),
					result,
					resolve: (r) => resolve(r)
				});
				if (typeof rendered === "string") return react.default.createElement(react.default.Fragment, null, rendered);
				return rendered;
			}, []),
			handler: (0, react.useCallback)(({ event, resolve }) => {
				var _actionRef$current$ha, _actionRef$current;
				return (_actionRef$current$ha = (_actionRef$current = actionRef.current).handler) === null || _actionRef$current$ha === void 0 ? void 0 : _actionRef$current$ha.call(_actionRef$current, {
					event: toV1Event(event),
					resolve: (r) => resolve(r)
				});
			}, []),
			enabled: (0, react.useCallback)((event) => {
				if (!actionRef.current.enabled) return true;
				return actionRef.current.enabled({
					eventValue: toV1Event(event).value,
					agentMetadata: metadataRef.current
				});
			}, []),
			agentId: resolvedAgentId
		});
	}

//#endregion
//#region src/hooks/use-copilot-additional-instructions.ts
/**
	* `useCopilotAdditionalInstructions` is a React hook that provides additional instructions
	* to the Copilot.
	*
	* ## Usage
	*
	* ### Simple Usage
	*
	* In its most basic usage, useCopilotAdditionalInstructions accepts a single string argument
	* representing the instructions to be added to the Copilot.
	*
	* ```tsx
	* import { useCopilotAdditionalInstructions } from "@copilotkit/react-core";
	*
	* export function MyComponent() {
	*   useCopilotAdditionalInstructions({
	*     instructions: "Do not answer questions about the weather.",
	*   });
	* }
	* ```
	*
	* ### Conditional Usage
	*
	* You can also conditionally add instructions based on the state of your app.
	*
	* ```tsx
	* import { useCopilotAdditionalInstructions } from "@copilotkit/react-core";
	*
	* export function MyComponent() {
	*   const [showInstructions, setShowInstructions] = useState(false);
	*
	*   useCopilotAdditionalInstructions({
	*     available: showInstructions ? "enabled" : "disabled",
	*     instructions: "Do not answer questions about the weather.",
	*   });
	* }
	* ```
	*/
	/**
	* Adds the given instructions to the Copilot context.
	*/
	function useCopilotAdditionalInstructions({ instructions, available = "enabled" }, dependencies) {
		const { setAdditionalInstructions } = useCopilotContext();
		(0, react.useEffect)(() => {
			if (available === "disabled") return;
			setAdditionalInstructions((prevInstructions) => [...prevInstructions || [], instructions]);
			return () => {
				setAdditionalInstructions((prevInstructions) => (prevInstructions === null || prevInstructions === void 0 ? void 0 : prevInstructions.filter((instruction) => instruction !== instructions)) || []);
			};
		}, [
			available,
			instructions,
			setAdditionalInstructions,
			...dependencies || []
		]);
	}

//#endregion
//#region src/hooks/use-default-tool.ts
	function useDefaultTool(tool, dependencies) {
		useCopilotAction({
			...tool,
			name: "*"
		}, dependencies);
	}

//#endregion
//#region src/hooks/use-copilot-chat-suggestions.tsx
/**
	* <Callout type="warning">
	*   useCopilotChatSuggestions is experimental. The interface is not final and
	*   can change without notice.
	* </Callout>
	*
	* `useCopilotReadable` is a React hook that provides app-state and other information
	* to the Copilot. Optionally, the hook can also handle hierarchical state within your
	* application, passing these parent-child relationships to the Copilot.
	*
	* <br/>
	* <img src="https://cdn.copilotkit.ai/docs/copilotkit/images/use-copilot-chat-suggestions/use-copilot-chat-suggestions.gif" width="500" />
	*
	* ## Usage
	*
	* ### Install Dependencies
	*
	* This component is part of the [@copilotkit/react-ui](https://npmjs.com/package/@copilotkit/react-ui) package.
	*
	* ```shell npm2yarn \"@copilotkit/react-ui"\
	* npm install @copilotkit/react-core @copilotkit/react-ui
	* ```
	*
	* ### Simple Usage
	*
	* ```tsx
	* import { useCopilotChatSuggestions } from "@copilotkit/react-ui";
	*
	* export function MyComponent() {
	*   const [employees, setEmployees] = useState([]);
	*
	*   useCopilotChatSuggestions({
	*     instructions: `The following employees are on duty: ${JSON.stringify(employees)}`,
	*   });
	* }
	* ```
	*
	* ### Dependency Management
	*
	* ```tsx
	* import { useCopilotChatSuggestions } from "@copilotkit/react-ui";
	*
	* export function MyComponent() {
	*   useCopilotChatSuggestions(
	*     {
	*       instructions: "Suggest the most relevant next actions.",
	*     },
	*     [appState],
	*   );
	* }
	* ```
	*
	* In the example above, the suggestions are generated based on the given instructions.
	* The hook monitors `appState`, and updates suggestions accordingly whenever it changes.
	*
	* ### Behavior and Lifecycle
	*
	* The hook registers the configuration with the chat context upon component mount and
	* removes it on unmount, ensuring a clean and efficient lifecycle management.
	*/
	function useCopilotChatSuggestions(config, dependencies = []) {
		var _existingConfig$agent, _ref;
		const existingConfig = (0, _copilotkitnext_react.useCopilotChatConfiguration)();
		const resolvedAgentId = (_existingConfig$agent = existingConfig === null || existingConfig === void 0 ? void 0 : existingConfig.agentId) !== null && _existingConfig$agent !== void 0 ? _existingConfig$agent : "default";
		const available = (_ref = config.available === "enabled" ? "always" : config.available) !== null && _ref !== void 0 ? _ref : "before-first-message";
		(0, _copilotkitnext_react.useConfigureSuggestions)({
			...config,
			available,
			consumerAgentId: resolvedAgentId
		}, dependencies);
	}

//#endregion
//#region src/types/frontend-action.ts
	function processActionsForRuntimeRequest(actions) {
		return actions.filter((action) => action.available !== _copilotkit_runtime_client_gql.ActionInputAvailability.Disabled && action.disabled !== true && action.name !== "*" && action.available != "frontend" && !action.pairedAction).map((action) => {
			let available = _copilotkit_runtime_client_gql.ActionInputAvailability.Enabled;
			if (action.disabled) available = _copilotkit_runtime_client_gql.ActionInputAvailability.Disabled;
			else if (action.available === "disabled") available = _copilotkit_runtime_client_gql.ActionInputAvailability.Disabled;
			else if (action.available === "remote") available = _copilotkit_runtime_client_gql.ActionInputAvailability.Remote;
			return {
				name: action.name,
				description: action.description || "",
				jsonSchema: JSON.stringify((0, _copilotkit_shared.actionParametersToJsonSchema)(action.parameters || [])),
				available
			};
		});
	}

//#endregion
//#region src/lib/copilot-task.ts
/**
	* This class is used to execute one-off tasks, for example on button press. It can use the context available via [useCopilotReadable](/reference/v1/hooks/useCopilotReadable) and the actions provided by [useCopilotAction](/reference/v1/hooks/useCopilotAction), or you can provide your own context and actions.
	*
	* ## Example
	* In the simplest case, use CopilotTask in the context of your app by giving it instructions on what to do.
	*
	* ```tsx
	* import { CopilotTask, useCopilotContext } from "@copilotkit/react-core";
	*
	* export function MyComponent() {
	*   const context = useCopilotContext();
	*
	*   const task = new CopilotTask({
	*     instructions: "Set a random message",
	*     actions: [
	*       {
	*         name: "setMessage",
	*       description: "Set the message.",
	*       argumentAnnotations: [
	*         {
	*           name: "message",
	*           type: "string",
	*           description:
	*             "A message to display.",
	*           required: true,
	*         },
	*       ],
	*      }
	*     ]
	*   });
	*
	*   const executeTask = async () => {
	*     await task.run(context, action);
	*   }
	*
	*   return (
	*     <>
	*       <button onClick={executeTask}>
	*         Execute task
	*       </button>
	*     </>
	*   )
	* }
	* ```
	*
	* Have a look at the [Presentation Example App](https://github.com/CopilotKit/CopilotKit/blob/main/src/v1.x/examples/next-openai/src/app/presentation/page.tsx) for a more complete example.
	*/
	var CopilotTask = class {
		constructor(config) {
			this.instructions = config.instructions;
			this.actions = config.actions || [];
			this.includeCopilotReadable = config.includeCopilotReadable !== false;
			this.includeCopilotActions = config.includeCopilotActions !== false;
			this.forwardedParameters = config.forwardedParameters;
		}
		/**
		* Run the task.
		* @param context The CopilotContext to use for the task. Use `useCopilotContext` to obtain the current context.
		* @param data The data to use for the task.
		*/
		async run(context, data) {
			var _this$forwardedParame, _response$data;
			const actions = this.includeCopilotActions ? Object.assign({}, context.actions) : {};
			for (const fn of this.actions) actions[fn.name] = fn;
			let contextString = "";
			if (data) contextString = (typeof data === "string" ? data : JSON.stringify(data)) + "\n\n";
			if (this.includeCopilotReadable) contextString += context.getContextString([], defaultCopilotContextCategories);
			const messages = [new _copilotkit_runtime_client_gql.TextMessage({
				content: taskSystemMessage(contextString, this.instructions),
				role: _copilotkit_runtime_client_gql.Role.System
			})];
			const response = await new _copilotkit_runtime_client_gql.CopilotRuntimeClient({
				url: context.copilotApiConfig.chatApiEndpoint,
				publicApiKey: context.copilotApiConfig.publicApiKey,
				headers: context.copilotApiConfig.headers,
				credentials: context.copilotApiConfig.credentials
			}).generateCopilotResponse({
				data: {
					frontend: {
						actions: processActionsForRuntimeRequest(Object.values(actions)),
						url: window.location.href
					},
					messages: (0, _copilotkit_runtime_client_gql.convertMessagesToGqlInput)((0, _copilotkit_runtime_client_gql.filterAgentStateMessages)(messages)),
					metadata: { requestType: _copilotkit_runtime_client_gql.CopilotRequestType.Task },
					forwardedParameters: {
						toolChoice: "required",
						...(_this$forwardedParame = this.forwardedParameters) !== null && _this$forwardedParame !== void 0 ? _this$forwardedParame : {}
					}
				},
				properties: context.copilotApiConfig.properties
			}).toPromise();
			const functionCallHandler = context.getFunctionCallHandler(actions);
			const functionCalls = (0, _copilotkit_runtime_client_gql.convertGqlOutputToMessages)(((_response$data = response.data) === null || _response$data === void 0 || (_response$data = _response$data.generateCopilotResponse) === null || _response$data === void 0 ? void 0 : _response$data.messages) || []).filter((m) => m.isActionExecutionMessage());
			for (const functionCall of functionCalls) await functionCallHandler({
				messages,
				name: functionCall.name,
				args: functionCall.arguments
			});
		}
	};
	function taskSystemMessage(contextString, instructions) {
		return `
Please act as an efficient, competent, conscientious, and industrious professional assistant.

Help the user achieve their goals, and you do so in a way that is as efficient as possible, without unnecessary fluff, but also without sacrificing professionalism.
Always be polite and respectful, and prefer brevity over verbosity.

The user has provided you with the following context:
\`\`\`
${contextString}
\`\`\`

They have also provided you with functions you can call to initiate actions on their behalf.

Please assist them as best you can.

This is not a conversation, so please do not ask questions. Just call a function without saying anything else.

The user has given you the following task to complete:

\`\`\`
${instructions}
\`\`\`
`;
	}

//#endregion
exports.CoAgentStateRendersContext = CoAgentStateRendersContext;
exports.CoAgentStateRendersProvider = CoAgentStateRendersProvider;
exports.CopilotContext = CopilotContext;
exports.CopilotKit = CopilotKit;
exports.CopilotMessagesContext = CopilotMessagesContext;
exports.CopilotTask = CopilotTask;
exports.SUGGESTION_RETRY_CONFIG = SUGGESTION_RETRY_CONFIG;
exports.ThreadsContext = ThreadsContext;
exports.ThreadsProvider = ThreadsProvider;
exports.defaultCopilotContextCategories = defaultCopilotContextCategories;
exports.shouldShowDevConsole = shouldShowDevConsole;
exports.useCoAgent = useCoAgent;
exports.useCoAgentStateRender = useCoAgentStateRender;
exports.useCoAgentStateRenders = useCoAgentStateRenders;
exports.useCopilotAction = useCopilotAction;
exports.useCopilotAdditionalInstructions = useCopilotAdditionalInstructions;
exports.useCopilotAuthenticatedAction_c = useCopilotAuthenticatedAction_c;
exports.useCopilotChat = useCopilotChat;
exports.useCopilotChatHeadless_c = useCopilotChatHeadless_c;
exports.useCopilotChatInternal = useCopilotChatInternal;
exports.useCopilotChatSuggestions = useCopilotChatSuggestions;
exports.useCopilotContext = useCopilotContext;
exports.useCopilotMessagesContext = useCopilotMessagesContext;
exports.useCopilotReadable = useCopilotReadable;
exports.useCopilotRuntimeClient = useCopilotRuntimeClient;
exports.useDefaultTool = useDefaultTool;
exports.useFrontendTool = useFrontendTool;
exports.useHumanInTheLoop = useHumanInTheLoop;
exports.useLangGraphInterrupt = useLangGraphInterrupt;
exports.useLazyToolRenderer = useLazyToolRenderer;
exports.useMakeCopilotDocumentReadable = useMakeCopilotDocumentReadable;
exports.useRenderToolCall = useRenderToolCall;
exports.useThreads = useThreads;
});
