// @copilotkit/a2ui-renderer/dist/react-renderer/components/content/AudioPlayer.mjs
import { classMapToString, stylesToObject } from "../../theme/utils.mjs";
import "../../lib/utils.mjs";
import { useA2UIComponent } from "../../hooks/useA2UIComponent.mjs";
import { memo } from "react";
import { jsx, jsxs } from "react/jsx-runtime";

//#region src/react-renderer/components/content/AudioPlayer.tsx
/**
* AudioPlayer component - renders an audio player with optional description.
*/
const AudioPlayer = memo(function AudioPlayer({ node, surfaceId }) {
	const { theme, resolveString } = useA2UIComponent(node, surfaceId);
	const props = node.properties;
	const url = resolveString(props.url);
	const description = resolveString(props.description ?? null);
	if (!url) return null;
	return /* @__PURE__ */ jsx("div", {
		className: "a2ui-audio",
		style: node.weight !== void 0 ? { "--weight": node.weight } : {},
		children: /* @__PURE__ */ jsxs("section", {
			className: classMapToString(theme.components.AudioPlayer),
			style: stylesToObject(theme.additionalStyles?.AudioPlayer),
			children: [description && /* @__PURE__ */ jsx("p", { children: description }), /* @__PURE__ */ jsx("audio", {
				src: url,
				controls: true
			})]
		})
	});
});

//#endregion
export { AudioPlayer as default };
