// @copilotkit/a2ui-renderer/dist/react-renderer/components/content/AudioPlayer.cjs
const require_runtime = require('../../../_virtual/_rolldown/runtime.cjs');
const require_utils = require('../../theme/utils.cjs');
require('../../lib/utils.cjs');
const require_useA2UIComponent = require('../../hooks/useA2UIComponent.cjs');
let react = require("react");
let react_jsx_runtime = require("react/jsx-runtime");

//#region src/react-renderer/components/content/AudioPlayer.tsx
/**
* AudioPlayer component - renders an audio player with optional description.
*/
const AudioPlayer = (0, react.memo)(function AudioPlayer({ node, surfaceId }) {
	const { theme, resolveString } = require_useA2UIComponent.useA2UIComponent(node, surfaceId);
	const props = node.properties;
	const url = resolveString(props.url);
	const description = resolveString(props.description ?? null);
	if (!url) return null;
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
		className: "a2ui-audio",
		style: node.weight !== void 0 ? { "--weight": node.weight } : {},
		children: /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("section", {
			className: require_utils.classMapToString(theme.components.AudioPlayer),
			style: require_utils.stylesToObject(theme.additionalStyles?.AudioPlayer),
			children: [description && /* @__PURE__ */ (0, react_jsx_runtime.jsx)("p", { children: description }), /* @__PURE__ */ (0, react_jsx_runtime.jsx)("audio", {
				src: url,
				controls: true
			})]
		})
	});
});

//#endregion
exports.default = AudioPlayer;
