// @copilotkit/a2ui-renderer/dist/react-renderer/components/content/Image.cjs
const require_runtime = require('../../../_virtual/_rolldown/runtime.cjs');
const require_utils = require('../../theme/utils.cjs');
const require_utils$1 = require('../../lib/utils.cjs');
const require_useA2UIComponent = require('../../hooks/useA2UIComponent.cjs');
let react = require("react");
let react_jsx_runtime = require("react/jsx-runtime");

//#region src/react-renderer/components/content/Image.tsx
/**
* Image component - renders an image from a URL with optional sizing and fit modes.
*
* Supports usageHint values: icon, avatar, smallFeature, mediumFeature, largeFeature, header
* Supports fit values: contain, cover, fill, none, scale-down (maps to object-fit via CSS variable)
*/
const Image = (0, react.memo)(function Image({ node, surfaceId }) {
	const { theme, resolveString } = require_useA2UIComponent.useA2UIComponent(node, surfaceId);
	const props = node.properties;
	const url = resolveString(props.url)?.replace("//via.placeholder.com/", "//placehold.co/");
	const usageHint = props.usageHint;
	const fit = props.fit ?? "fill";
	const classes = require_utils$1.mergeClassMaps(theme.components.Image.all, usageHint ? theme.components.Image[usageHint] : {});
	const style = {
		...require_utils.stylesToObject(theme.additionalStyles?.Image),
		"--object-fit": fit
	};
	if (!url) return null;
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
		className: "a2ui-image",
		style: node.weight !== void 0 ? { "--weight": node.weight } : {},
		children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)("section", {
			className: require_utils.classMapToString(classes),
			style,
			children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)("img", {
				src: url,
				alt: ""
			})
		})
	});
});

//#endregion
exports.default = Image;
