// @copilotkit/a2ui-renderer/dist/react-renderer/components/content/Image.mjs
import { classMapToString, stylesToObject } from "../../theme/utils.mjs";
import { mergeClassMaps } from "../../lib/utils.mjs";
import { useA2UIComponent } from "../../hooks/useA2UIComponent.mjs";
import { memo } from "react";
import { jsx } from "react/jsx-runtime";

//#region src/react-renderer/components/content/Image.tsx
/**
* Image component - renders an image from a URL with optional sizing and fit modes.
*
* Supports usageHint values: icon, avatar, smallFeature, mediumFeature, largeFeature, header
* Supports fit values: contain, cover, fill, none, scale-down (maps to object-fit via CSS variable)
*/
const Image = memo(function Image({ node, surfaceId }) {
	const { theme, resolveString } = useA2UIComponent(node, surfaceId);
	const props = node.properties;
	const url = resolveString(props.url)?.replace("//via.placeholder.com/", "//placehold.co/");
	const usageHint = props.usageHint;
	const fit = props.fit ?? "fill";
	const classes = mergeClassMaps(theme.components.Image.all, usageHint ? theme.components.Image[usageHint] : {});
	const style = {
		...stylesToObject(theme.additionalStyles?.Image),
		"--object-fit": fit
	};
	if (!url) return null;
	return /* @__PURE__ */ jsx("div", {
		className: "a2ui-image",
		style: node.weight !== void 0 ? { "--weight": node.weight } : {},
		children: /* @__PURE__ */ jsx("section", {
			className: classMapToString(classes),
			style,
			children: /* @__PURE__ */ jsx("img", {
				src: url,
				alt: ""
			})
		})
	});
});

//#endregion
export { Image as default };
