// @copilotkit/a2ui-renderer/dist/react-renderer/components/content/Text.mjs
import { classMapToString, stylesToObject } from "../../theme/utils.mjs";
import { mergeClassMaps } from "../../lib/utils.mjs";
import { useA2UIComponent } from "../../hooks/useA2UIComponent.mjs";
import { memo, useMemo } from "react";
import { jsx } from "react/jsx-runtime";
import MarkdownIt from "markdown-it";

//#region src/react-renderer/components/content/Text.tsx
function isHintedStyles(styles) {
	if (typeof styles !== "object" || !styles || Array.isArray(styles)) return false;
	return [
		"h1",
		"h2",
		"h3",
		"h4",
		"h5",
		"caption",
		"body"
	].some((v) => v in styles);
}
/**
* Markdown-it instance for rendering markdown text.
* Uses synchronous import to ensure availability at first render (matches Lit renderer).
*
* Configuration matches Lit's markdown directive (uses MarkdownIt defaults):
* - html: false (default) - Security: disable raw HTML
* - linkify: false (default) - Don't auto-convert URLs/emails to links
* - breaks: false (default) - Don't convert \n to <br>
* - typographer: false (default) - Don't use smart quotes/dashes
*/
const markdownRenderer = new MarkdownIt();
/**
* Apply theme classes to markdown HTML elements.
* Replaces default element tags with themed versions.
*/
function applyMarkdownTheme(html, markdownTheme) {
	if (!markdownTheme) return html;
	const replacements = [];
	for (const [element, classes] of Object.entries(markdownTheme)) {
		if (!classes || Array.isArray(classes) && classes.length === 0) continue;
		const classString = Array.isArray(classes) ? classes.join(" ") : classMapToString(classes);
		if (!classString) continue;
		const tagRegex = new RegExp(`<${element}(?=\\s|>|/>)`, "gi");
		replacements.push([tagRegex, `<${element} class="${classString}"`]);
	}
	let result = html;
	for (const [regex, replacement] of replacements) result = result.replace(regex, replacement);
	return result;
}
/**
* Text component - renders text content with markdown support.
*
* Structure mirrors Lit's Text component:
*   <div class="a2ui-text">      ← :host equivalent
*     <section class="...">      ← theme classes
*       <h2>...</h2>             ← rendered markdown content
*     </section>
*   </div>
*
* Text is parsed as markdown and rendered as HTML (matches Lit renderer behavior).
* Supports usageHint values: h1, h2, h3, h4, h5, caption, body
*
* Markdown features supported:
* - **Bold** and *italic* text
* - Lists (ordered and unordered)
* - `inline code` and code blocks
* - [Links](url) (auto-linkified URLs too)
* - Blockquotes
* - Horizontal rules
*
* Note: Raw HTML is disabled for security.
*/
const Text = memo(function Text({ node, surfaceId }) {
	const { theme, resolveString } = useA2UIComponent(node, surfaceId);
	const props = node.properties;
	const textValue = resolveString(props.text);
	const usageHint = props.usageHint;
	const classes = mergeClassMaps(theme.components.Text.all, usageHint ? theme.components.Text[usageHint] : {});
	const additionalStyles = useMemo(() => {
		const textStyles = theme.additionalStyles?.Text;
		if (!textStyles) return void 0;
		if (isHintedStyles(textStyles)) return stylesToObject(textStyles[usageHint ?? "body"]);
		return stylesToObject(textStyles);
	}, [theme.additionalStyles?.Text, usageHint]);
	const renderedContent = useMemo(() => {
		if (textValue === null || textValue === void 0) return null;
		let markdownText = textValue;
		switch (usageHint) {
			case "h1":
				markdownText = `# ${markdownText}`;
				break;
			case "h2":
				markdownText = `## ${markdownText}`;
				break;
			case "h3":
				markdownText = `### ${markdownText}`;
				break;
			case "h4":
				markdownText = `#### ${markdownText}`;
				break;
			case "h5":
				markdownText = `##### ${markdownText}`;
				break;
			case "caption":
				markdownText = `*${markdownText}*`;
				break;
			default: break;
		}
		return { __html: applyMarkdownTheme(markdownRenderer.render(markdownText), theme.markdown) };
	}, [
		textValue,
		theme.markdown,
		usageHint
	]);
	if (!renderedContent) return null;
	return /* @__PURE__ */ jsx("div", {
		className: "a2ui-text",
		style: node.weight !== void 0 ? { "--weight": node.weight } : {},
		children: /* @__PURE__ */ jsx("section", {
			className: classMapToString(classes),
			style: additionalStyles,
			dangerouslySetInnerHTML: renderedContent
		})
	});
});

//#endregion
export { Text as default };
