// @copilotkit/a2ui-renderer/dist/react-renderer/components/content/Video.mjs
import { classMapToString, stylesToObject } from "../../theme/utils.mjs";
import "../../lib/utils.mjs";
import { useA2UIComponent } from "../../hooks/useA2UIComponent.mjs";
import { memo } from "react";
import { jsx } from "react/jsx-runtime";

//#region src/react-renderer/components/content/Video.tsx
/**
* Check if a URL is a YouTube URL and extract the video ID.
*/
function getYouTubeVideoId(url) {
	for (const pattern of [/(?:youtube\.com\/watch\?v=|youtu\.be\/|youtube\.com\/embed\/)([^&\s?]+)/]) {
		const match = url.match(pattern);
		if (match && match.length > 1) return match[1];
	}
	return null;
}
/**
* Video component - renders a video player.
*
* Supports regular video URLs and YouTube URLs (renders as embedded iframe).
*/
const Video = memo(function Video({ node, surfaceId }) {
	const { theme, resolveString } = useA2UIComponent(node, surfaceId);
	const props = node.properties;
	const url = resolveString(props.url);
	if (!url) return null;
	const youtubeId = getYouTubeVideoId(url);
	return /* @__PURE__ */ jsx("div", {
		className: "a2ui-video",
		style: node.weight !== void 0 ? { "--weight": node.weight } : {},
		children: /* @__PURE__ */ jsx("section", {
			className: classMapToString(theme.components.Video),
			style: stylesToObject(theme.additionalStyles?.Video),
			children: youtubeId ? /* @__PURE__ */ jsx("iframe", {
				src: `https://www.youtube.com/embed/${youtubeId}`,
				title: "YouTube video player",
				allow: "accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture",
				allowFullScreen: true,
				style: {
					border: "none",
					width: "100%",
					aspectRatio: "16/9"
				}
			}) : /* @__PURE__ */ jsx("video", {
				src: url,
				controls: true
			})
		})
	});
});

//#endregion
export { Video as default };
