// @copilotkit/a2ui-renderer/dist/react-renderer/components/content/Video.cjs
const require_runtime = require('../../../_virtual/_rolldown/runtime.cjs');
const require_utils = require('../../theme/utils.cjs');
require('../../lib/utils.cjs');
const require_useA2UIComponent = require('../../hooks/useA2UIComponent.cjs');
let react = require("react");
let react_jsx_runtime = require("react/jsx-runtime");

//#region src/react-renderer/components/content/Video.tsx
/**
* Check if a URL is a YouTube URL and extract the video ID.
*/
function getYouTubeVideoId(url) {
	for (const pattern of [/(?:youtube\.com\/watch\?v=|youtu\.be\/|youtube\.com\/embed\/)([^&\s?]+)/]) {
		const match = url.match(pattern);
		if (match && match.length > 1) return match[1];
	}
	return null;
}
/**
* Video component - renders a video player.
*
* Supports regular video URLs and YouTube URLs (renders as embedded iframe).
*/
const Video = (0, react.memo)(function Video({ node, surfaceId }) {
	const { theme, resolveString } = require_useA2UIComponent.useA2UIComponent(node, surfaceId);
	const props = node.properties;
	const url = resolveString(props.url);
	if (!url) return null;
	const youtubeId = getYouTubeVideoId(url);
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
		className: "a2ui-video",
		style: node.weight !== void 0 ? { "--weight": node.weight } : {},
		children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)("section", {
			className: require_utils.classMapToString(theme.components.Video),
			style: require_utils.stylesToObject(theme.additionalStyles?.Video),
			children: youtubeId ? /* @__PURE__ */ (0, react_jsx_runtime.jsx)("iframe", {
				src: `https://www.youtube.com/embed/${youtubeId}`,
				title: "YouTube video player",
				allow: "accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture",
				allowFullScreen: true,
				style: {
					border: "none",
					width: "100%",
					aspectRatio: "16/9"
				}
			}) : /* @__PURE__ */ (0, react_jsx_runtime.jsx)("video", {
				src: url,
				controls: true
			})
		})
	});
});

//#endregion
exports.default = Video;
