// @copilotkit/a2ui-renderer/dist/react-renderer/lib/utils.cjs
const require_runtime = require('../../_virtual/_rolldown/runtime.cjs');
const require_utils = require('../theme/utils.cjs');
let _a2ui_lit_0_8 = require("@a2ui/lit/0.8");
let clsx = require("clsx");

//#region src/react-renderer/lib/utils.ts
/**
* Utility function to merge class names.
* Combines clsx for conditional classes.
*
* @param inputs - Class values to merge
* @returns Merged class name string
*
* @example
* cn('base-class', condition && 'conditional-class', { 'object-class': true })
*/
function cn(...inputs) {
	return (0, clsx.clsx)(inputs);
}
/**
* Merges multiple class maps into a single class map.
* Uses Lit's Styles.merge() function directly for consistency.
*
* Lit's merge handles prefix conflicts: if you have 'layout-p-2' and 'layout-p-4',
* only the latter is kept (same prefix 'layout-p-' means they conflict).
*
* @param maps - Class maps to merge
* @returns A merged class map
*/
function mergeClassMaps(...maps) {
	const validMaps = maps.filter((m) => m !== void 0);
	if (validMaps.length === 0) return {};
	return _a2ui_lit_0_8.Styles.merge(...validMaps);
}

//#endregion
exports.cn = cn;
exports.mergeClassMaps = mergeClassMaps;
