// @copilotkitnext/react/dist/providers/CopilotChatConfigurationProvider.cjs
const require_runtime = require('../_virtual/_rolldown/runtime.cjs');
let react = require("react");
react = require_runtime.__toESM(react);
let _copilotkitnext_shared = require("@copilotkitnext/shared");
let react_jsx_runtime = require("react/jsx-runtime");

//#region src/providers/CopilotChatConfigurationProvider.tsx
const CopilotChatDefaultLabels = {
	chatInputPlaceholder: "Type a message...",
	chatInputToolbarStartTranscribeButtonLabel: "Transcribe",
	chatInputToolbarCancelTranscribeButtonLabel: "Cancel",
	chatInputToolbarFinishTranscribeButtonLabel: "Finish",
	chatInputToolbarAddButtonLabel: "Add photos or files",
	chatInputToolbarToolsButtonLabel: "Tools",
	assistantMessageToolbarCopyCodeLabel: "Copy",
	assistantMessageToolbarCopyCodeCopiedLabel: "Copied",
	assistantMessageToolbarCopyMessageLabel: "Copy",
	assistantMessageToolbarThumbsUpLabel: "Good response",
	assistantMessageToolbarThumbsDownLabel: "Bad response",
	assistantMessageToolbarReadAloudLabel: "Read aloud",
	assistantMessageToolbarRegenerateLabel: "Regenerate",
	userMessageToolbarCopyMessageLabel: "Copy",
	userMessageToolbarEditMessageLabel: "Edit",
	chatDisclaimerText: "AI can make mistakes. Please verify important information.",
	chatToggleOpenLabel: "Open chat",
	chatToggleCloseLabel: "Close chat",
	modalHeaderTitle: "CopilotKit Chat",
	welcomeMessageText: "How can I help you today?"
};
const CopilotChatConfiguration = (0, react.createContext)(null);
const CopilotChatConfigurationProvider = ({ children, labels, agentId, threadId, isModalDefaultOpen }) => {
	const parentConfig = (0, react.useContext)(CopilotChatConfiguration);
	const mergedLabels = (0, react.useMemo)(() => ({
		...CopilotChatDefaultLabels,
		...parentConfig?.labels ?? {},
		...labels ?? {}
	}), [labels, parentConfig?.labels]);
	const resolvedAgentId = agentId ?? parentConfig?.agentId ?? _copilotkitnext_shared.DEFAULT_AGENT_ID;
	const resolvedThreadId = (0, react.useMemo)(() => {
		if (threadId) return threadId;
		if (parentConfig?.threadId) return parentConfig.threadId;
		return (0, _copilotkitnext_shared.randomUUID)();
	}, [threadId, parentConfig?.threadId]);
	const [internalModalOpen, setInternalModalOpen] = (0, react.useState)(isModalDefaultOpen ?? true);
	const resolvedIsModalOpen = parentConfig?.isModalOpen ?? internalModalOpen;
	const resolvedSetModalOpen = parentConfig?.setModalOpen ?? setInternalModalOpen;
	const configurationValue = (0, react.useMemo)(() => ({
		labels: mergedLabels,
		agentId: resolvedAgentId,
		threadId: resolvedThreadId,
		isModalOpen: resolvedIsModalOpen,
		setModalOpen: resolvedSetModalOpen
	}), [
		mergedLabels,
		resolvedAgentId,
		resolvedThreadId,
		resolvedIsModalOpen,
		resolvedSetModalOpen
	]);
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)(CopilotChatConfiguration.Provider, {
		value: configurationValue,
		children
	});
};
const useCopilotChatConfiguration = () => {
	return (0, react.useContext)(CopilotChatConfiguration);
};

//#endregion
exports.CopilotChatConfigurationProvider = CopilotChatConfigurationProvider;
exports.CopilotChatDefaultLabels = CopilotChatDefaultLabels;
exports.useCopilotChatConfiguration = useCopilotChatConfiguration;
