// @copilotkitnext/react/dist/index.mjs
"use client";

import { CopilotChatConfigurationProvider, useCopilotChatConfiguration } from "./providers/CopilotChatConfigurationProvider.mjs";
import { AudioRecorderError, CopilotChatAudioRecorder } from "./components/chat/CopilotChatAudioRecorder.mjs";
import CopilotChatInput_default from "./components/chat/CopilotChatInput.mjs";
import { CopilotKitInspector } from "./components/CopilotKitInspector.mjs";
import { MCPAppsActivityContentSchema, MCPAppsActivityRenderer, MCPAppsActivityType } from "./components/MCPAppsActivityRenderer.mjs";
import "./providers/index.mjs";
import { createA2UIMessageRenderer } from "./a2ui/A2UIMessageRenderer.mjs";
import { CopilotKitCoreReact } from "./lib/react-core.mjs";
import { CopilotKitProvider, useCopilotKit } from "./providers/CopilotKitProvider.mjs";
import { useRenderToolCall } from "./hooks/use-render-tool-call.mjs";
import { useRenderCustomMessages } from "./hooks/use-render-custom-messages.mjs";
import { useRenderActivityMessage } from "./hooks/use-render-activity-message.mjs";
import { useFrontendTool } from "./hooks/use-frontend-tool.mjs";
import { useComponent } from "./hooks/use-component.mjs";
import { defineToolCallRenderer } from "./types/defineToolCallRenderer.mjs";
import { useRenderTool } from "./hooks/use-render-tool.mjs";
import { useDefaultRenderTool } from "./hooks/use-default-render-tool.mjs";
import { useHumanInTheLoop } from "./hooks/use-human-in-the-loop.mjs";
import { UseAgentUpdate, useAgent } from "./hooks/use-agent.mjs";
import { useAgentContext } from "./hooks/use-agent-context.mjs";
import { useSuggestions } from "./hooks/use-suggestions.mjs";
import { useConfigureSuggestions } from "./hooks/use-configure-suggestions.mjs";
import { useInterrupt } from "./hooks/use-interrupt.mjs";
import "./hooks/index.mjs";
import CopilotChatToolCallsView from "./components/chat/CopilotChatToolCallsView.mjs";
import CopilotChatAssistantMessage_default from "./components/chat/CopilotChatAssistantMessage.mjs";
import CopilotChatUserMessage_default from "./components/chat/CopilotChatUserMessage.mjs";
import CopilotChatReasoningMessage_default from "./components/chat/CopilotChatReasoningMessage.mjs";
import CopilotChatSuggestionPill from "./components/chat/CopilotChatSuggestionPill.mjs";
import CopilotChatSuggestionView from "./components/chat/CopilotChatSuggestionView.mjs";
import CopilotChatMessageView from "./components/chat/CopilotChatMessageView.mjs";
import CopilotChatView_default from "./components/chat/CopilotChatView.mjs";
import { CopilotChat } from "./components/chat/CopilotChat.mjs";
import CopilotChatToggleButton, { DefaultCloseIcon, DefaultOpenIcon } from "./components/chat/CopilotChatToggleButton.mjs";
import { CopilotModalHeader } from "./components/chat/CopilotModalHeader.mjs";
import { CopilotSidebarView } from "./components/chat/CopilotSidebarView.mjs";
import { CopilotPopupView } from "./components/chat/CopilotPopupView.mjs";
import CopilotSidebar from "./components/chat/CopilotSidebar.mjs";
import CopilotPopup from "./components/chat/CopilotPopup.mjs";
import { WildcardToolCallRender } from "./components/WildcardToolCallRender.mjs";
import "./components/index.mjs";
import "./types/index.mjs";

export * from "@ag-ui/client"

export { AudioRecorderError, CopilotChat, CopilotChatAssistantMessage_default as CopilotChatAssistantMessage, CopilotChatAudioRecorder, CopilotChatConfigurationProvider, CopilotChatInput_default as CopilotChatInput, CopilotChatMessageView, CopilotChatReasoningMessage_default as CopilotChatReasoningMessage, CopilotChatSuggestionPill, CopilotChatSuggestionView, CopilotChatToggleButton, DefaultCloseIcon as CopilotChatToggleButtonCloseIcon, DefaultOpenIcon as CopilotChatToggleButtonOpenIcon, CopilotChatToolCallsView, CopilotChatUserMessage_default as CopilotChatUserMessage, CopilotChatView_default as CopilotChatView, CopilotKitCoreReact, CopilotKitInspector, CopilotKitProvider, CopilotModalHeader, CopilotPopup, CopilotPopupView, CopilotSidebar, CopilotSidebarView, MCPAppsActivityContentSchema, MCPAppsActivityRenderer, MCPAppsActivityType, UseAgentUpdate, WildcardToolCallRender, createA2UIMessageRenderer, defineToolCallRenderer, useAgent, useAgentContext, useComponent, useConfigureSuggestions, useCopilotChatConfiguration, useCopilotKit, useDefaultRenderTool, useFrontendTool, useHumanInTheLoop, useInterrupt, useRenderActivityMessage, useRenderCustomMessages, useRenderTool, useRenderToolCall, useSuggestions };
