// @copilotkitnext/react/dist/index.cjs
"use client";

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
const require_CopilotChatConfigurationProvider = require('./providers/CopilotChatConfigurationProvider.cjs');
const require_CopilotChatAudioRecorder = require('./components/chat/CopilotChatAudioRecorder.cjs');
const require_CopilotChatInput = require('./components/chat/CopilotChatInput.cjs');
const require_CopilotKitInspector = require('./components/CopilotKitInspector.cjs');
const require_MCPAppsActivityRenderer = require('./components/MCPAppsActivityRenderer.cjs');
require('./providers/index.cjs');
const require_A2UIMessageRenderer = require('./a2ui/A2UIMessageRenderer.cjs');
const require_react_core = require('./lib/react-core.cjs');
const require_CopilotKitProvider = require('./providers/CopilotKitProvider.cjs');
const require_use_render_tool_call = require('./hooks/use-render-tool-call.cjs');
const require_use_render_custom_messages = require('./hooks/use-render-custom-messages.cjs');
const require_use_render_activity_message = require('./hooks/use-render-activity-message.cjs');
const require_use_frontend_tool = require('./hooks/use-frontend-tool.cjs');
const require_use_component = require('./hooks/use-component.cjs');
const require_defineToolCallRenderer = require('./types/defineToolCallRenderer.cjs');
const require_use_render_tool = require('./hooks/use-render-tool.cjs');
const require_use_default_render_tool = require('./hooks/use-default-render-tool.cjs');
const require_use_human_in_the_loop = require('./hooks/use-human-in-the-loop.cjs');
const require_use_agent = require('./hooks/use-agent.cjs');
const require_use_agent_context = require('./hooks/use-agent-context.cjs');
const require_use_suggestions = require('./hooks/use-suggestions.cjs');
const require_use_configure_suggestions = require('./hooks/use-configure-suggestions.cjs');
const require_use_interrupt = require('./hooks/use-interrupt.cjs');
require('./hooks/index.cjs');
const require_CopilotChatToolCallsView = require('./components/chat/CopilotChatToolCallsView.cjs');
const require_CopilotChatAssistantMessage = require('./components/chat/CopilotChatAssistantMessage.cjs');
const require_CopilotChatUserMessage = require('./components/chat/CopilotChatUserMessage.cjs');
const require_CopilotChatReasoningMessage = require('./components/chat/CopilotChatReasoningMessage.cjs');
const require_CopilotChatSuggestionPill = require('./components/chat/CopilotChatSuggestionPill.cjs');
const require_CopilotChatSuggestionView = require('./components/chat/CopilotChatSuggestionView.cjs');
const require_CopilotChatMessageView = require('./components/chat/CopilotChatMessageView.cjs');
const require_CopilotChatView = require('./components/chat/CopilotChatView.cjs');
const require_CopilotChat = require('./components/chat/CopilotChat.cjs');
const require_CopilotChatToggleButton = require('./components/chat/CopilotChatToggleButton.cjs');
const require_CopilotModalHeader = require('./components/chat/CopilotModalHeader.cjs');
const require_CopilotSidebarView = require('./components/chat/CopilotSidebarView.cjs');
const require_CopilotPopupView = require('./components/chat/CopilotPopupView.cjs');
const require_CopilotSidebar = require('./components/chat/CopilotSidebar.cjs');
const require_CopilotPopup = require('./components/chat/CopilotPopup.cjs');
const require_WildcardToolCallRender = require('./components/WildcardToolCallRender.cjs');
require('./components/index.cjs');
require('./types/index.cjs');

exports.AudioRecorderError = require_CopilotChatAudioRecorder.AudioRecorderError;
Object.defineProperty(exports, 'CopilotChat', {
  enumerable: true,
  get: function () {
    return require_CopilotChat.CopilotChat;
  }
});
exports.CopilotChatAssistantMessage = require_CopilotChatAssistantMessage;
exports.CopilotChatAudioRecorder = require_CopilotChatAudioRecorder.CopilotChatAudioRecorder;
exports.CopilotChatConfigurationProvider = require_CopilotChatConfigurationProvider.CopilotChatConfigurationProvider;
exports.CopilotChatInput = require_CopilotChatInput;
exports.CopilotChatMessageView = require_CopilotChatMessageView;
exports.CopilotChatReasoningMessage = require_CopilotChatReasoningMessage;
exports.CopilotChatSuggestionPill = require_CopilotChatSuggestionPill;
exports.CopilotChatSuggestionView = require_CopilotChatSuggestionView;
exports.CopilotChatToggleButton = require_CopilotChatToggleButton;
exports.CopilotChatToggleButtonCloseIcon = require_CopilotChatToggleButton.DefaultCloseIcon;
exports.CopilotChatToggleButtonOpenIcon = require_CopilotChatToggleButton.DefaultOpenIcon;
exports.CopilotChatToolCallsView = require_CopilotChatToolCallsView;
exports.CopilotChatUserMessage = require_CopilotChatUserMessage;
exports.CopilotChatView = require_CopilotChatView;
exports.CopilotKitCoreReact = require_react_core.CopilotKitCoreReact;
exports.CopilotKitInspector = require_CopilotKitInspector.CopilotKitInspector;
exports.CopilotKitProvider = require_CopilotKitProvider.CopilotKitProvider;
Object.defineProperty(exports, 'CopilotModalHeader', {
  enumerable: true,
  get: function () {
    return require_CopilotModalHeader.CopilotModalHeader;
  }
});
exports.CopilotPopup = require_CopilotPopup;
Object.defineProperty(exports, 'CopilotPopupView', {
  enumerable: true,
  get: function () {
    return require_CopilotPopupView.CopilotPopupView;
  }
});
exports.CopilotSidebar = require_CopilotSidebar;
Object.defineProperty(exports, 'CopilotSidebarView', {
  enumerable: true,
  get: function () {
    return require_CopilotSidebarView.CopilotSidebarView;
  }
});
exports.MCPAppsActivityContentSchema = require_MCPAppsActivityRenderer.MCPAppsActivityContentSchema;
exports.MCPAppsActivityRenderer = require_MCPAppsActivityRenderer.MCPAppsActivityRenderer;
exports.MCPAppsActivityType = require_MCPAppsActivityRenderer.MCPAppsActivityType;
exports.UseAgentUpdate = require_use_agent.UseAgentUpdate;
exports.WildcardToolCallRender = require_WildcardToolCallRender.WildcardToolCallRender;
exports.createA2UIMessageRenderer = require_A2UIMessageRenderer.createA2UIMessageRenderer;
exports.defineToolCallRenderer = require_defineToolCallRenderer.defineToolCallRenderer;
exports.useAgent = require_use_agent.useAgent;
exports.useAgentContext = require_use_agent_context.useAgentContext;
exports.useComponent = require_use_component.useComponent;
exports.useConfigureSuggestions = require_use_configure_suggestions.useConfigureSuggestions;
exports.useCopilotChatConfiguration = require_CopilotChatConfigurationProvider.useCopilotChatConfiguration;
exports.useCopilotKit = require_CopilotKitProvider.useCopilotKit;
exports.useDefaultRenderTool = require_use_default_render_tool.useDefaultRenderTool;
exports.useFrontendTool = require_use_frontend_tool.useFrontendTool;
exports.useHumanInTheLoop = require_use_human_in_the_loop.useHumanInTheLoop;
exports.useInterrupt = require_use_interrupt.useInterrupt;
exports.useRenderActivityMessage = require_use_render_activity_message.useRenderActivityMessage;
exports.useRenderCustomMessages = require_use_render_custom_messages.useRenderCustomMessages;
exports.useRenderTool = require_use_render_tool.useRenderTool;
exports.useRenderToolCall = require_use_render_tool_call.useRenderToolCall;
exports.useSuggestions = require_use_suggestions.useSuggestions;
var _ag_ui_client = require("@ag-ui/client");
Object.keys(_ag_ui_client).forEach(function (k) {
  if (k !== 'default' && !Object.prototype.hasOwnProperty.call(exports, k)) Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () { return _ag_ui_client[k]; }
  });
});
