// @copilotkit/shared/dist/index.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
const require_conditions = require('./utils/conditions.cjs');
const require_console_styling = require('./utils/console-styling.cjs');
const require_errors = require('./utils/errors.cjs');
const require_json_schema = require('./utils/json-schema.cjs');
const require_random_id = require('./utils/random-id.cjs');
const require_requests = require('./utils/requests.cjs');
const require_index = require('./utils/index.cjs');
const require_index$1 = require('./constants/index.cjs');
const require_package = require('./package.cjs');
const require_telemetry_client = require('./telemetry/telemetry-client.cjs');

//#region src/index.ts
const COPILOTKIT_VERSION = require_package.version;

//#endregion
exports.BANNER_ERROR_NAMES = require_errors.BANNER_ERROR_NAMES;
exports.COPILOTKIT_VERSION = COPILOTKIT_VERSION;
exports.COPILOT_CLOUD_API_URL = require_index$1.COPILOT_CLOUD_API_URL;
exports.COPILOT_CLOUD_CHAT_URL = require_index$1.COPILOT_CLOUD_CHAT_URL;
exports.COPILOT_CLOUD_ERROR_NAMES = require_errors.COPILOT_CLOUD_ERROR_NAMES;
exports.COPILOT_CLOUD_PUBLIC_API_KEY_HEADER = require_index$1.COPILOT_CLOUD_PUBLIC_API_KEY_HEADER;
exports.COPILOT_CLOUD_VERSION = require_index$1.COPILOT_CLOUD_VERSION;
exports.ConfigurationError = require_errors.ConfigurationError;
exports.ConsoleColors = require_console_styling.ConsoleColors;
exports.ConsoleStyles = require_console_styling.ConsoleStyles;
exports.CopilotKitAgentDiscoveryError = require_errors.CopilotKitAgentDiscoveryError;
exports.CopilotKitApiDiscoveryError = require_errors.CopilotKitApiDiscoveryError;
exports.CopilotKitError = require_errors.CopilotKitError;
exports.CopilotKitErrorCode = require_errors.CopilotKitErrorCode;
exports.CopilotKitLowLevelError = require_errors.CopilotKitLowLevelError;
exports.CopilotKitMisuseError = require_errors.CopilotKitMisuseError;
exports.CopilotKitRemoteEndpointDiscoveryError = require_errors.CopilotKitRemoteEndpointDiscoveryError;
exports.CopilotKitVersionMismatchError = require_errors.CopilotKitVersionMismatchError;
exports.ERROR_CONFIG = require_errors.ERROR_CONFIG;
exports.ERROR_NAMES = require_errors.ERROR_NAMES;
exports.ErrorVisibility = require_errors.ErrorVisibility;
exports.MissingPublicApiKeyError = require_errors.MissingPublicApiKeyError;
exports.ResolvedCopilotKitError = require_errors.ResolvedCopilotKitError;
exports.Severity = require_errors.Severity;
exports.TelemetryClient = require_telemetry_client.TelemetryClient;
exports.UpgradeRequiredError = require_errors.UpgradeRequiredError;
exports.actionParametersToJsonSchema = require_json_schema.actionParametersToJsonSchema;
exports.convertJsonSchemaToZodSchema = require_json_schema.convertJsonSchemaToZodSchema;
exports.dataToUUID = require_random_id.dataToUUID;
exports.ensureStructuredError = require_errors.ensureStructuredError;
exports.executeConditions = require_conditions.executeConditions;
exports.getPossibleVersionMismatch = require_errors.getPossibleVersionMismatch;
exports.getZodParameters = require_json_schema.getZodParameters;
exports.isMacOS = require_index.isMacOS;
exports.isStructuredCopilotKitError = require_errors.isStructuredCopilotKitError;
exports.isTelemetryDisabled = require_telemetry_client.isTelemetryDisabled;
exports.isValidUUID = require_random_id.isValidUUID;
exports.jsonSchemaToActionParameters = require_json_schema.jsonSchemaToActionParameters;
exports.logCopilotKitPlatformMessage = require_console_styling.logCopilotKitPlatformMessage;
exports.logStyled = require_console_styling.logStyled;
exports.parseJson = require_index.parseJson;
exports.publicApiKeyRequired = require_console_styling.publicApiKeyRequired;
exports.randomId = require_random_id.randomId;
exports.randomUUID = require_random_id.randomUUID;
exports.readBody = require_requests.readBody;
exports.styledConsole = require_console_styling.styledConsole;
exports.tryMap = require_index.tryMap;
